
library('png')
library('GenomicRanges')
library(pheatmap)
library(tidyverse)

library(grid)

library(ComplexHeatmap)

library(ggplot2)


#########set colorset
color_gradient_biasedBR  <- readRDS('../../../../../02.snapATAC_harmony/chromVAR_TF_specific/full12526/color_gradient_biasedBR.rds')
#plot
barplot(1:length(color_gradient_biasedBR),col=rev(color_gradient_biasedBR) )


color_set_yellowbrick.flat <- readRDS('../../../../../03.snRNA_snATAC/liger/color_set_yellowbrick.flat.rds') 
#saveRDS(color_set_yellowbrick.flat,'complexHeatmap_interactive/color_set_yellowbrick.flat.rds')

colorset1 <- color_set_yellowbrick.flat[['RdYlGn.11']]
colorset2 <- rev(color_set_yellowbrick.flat[['YlGnBu.9']]) #or this
colorset3 <- rev(color_set_yellowbrick.flat[['RdPu.9']]) #use this !
colorset4 <- color_set_yellowbrick.flat[['Spectral.8']]
colorset5 <- color_set_yellowbrick.flat[['RdGy.11']]

#colorset <- colorset1

colorset <- color_set_yellowbrick.flat[['Reds.3']]

barplot(1:length(colorset),col=colorset )

colorset_go <- rev(color_set_yellowbrick.flat[['YlGnBu.9']])
colorset_pathway <- rev(color_set_yellowbrick.flat[['RdPu.9']])

barplot(1:length(colorset_go),col=colorset_go )
barplot(1:length(colorset_pathway),col=colorset_pathway )

##color set done




###1 collect i-cistarget result

enrichTable.list <- list()


for(i in c('c6','c3', 'c9' ,'c5', 'c7' ,'c1','c8','c2','c4') ){
#for(i in c('c2') ){
    cat('do collection for cluster ', i,'\n')
    dir <- paste0('dar_',i,'_overlap_p2g_auc_0.001')
    
    if(!dir.exists(dir) ){ stop(paste('not exist dir',dir,sep=' '))}
    
    tfmotif.df <- read.table(  paste0(dir,'/icistarget/statistics.tbl'), header =TRUE, stringsAsFactors = FALSE, sep='\t', comment.char = "")
    
    map.df <- read.table(  paste0(dir,'/icistarget/input_mapped_to_icistarget_regions.bed'), header =FALSE, stringsAsFactors = FALSE, sep='\t')
    colnames(map.df) <- c('chr','start','end','id','chr_meta','start_meta','end_meta','id_meta')
    map.df <- map.df[rowSums(is.na(map.df)) == 0,]
    stopifnot(nrow(map.df) != 0)
    
    
    #c2_DAR_overlap_p2g_0.005/icistarget/PWMs/cisbp__M0945.targets.tsv
    
    stopifnot(sum(duplicated(tfmotif.df$FeatureID ) )==0)
    
    enrichTable <- data.frame()
    
    for(j in seq_along(1:nrow(tfmotif.df)) ){
        
        tfmotif.line <- unlist(tfmotif.df[j,])
        
        motifid <-  tfmotif.line['FeatureID']
        
        tfid <-  tfmotif.line['FeatureAnnotations']
        if( tfid == '' ){next}
        tfid <- strsplit(x = tfid,split = ',')[[1]]
        
        
        topTargetMeta <- tfmotif.line['TopTargetIDs']
        if( topTargetMeta == '' ){next}
        topTargetMeta <- strsplit(x = topTargetMeta,split = ';')[[1]]
        
        candiTargetMeta <- tfmotif.line['CandidateTargetIDs']
        if( candiTargetMeta == '' ){next}
        candiTargetMeta <- strsplit(x = candiTargetMeta,split = ';')[[1]]
        
        nes <- tfmotif.line['NES']
        
        if(file.exists( paste0( dir,'/icistarget/PWMs/',motifid,'.targets.tsv'   )    ) ){
            nearGene.df <- read.table(paste0( dir,'/icistarget/PWMs/',motifid,'.targets.tsv'   ),sep='\t',stringsAsFactors = FALSE,header =FALSE)
            
            stopifnot( all.equal(candiTargetMeta,nearGene.df$V2,check.attributes = FALSE))
            #nearGene <- nearGene.df$V3
        
            stopifnot(  length(nearGene.df$V2 %in% map.df$id_meta) == length(nearGene.df$V2) )
            #idx <- match(nearGene.df$V2, map.df$id_meta    )
            idx <- which(map.df$id_meta %in% nearGene.df$V2 ) #find all (include dup) then merge
            
            map.df.sel <- map.df[idx,c('id','id_meta')]
            
            #link map.df.sel and nearGene.df by merge
            colnames(nearGene.df) <- c('x','id_meta','gene')
            df.merge <- merge(x=nearGene.df,y=map.df.sel, by = 'id_meta')
            
            peaks_in <- sapply(strsplit(df.merge$id,split = "\\|",perl = TRUE) ,  function(x){x[1]} )
        
            nearGene <- paste0( peaks_in ,'|', df.merge$gene )
            
        }else{nearGene <- 'na'}
        
        
        stopifnot(  length(topTargetMeta %in% map.df$id_meta) == length(topTargetMeta) )
        
        idx <- which(map.df$id_meta %in% topTargetMeta  )
        
        peaks <- sapply(strsplit(map.df[idx,'id'],split = "\\|",perl = TRUE) ,  function(x){x[1]} )
        #warning, order information lost
        
        enrichTable <- rbind.data.frame(enrichTable, data.frame(  
            'motif' = motifid, 
            'tf' = paste(tfid,collapse = ','),
            'NES' = nes,
            'topTargetMeta' = paste(topTargetMeta,collapse = ','),
            'peaks' = paste(peaks,collapse = ','),
            'nearGene' = paste(nearGene,collapse = ',')
            
            
         ) )
        
        
    }
    
    enrichTable.list[[i]] <- enrichTable
    
    
}


saveRDS(enrichTable.list,'enrichTable.list.rds')


enr.c6 <- enrichTable.list[['c6']]
names(sort(table(enr.c6$tf),decreasing = TRUE))

head(enr.c6[,1:3],n=50)

'SP1''TP53''TEAD1''MAZ''TP63''TEAD3,TEAD2,TEAD1,TEAD4''IRF8''NFYB''TP73''TEAD4''SP2''NFYA''TEAD3''TEAD2''KLF4''E2F4''EGR1''KLF5''SP4''ZNF148''WT1''PATZ1''SP3''PURA''MECOM,PRDM16,PRDM13''STAT3''IRF9''PPARA''ZSCAN10''ZNF575''IRF4''LEF1,TCF7,TCF7L2,TCF7L1''VEZF1''SP1,SP2,SP3,SP4''RELA''SP5''FOS''NFYC''E2F6''ZNF607,ZNF189,ZNF84,ZNF23,ZNF471,ZNF16,ZNF502,ZNF429''TP63,TP53,TP73''TP63,NKX2-1,TP53,TP73''SP110''ZMAT2''PPARG''JUN''TCEAL6''FGF19''RBM17''HLCS''APEX1,EP300''EP300''ZDHHC5''MCTP2''RXRA''ZNF281''PTPMT1''IKZF1''NKX2-2''UBB''ZSCAN29''ZCCHC14''SP7''GPD1''AKR1A1''EGR2,EGR3,EGR1,EGR4''PDLIM5''NAP1L1''ZIC1''CPTP''ANXA11''RAD21''EZR''KLF13''PAX5''IRF5''IRF2''TFAP2C''TRMO''SIN3A''ZNF808''ASCL2''SMAD3''EGR3''ASCC1''HDAC2''SP1,SP3,SP4''KLF11''ZNF460''TP53,HIC1''SUPT20H''SOCS4''GTF2I''TCF12''RBM7''IRF1''SREBF2''KLF15''SPIC''NELFE''NFYB,NFYA''PGAM2''POLR2A''ZNF467''NR3C1''ESRRA''SREBF1''SIX1''PDS5A''SP1,SP3''RFX5''NR2F2''EZH2''CENPB''ZNF263''KLF6''NFYC,NFYB,NFYA''BRF1''ZNF354C,MZF1''IRF2,IRF1,IRF7''ZNF503''SIX4''ZBTB7B''EBF1''EGR2''ZNF563''ZNF740''NFYC,POLE4''AFF4''MZF1'

#'TP53''NFYB''NFYA''IRF8''TEAD1''TP63''SP1''IRF5''E2F4''NFYC,POLE4''TP73''NFYC,NFYB,NFYA''RFX5''PBX3''NFYC''NFYB,POLE3''TEAD3''EZH2''RELA''E2F1''PPARA''NR2C2''YBX1''RXRA''POLR2A''IRF9''IRF4''IRF7''FOS''TFAP2C''CEBPZ''IRF3''PPARG''FOXN4''FOXI1''SP2''TEAD4''SMAD3''LEF1,TCF7,TCF7L2,TCF7L1''ZNF513''TEAD3,TEAD2,TEAD1,TEAD4''NFYB,NFYA''HLCS''SREBF1''RAD21''CHD2''FGF19''PBX1''NR2F2''TRMO''TFAP2B''SRF''PURA''NR2F6''SIN3A''HAND1''NR3C1''RBM17''SMAP2''MTHFD1''TFAP4''PIR''PEG3''NAP1L1''TFCP2''TP63,NKX2-1,TP53,TP73''TRIM21''PAX4,TCF7L1''TBP''HNF4A''ZNF394''TAF1''CXXC1''KLF17''SOX4''CTBP2''MECOM,PRDM16,PRDM13''ZNF580''NFIB''GTF2I''TCF12''ESR2''GFI1''PRDM5''NR2F1,HNF4A,NR2F2''STAT2''ZNF676''ZNF449''UGP2''EOMES,TEAD4''ZNF274''CBFA2T2''SAP30''EP300''SIX1'


#'TP53''TP63''TP73''POLR2A''TFAP2C''KLF6''EZH2''TFAP2A''SP7''E2F1''ZFY''KLF11''KLF17''SP5''KLF2''TAF1''SP6''KLF10''MBD2''SP1''NAP1L1''AKR1A1''RAD21''TRIM21''HAND1''PURA''NXPH3''RBM17''AFF4''PLAGL1''NR2C2''PPARG''TP63,NKX2-1,TP53,TP73''TRMO''CHD1''SMAP2''TP63,TP53,TP73''POLR3A''ZNF304''SMC3''HLCS''KLF4''TFCP2''SP8''SP9''E2F4''PPARA''TFAP2B''WT1''UBB''SP2''GPD1''ZNF676''EGR3''TCF12''NFYB''MTHFD1''SP3''LUZP1''ZCCHC14''PRNP''KLF5''TEAD4''NELFE''HDAC2''KLF13''MXI1''ZNF148''RBBP5''FOXN4''SREBF2''GTF2F1''KLF14''HMGA2''CXXC1''ZFX''SETBP1''UGP2''PGAM2''ESRRA''SREBF1''ZIC1''NFYC,NFYB,NFYA''SALL4''PIR''ZBTB7A''TCEAL6''SMAD3'


enr.c3 <- enrichTable.list[['c3']]
names(sort(table(enr.c3$tf),decreasing = TRUE))
head(enr.c3[,1:3],n=50)

'TP53''TEAD3,TEAD2,TEAD1,TEAD4''HNF4A''TEAD1''TP63''TP73''EP300''TEAD3''GRHL1''HLF''NR2C2''TEAD4''LEF1,TCF7,TCF7L2,TCF7L1''RXRA''PPARG''DBP''GRHL2''TEF''PPARA''STAT3''ERG''IRF9''TCF7L1''HNF4G''TP63,NKX2-1,TP53,TP73''TP63,TP53,TP73''PURA''FGF19''APEX1,EP300''JUN''NR2F6''MYB,TP53''NKX2-2''NR2F2''TFAP2C''HLCS''MECOM,PRDM16,PRDM13''TRMO''SP110''RBM7''PAX5''NFIL3''ZMAT2''PURA,TFAP2C''TP53,HIC1''RAD21''IRX4,TP53,TP73,TP63''MCTP2''ZDHHC5''TFAP2B''SIX1''CPTP''RBM17''TCF3''HNF4A,HNF4G,NR2F1,RXRA,NR2F2,PPARG''ANXA11''SMAD3''TCF12''PLG''PPARG,PPARD,RXRA,PPARA''TCEAL6''NAP1L1''GTF2I''EZH2''GRHPR''SOCS4''TEAD2''KLF4''PGAM2''CRX''ASCC1''ETS2''MTHFD1''ZNF43''RFX5''RELA''EN1''PDLIM5''ZNF460''SP7''GPD1''SPIB''CTBP2''AKR1A1''NR2F1,HNF4A,NR2F2''HSF4''LYL1''SP2''RBBP5''ZBTB33''IRF8''SIX4''TFAP4''SP1''IKZF1''E2F4''PTPMT1''TEAD4,HES7'

#'SP1''SP1,SP2,SP3,SP4''TFAP2C''TP63''TP53''EZH2''TP73''RXRA''TFAP2A''TEAD1''NR2C2''POLR2A''ZNF513''PPARA''MTF1''TFAP2B''GCM1,FIGLA''SMAD3''TEAD3''HNF4A''TCF12''PRDM5''E2F1''KLF4''TEAD3,TEAD2,TEAD1,TEAD4''SP5''IRF5''KLF11''SP7''KLF10''RAD21''FGF19''HLCS''SMAP2''PURA''MTHFD1''UGP2''MECOM,PRDM16,PRDM13''PEG3''TFCP2''RBM17''TRMO''TFAP4''TP63,NKX2-1,TP53,TP73''NR2F2''GTF2I''TRIM21''HAND1''PIR''TOB2''SALL4''NAP1L1''ESR2''SP110''NXPH3''KLF6''SMC3''IRF9''ZNF557''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''RELA''ZDHHC5''NANOG''GLIS3''ZBTB33''KLF8''AKR1A1''PPARG''EBF1''PLAGL1''TBP''HNF1B''NEUROD1''ZMAT2''ZIC2''SP9''SP8''ZNF148''TEAD4''FOXN4''NFIB''SOCS4''PDS5A''AFF4''SREBF1''NR3C1''KLF13''ZXDB''ZXDA''RARA''ZNF43''APEX1,EP300''NR2F6''PAX4,TCF7L1''IRF4''SP4''TFDP1''SIX1''RAX''ZNF180''RXRB''EGR2,EGR3,EGR1''NR2F1,HNF4G,HNF4A,NR2F2''ETS1''ZFP92''SP2''MCTP2''TAF1''SREBF2''POLR3A''WT1''IRF1''TEAD2''GRHPR''SOX4''VPS72''E2F6''HSF4''NR1H4''ZNF202''ZNF676''EOMES,TEAD4''SP3''STAT1'


#'TP53''SP1''TP73''TP63''POLR2A''KLF6''SP1,SP2,SP3,SP4''TFAP2C''KLF4''TFAP2A''SP7''MAZ''SP3''WT1''KLF11''TCF12''KLF5''ZFX''EZH2''SP5''E2F1''KLF17''KLF2''ZNF676''KLF10''MECP2''MBD2''KLF7''KLF13''AKR1A1''AFF4''TRIM21''TRMO''NAP1L1''PURA''SMAP2''NXPH3''SMC3''ZCCHC14''SALL4''HLCS''TFCP2''RAD21''HAND1''TP63,NKX2-1,TP53,TP73''PGAM2''TFAP2A,TFAP2C,TFAP2B''GPD1''GRHPR''RBM17''EZR''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''FGF19''CHD1''TFAP2B''MTHFD1''TEAD2''ZDHHC5''ZNF148''ZMAT2''UGP2''EBF1''MCTP2''EGR3''SETBP1''SP110''ZNF503''MECOM,PRDM16,PRDM13''HMGA2''SP8''PLAGL1''HOMEZ''SP9''BRF1''TAGLN2''ZNF672''ZSCAN10''PPARG''PRNP''PIR''MTF1''POLR3A''ZC3H11A''KLF14''NR2C2''FOXN4''TAF1''RARA''ZIC1''SP2''NR2F2''ZBTB14''HNF4A''SUZ12''SP4''UBB''KLF8''ZFP92''ZFY''SMARCA4''TCEAL6''ZNF281''HINFP''SREBF1''EGR2,EGR3,EGR1,EGR4,WT1'


enr.c9 <- enrichTable.list[['c9']]
names(sort(table(enr.c9$tf),decreasing = TRUE))
head(enr.c9[,1:3],n=50)

'TP53''EGR2,EGR3,EGR1,EGR4,WT1''TP63''TP73''KLF5''ZNF281''TCF7L2''PRDM1''MZF1''CEBPB''KLF4''SREBF2''ZIC1''MECOM,PRDM16,PRDM13''KLF3''KLF6''GRHL1''CEBPA''TFCP2''CEBPD''ZNF148''CHURC1''PAX5''RBPJ''MAZ''KLF1''FOXO4''STAT3''FOXO3''TP63,TP53,TP73''NAP1L1''PURA''SP7''TP53,SMAD3''PTPMT1''ASCC1''SP110''PDS5A''UBB''EZR''SALL2''BCL6''ZDHHC5''HINFP''RFX5''ZNF714''TFAP2C''HAND1''RBM17''ZNF217''PPARA''TRMO''PLG''AKR1A1''RUNX3,GABPB1''LUZP1''TBX15''ZSCAN10''ZNF341''ZNF526''AFF4''ZBTB7B''SMAD3''RAD21''SOCS4''TRIM21''GPD1''ZNF354C,MZF1''CHD1''MCTP2''FGF19''GATA3,TEAD4''E2F1''TAF1''TP53,HIC1''KLF2''PROX1''ZNF503''ZCCHC14''ETV5,CEBPD''SP2''ZIC2''IRF2''KLF16''ZNF607,ZNF189,ZNF84,ZNF23,ZNF471,ZNF16,ZNF502,ZNF429''PDLIM5''NXPH3''NANOS1''ASCL2''ZNF134''MTHFD1''TP53,ZBTB14''SOX4''ZNF579''ZMIZ1''ERF,DLX3''OVOL2''TEF''SP1''POLR2A''ZBTB33''RELA''MYB,TP53''GTF2I''CPSF4''ZNF460''RCOR1''SPIB''SP4''SALL4''EP300''ZNF418''MECP2''RFX5,ETV2''OTUD4''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA,PAX4''PPARG''SPDEF,TEAD4'

#'NFE2L2''CEBPB''NFE2''ZNF683''MECOM,PRDM16,PRDM13''KLF6''JUND''PRDM1''BACH1''BACH2''TP53''TP63''POLR2A''PURA''EBF1''SMARCB1''EZH2''TP73''SP6''ZNF134''ZNF652''CEBPD''CEBPE''BRF2''JUN''SIN3A''HIF1A''KLF4''EZR''RAD21''FGF19''HSF4''HLCS''PLG''POLR3A''HNF1B''APEX1,EP300''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''RELA''SMC3''SP110''NAP1L1''PLAGL1''PAX5''RBM7''TCF12''HAND1''SP3''SMAD3''RBM17''ASCC1''TRIM21''BCL11A''ETV2,SOX15''GCM1,HOXA2''GTF2F1''ZNF616''SP2''ANXA11''TFCP2''AFF4''EVX1,EVX2''TRMO''UBTF''FOXR1,FOXR2,FOXN4,FOXN1,FOXN2,FOXN3''SRF''KLF5''SMARCA4''AKR1A1''ZNF554''FOSL1''TOB2''ARID3A,USF1,GTF2I''ZDHHC5''THRB''PURA,EP300''SPIB''ZFY''ZNF239''ZNF263''ZNF510''SOD1''ZIC1''ZXDA''AHR,ARNT,HIF1A,ARNT2,AHRR''ZXDB''SMAP2''HNF4A''MAFK,NFE2''ZNF83''PTPMT1''ZBTB17''ZBTB14''UGP2''RXRA''TAF1''E2F2''RBPJ,RBPJL''SMAD4''ZNF589''IRF2''CHD1''ZNF354C,MZF1''IRF4''ETV2,CEBPD''KLF1''VEZF1''UBB''PDLIM5''ZXDC''SP100''BCL6''PAX4,IKZF2''SMARCC2''E2F1''ZNF672''ZNF579''GCM1,ELK3''ZMAT2''ZNF341''ZNF503''ZNF595''ZSCAN22''MAZ''ZNF709''ZBTB2''ATF4''ZNF48''FOSL2''CEBPG''PRDM5''KLF2''FOSB,JUNB,FOSL1,JUND,FOS''RBPJ''NFE2L1,MAFG''SPI1''HOXA3''ZNF670''ETV5,CEBPD''PURA,ZBTB14''ZBTB33''ZNF621''ZNF460''GTF2I''PRDM16''ETV2,GSC2''FOS''MYCN''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA,PAX4''SP9''MECP2''SP5''CUX1'

#'SP1''KLF6''KLF7''MECOM,PRDM16,PRDM13''EZH2''KLF11''KLF13''SP5''KLF10''SP3''SP6''E2F1''KLF2''TP63''ZBTB7A''MAZ''EGR1''WT1''KLF17''FOSL2''MBD2''HNF1B''KLF15''UBB''KLF8''PTPMT1''RAD21''ZCCHC14''MXI1''CREB1''ZNF202''RELA''RBBP5''AKR1A1''AFF4''JUND''PURA''EBF1''RBM17''NAP1L1''NELFE''HLCS''POLR3A''SP110''SMC3''HNF4A''TFAP2C''HAND1''TOB2''SP2''ZNF48''EGR2,EGR3,EGR1''ERF,HES7''GCM1,ELK3''SP9''CTCF''TP53''FGF19''ATF3''TP73''SP8''GCM1,ELK1''EZR''ZNF341''TRMO''ZNF460''PLAGL1''ATF2''RBM7''HDAC2''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''KLF12''ZNF219''POLR2A''ZNF134''ZIC1''GPD1''ATF4''ZDHHC5''ZNF66''SREBF2''TFAP2A''KLF4''ZNF503''ZNF595''JUNB''SMPX''SREBF1''TBP''ZBTB14''UBTF''PAX5''CHURC1''SMAD1''SMARCB1''SP7''ZNF281''NXPH3''NFE2L2,NFE2L3,BACH2,BACH1,PURA,MAFK,NFE2,MAF,MAFB,NFE2L1,MAFF,MAFG''EGR2,EGR3,EGR1,EGR4''FOXN3''ZNF676''PLAG1''TRIM21''ZFP64''PRDM5''JUN''KLF14''GLIS2'

enr.c5 <- enrichTable.list[['c5']]
names(sort(table(enr.c5$tf),decreasing = TRUE)) #ESR, PPARG
head(enr.c5[,1:3],n=50)

'TEAD1''RFX5''TFAP2C''TEAD4''ELF5''SPDEF''SMAD3''TEAD2''TEAD3''SPI1''TEAD3,TEAD2,TEAD1,TEAD4''HNF4A''SMAD4''ELF1''TP73''ETS1''MECOM,PRDM16,PRDM13''TFAP2A''FOXC1''SMAD2''ERG''SMAD1''FLI1''TFCP2''KLF4''RXRA''SPIB''FOXP2''EHF''NAP1L1''RBM17''PURA''HAND1''FGF19''RAD21''HLCS''UBP1''ZNF557''HSF4''SMAD9,SMAD4,SMAD5,SMAD6,SMAD1,SMAD2,SMAD3''SP2''AKR1A1''ZFX''KLF16''NXPH3''NKX2-2''CEBPG''SMAP2''TRMO''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''NANOG''TRIM21''SMAD5''GRHL2''RUNX1''PIR''NR2C2''PAX5''DUS3L''MYB,ZEB1''PURA,TFAP2C''ZNF91''EVX1,EVX2''AFF4''EBF1''PLG''FOXI1''PRDM16''SMC3''KDM5A''PLAGL1''TCF12''PGR''IKZF1''KLF6''ZNF503''KLF5''HDAC1''RBM7''JUN''EZH2''BCL6''NR4A2''SOCS4''VDR,FLI1''HINFP''EGR2''WISP2''FOXO1,ELK1''FOXB1''UBB''SP110''ZNF776''RORA''KLF12''RELA''MCTP2''ZNF483''SMARCA4''KLF3''PTPMT1''TP63,NKX2-1,TP53,TP73''ZXDA''ZDHHC5''TFAP4''GPD1''ZSCAN26''RBPJ''NFIC,NFIB,NFIA''NR2F6''FOXI1,ETV5''ZBTB7A''NKX2-1''FOXI1,ERF''HNRNPH3''FIZ1''DEAF1''GRHL1''TP53''ZEB2''GLIS3''ZIC1''ESR2''APEX1,EP300''FOXK1''NR4A3'

#'ELF1''ETS1''EGR1''TEAD4''TEAD2''SPI1''KLF4''ERG''EGR2''RUNX2,RUNX3,RUNX1''INSM1''ELK1''KLF18,KLF17''RUNX3''MECOM,PRDM16,PRDM13''SPIB''RORA''RARB''EGR3''BRCA1''PURA''SMAD3''GRHL1''TEAD3,TEAD2,TEAD1,TEAD4''ZBTB33''TEAD3''PAX4''KLF12''VEZF1''GMEB2''ELF2''DPF1''ESR1''BHLHE40''ZNF354A''ZBTB45''TAF1''ELK4''ZNF263''GABPA''FOXF2''CHD2''RUNX2''TCF4''FOXA1''RXRA''NAP1L1''HLCS''RBM17''TRIM21''SMAP2''TRMO''RAD21''HNF4A''FGF19''SMAD1''HAND1''UBP1''NANOG''TFAP2C''RBM7''ESR2''SMC3''FOXI1,ELK1''EZR''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''GLIS3''AKR1A1''APEX1,EP300''HSF4''DUS3L''KDM5A''EVX1,EVX2''SMAD5''FLI1''ETV5''ZNF837''CTBP1''KLF5''ZBTB17''KLF6''ZNF23''SP110''ZNF557''NXPH3''TCF12''AFF4''PAX5''SMAD2''ZNF71''ZNF66''ZBTB2''SMARCA4''CUX1''ZNF554''RARA,RARB,RARG,NR2F1,RXRA,RXRB,NR2F2,NR1I3,NR1I2,NR1H2,NR1H3''MCTP2''AHR,ARNT2,ARNT''FHL2''IRF1,NR3C1''PURA,EP300''ZFX''ZNF445''WT1,YY1''TCEAL6''PATZ1''RELA''ESRRG''SMAD4,ELF5''EIF5A2''PRDM5''OSR2''FOXS1,FOXC2,FOXC1''PRDM15''ZDHHC5''ZNF671''SPIC''PROX1,PROX2''ZNF708''ZNF563''ETS2''ESRRB''ZNF341''TAGLN2''GRHL2''TBX21,ERF''TFAP4''IKZF1''ARNT''ZNF546''PTPMT1''GABPB1''CEBPG''GPD1''CREB1''EOMES''SMAD4''MEF2C,MEF2B,MEF2A,BORCS8-MEF2B,MEF2D''HSF1''ZNF579''NR2E1''ZNF513''ETV4''WRNIP1''ETV1''POLR2A''ZNF26''CBFB''ZMAT2''GFI1B''HSPA1L''SRF''MAZ''SP3''UGP2''ZFY''ZNF483''TFCP2''ELF1,ELF2,ELF4''MEIS3,MEIS2,MEIS1,TGIF2,TGIF2LY,TGIF2LX,TGIF1,PKNOX1,PKNOX2''EBF1''RBBP5''RORC,RORB,RORA''POLR3A''HDAC2''CTBP2''STAT1''XBP1''EGR2,EGR3,EGR1''SREBF2''TP63,NKX2-1,TP53,TP73''ZNF639''CAT''ZNF267''MAFA,IKZF1''ZNF502''TBX21'

#'ELF1''GABPA''ELK1''FLI1''ELK4''SPI1''TP63''ETV4''TFAP2C''TFAP2A''TP73''TEAD2''TEAD3,TEAD2,TEAD1,TEAD4''ELF1,ELF2,ELF4''ERG''ETS1''ZBTB33''HNF4A''ZFX''ZFY''KLF6''PAX4''ZNF711''GRHL1''SP5''GRHL2''KLF12''KLF18,KLF17''TEAD4''TEAD3''HAND1''PURA''NAP1L1''RBM17''SMAP2''RAD21''TRIM21''TFCP2''NXPH3''TRMO''FGF19''AKR1A1''SMC3''SMAD3''HLCS''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''SMAD1''PIR''ZNF257''TAGLN2''PLAGL1''RBM7''AFF4''NANOG''PRDM16''KLF4''ESR2''APEX1,EP300''MTHFD1''SPIB''CUX1''SPIC''EZR''SOD1''FEV''UGP2''MRPL2''RELA''EBF1''UBP1''PAX5''GRHPR''PEG3''ANXA11''THRA''ZNF554''FOXS1,FOXC2,FOXC1''SMPX''ETV7''ZBTB2''MZF1''KLF2''ETS2''EGR3''SP110''PURA,EP300''KLF3''HNF1B''ELF2''FOXI1,ELK1''ZNF396''MAF''CBX7''SMAD4''ETV1''AGGF1''ZDHHC5''GPD1''POLR3A''WRNIP1''GABPB1,GABPA,ELK1,ELK4''CTCF''TOB2''PGR''RARB''SMAD5''KDM5A''GCM1,FIGLA''SP8''NFE2L2,NFE2L3,BACH2,BACH1,PURA,MAFK,NFE2,MAF,MAFB,NFE2L1,MAFF,MAFG''PURA,TFAP2C''SP9''TP53''FEV,ETV2,FLI1,ETS2,ERF,ERG,GABPA,ETS1''CRTC2''RAB7A''ZNF560''ZIC1''RARG''TFAP4''ASCC1''MCTP2''DUS3L''POLR2A''LIN28B,LIN28A''PPARG''OSR1''GABPB1''NR1H3''EGR2''JUN''SIRT6''ZNF217''HSF4''RARA''TCF12''ZNF521''KLF10''TFAP2E''EP300''RXRA''FOXN3'


enr.c7 <- enrichTable.list[['c7']]
names(sort(table(enr.c7$tf),decreasing = TRUE))
head(enr.c7[,1:3],n=50)


'MEF2A''TP53''SRF''RELA''TP73''TP63''TEAD4''SNAI2,SNAI3,SNAI1''TEAD3,TEAD2,TEAD1,TEAD4''TEAD1''MYOD1''TCF12''TFAP2C''MEF2D''FOXQ1,FOXJ1,FOXJ2,FOXJ3,FOXK2,FOXK1,FOXG1,FOXL2,FOXF1,FOXF2,FOXI1,FOXI3,FOXI2''MEF2C,MEF2B,MEF2A,BORCS8-MEF2B,MEF2D''MYOG''TEAD3''E2F1''TFAP2A''ASCL2''NR3C1''E2F2''FIGLA''ZEB1''FLI1''FOXG1''PRDM1''TCF4,TCF3,TCF12''MEF2C''VDR''PAX7,PAX3''ETV3''TP63,TP53,TP73''TP63,NKX2-1,TP53,TP73''EP300''PURA,TFAP2C''IRF3''HSF4''TEAD2''TP53,HIC1''HDAC2''TP53,SMAD3''HSFY1''TCF7''ZBTB14''TFAP2B''ESRRA''ERG''SMAD4''HSFY2''SPIB,TEAD4''PURA''ZNF160''ZNF582''PLG''MEF2C,MEF2B,MEF2A,MEF2D''SOX13''PTF1A''SMAD3''E2F3,HES7''ZNF454''ETV2,PITX1''REL''EGR3''SREBF2,SREBF1,ZNF354C''IKZF1''ZNF8''FOXF1,FOXF2''NFKB1,BCL3''EGR1''ELF3''SMAD4,NR3C1''EGR4''FOXD3''WISP2''NR1D1''SOX5''MBD1''ZNF404''APEX1,EP300''FOXI1,FLI1''LEF1''MEF2C,MEF2A,MEF2D''THAP1''NFKB1''TFCP2''KLF5''ZNF519''SP7''MESP1''IRF1,LTF''MEF2B''WT1''ID4''FOXA3'

#'RELA''SRF''SP1''PITX2,PITX3,PITX1''OTX2''MYBL2''GSC,GSC2''TP73''NFKB1''SP1,SP2,SP3,SP4''IRF5''NFKB2''TEAD4''MECOM,PRDM16,PRDM13''MYBL1''POLR2A''OTX1''TP53''PLAGL2''NFKB1,NFKB2''PITX1''PITX2''NFYA''EGR1''ZNF350''RHOXF2''IRX5,IRX4,IRX6,IRX1,IRX3,IRX2''PITX3''ZNF622''ZSCAN2''TEAD1''ZNF274''CXXC1''GCM1,PITX1''SP2''MYF6''ZNF628''SP1,SP2''POLR3A''TBX20''ZNF716''EGR4''PBX3''IRF1''TFAP2D''HAND1''ZFX''ETV5,CLOCK''ZNF256''ZBTB8B''HSF4''NFKB1,BCL3''STAT1''ZNF23''WRNIP1''UBB''REL''ZNF404''NFKB1,NFKB2,RELA,POLD2,REL''NFYB''GCM1,FIGLA''ETV2,SOX15''MYB''TP63''ZFY''ZNF202''RBM17''ETV7''NR2C2''SREBF2,ERF''TFCP2''ZNF546''PAX5''HCFC2''NR1D1''ZNF134''TBP''ZBTB18''TFAP2C''TRIM28''ZNF48''SP4''RUNX3,GABPB1''ELF1,MYBL1''ZNF486''NR3C1''PRDM5''RXRA''ELF3''HINFP''AKR1A1''ETV2,FIGLA''WISP2''ZKSCAN3''HMBOX1''KLF15''ELK1,ETV7''NKX2-5''MAZ''OVOL3,OVOL2,OVOL1''KLF7''BCL3''ZNF516''HOXB2,RFX5''ZNF729''ELK1''KLF12''SETDB1''CENPB''ZNF555''CRX''NFKB1,NFKB2,RELA,REL,RELB''HLCS''ETS1''KLF5''ZNF382''FOXM1''ZNF709''KDM5B''MCTP2''TAF7,STAT6,TBP''TP63,TP53,TP73''ZNF44''IRF6''RCOR1''ELF3,POU5F1''ATF1'

#'SP1''EGR2,EGR3,EGR1,EGR4,WT1''TP53''SP1,SP2,SP3,SP4''SP2''TP63''ZNF740''EGR1''TP73''SP4''MZF1''MAZ''ZFY''NFYB''NFYA''SP8,SP9,SP6,SP7''KLF11''ZNF23''EZH2''ZNF202''CHURC1''IRF1''SP1,SP2''PBX3''POLR3A''PLAGL1''FOS''E2F4''ZNF281''KLF12''WT1''SP1,SP3,SP4''ZNF48''SP3''NELFE''ZBTB12''PRNP''ESRRA''ZNF563''TFAP2C''CXXC1''OVOL2''TBP''TEF''RAD21''ZBTB7B''KLF3''SREBF2,WT1''TAF1''ZDHHC5''SREBF2''PURA''KLF15''SP7''SP100''LUZP1''ZFX''TRMO''ZNF546''NFKB2''TBX15''SMAD3''SMARCA4''KLF16''ZNF789''SREBF1''NFIL3''SMC3''KDM2B''ZNF716''GPD1''KLF10''ZNF711''ZMIZ1''ZBTB14''KLF13,KLF12,KLF11,KLF10,KLF17,KLF16,KLF15,KLF7,KLF6,KLF5,KLF4,KLF3,KLF2,KLF1,KLF9,KLF8''RBBP5''NANOG''HDAC2''KLF6''GRHL1,GRHL2,GRHL3''ZNF219''STAT1''ZNF460''MECOM,PRDM16,PRDM13''HOXA2''ETV3''KLF7''PGAM2''ZNF148''ARID5B'



enr.c1 <- enrichTable.list[['c1']]
names(sort(table(enr.c1$tf),decreasing = TRUE))
head(enr.c2[,1:3],n=20)

'JUN''JUND''FOS''FOSL2''JUNB''FOSL1''JDP2''CEBPB''KLF5''HNF4A''BACH1''NFE2''JUN,FOS''BACH2''ATF3''EP300''FOSB''RXRA''CEBPD''SMARCC1''STAT3''KLF1''TFCP2''RORA''FOS,FOSB,JUNB,JUN,JUND,FOSL1,FOSL2''CEBPA''KLF4''HNF4G''PPARG''MXI1''ATF7''RCOR1''GRHL1''TCF12''JUNB,JUN,JUND''BATF''MAF''NR2C2''EPAS1''PURA''NR2F6''ZEB1''SREBF2''NFE2L2''ZNF214,ZNF222,ZNF223,ZNF224,ZNF239,ZNF226,ZNF221,ZNF234,ZNF664,ZNF235,ZNF155,ZNF233,ZNF230,ZNF284,ZNF285,ZNF112,ZNF227''TCF7L2''ZNF554''FOSB,JUNB,FOSL1,JUND,FOS''SIN3A''MYC''TFAP2C''SMARCB1''MEF2A''DUS3L''GATA2''BATF,FOS,JUNB,JUN,JUND,FOSL1,FOSL2''NFE2L2,BACH2,BACH1,MAFK,NFE2,MAF,MAFB,NRF1,NFE2L1,MAFG''NAP1L1''NR1H4''FOS,FOSB,JUNB,JUN,JUND,FOSL1''BCL3''EVX1,EVX2''PAX5''FGF19''ZFP62''EIF5A2''NXPH3''ZNF503''SMAD3''KLF8''MEF2C''KDM5A''RBM17''NR2F1,HNF4A,NR2F2''POLR3A''CRTC2''TRIM21''AFF4''SALL4''KLF17''RAB7A''ZNF384''ZBTB14,YY2''RELA''HLCS''HAND1''PDS5A''KLF6''CELF6''PPARG,PPARD,RXRA,PPARA''AKR1A1''PIR''PPARD''MCTP2''HNF4A,HNF4G,NR2F1,RXRA,NR2F2,PPARG''KLF7,KLF6''SMARCA4''ZBTB7B''ZBTB7A''ZNF708''CEBPG''TERF1''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''ZNF660''NR2F2''NFIC''PPARGC1A''ZFY''ZFX''KLF12''SMAP2''FHL2''MAFB''FLI1,TCF3''RAD21''MECOM,PRDM16,PRDM13''SRF''SOCS4''PPARA''RREB1''ATF2''ZNF658''ZNF658B''VDR''ARNT2,ARNT''SREBF1''RFX1''AVEN''TBX2''UGP2''MYBL1,MYB,MYBL2,SNAPC4''CREM''RARA''NR4A3''SREBF2,SREBF1''HLF''NR2F1''MEIS2''THAP1''ZBTB49''FOXO3''ESRRA''TP53''ATF1'

#'JUND''JUN''FOSL2''ZNF554''FOS''JUNB''FOSL1''KLF4''BACH1''KLF5''NFE2''BACH2''MYBL2''JDP2''ATF3''MEIS2''TFAP2C''SREBF2''FOS,FOSB,JUNB,JUN,JUND,FOSL1,FOSL2''SP1''KLF1''RXRA''SMARCC1''RCOR1''JUN,FOS''ATF2''OVOL2''SMAD3''PURA''SMC3''MECOM,PRDM16,PRDM13''THAP1''PAX5''SMARCB1''TFCP2''ZXDA''ZXDB''PLAGL1''KLF3''ZBTB7A''SPI1''STAT3''MAZ''NFE2L2''MAF''RELA''CUX1''MZF1''ASCL2,ASCL1''BATF''ATF7''SMAD1''FOSB''NAP1L1''EVX1,EVX2''HNF4A''RBM17''FGF19''DUS3L''TRIM21''NXPH3''SMAP2''HLCS''RAD21''AKR1A1''AFF4''SP110''CELF6''TCF12''CERS4''FOSB,JUNB,FOSL1,JUND,FOS''HAND1''MTHFD1''MCTP2''ZNF503''FOXN3''ZCCHC17''KLF8''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''KDM5A''ZNF28''ZNF430''GPD1''TEAD2''ZNF671''NR0B1,SREBF1''SOD1''KLF13,KLF12,KLF11,KLF10,KLF17,KLF16,KLF15,KLF7,KLF6,KLF5,KLF4,KLF3,KLF2,KLF1,KLF9,KLF8''TCF4,TCF3,TCF12''FHL2''APEX1,EP300''SOCS4''ELK1''ZNF780A''ERG''GPAM''EBF1''KLF6''RAB7A''SPIC''TCF7L2''ZNF583''ZNF148''ZNF770''BATF,FOS,JUNB,JUN,JUND,FOSL1,FOSL2''PDS5A''UBB''ZNF692''NKX2-2''KLF2''CHD1''VEZF1''ZNF711''TRMO''PIR''SALL4''PLG''RORB''RBM7''SPIB''FOS,FOSB,JUNB,JUN,JUND,FOSL1''LYL1''ZBTB41''MYC''NR1H4,RXRA,MZF1''ZBTB7B''ATF4''ZFX''HIVEP2''MAFB''RELB,RELA,REL''STAT6,STAT4,STAT2''YY1,MZF1''HDAC1''PGAM2''NR2C2''STAT1''PRDM16''ZNF846''POLR3A''RARG''EBF1,E2F1''ZXDC''ETV7''EZH2''RXRG''PATZ1''NR1D1''PRDM10''ZNF814''RORA''TGIF2LY,TGIF2LX,TGIF2,TGIF1''ETS1''ATF1''HNRNPH3''MESP1''HINFP''SREBF2,SREBF1,ZNF354C''ZSCAN20''NR4A2''ZNF480''EP300''MYBL1''SREBF2,SREBF1''HSF4''TFAP2B''UGP2''NFE4''AHR,ARNT,HIF1A,ARNT2,AHRR''ZNF682''MEIS1''AGGF1''MYBL1,MYB,MYBL2,SNAPC4''CHURC1''NFATC3''GMEB2''EIF5A2''SMAD4''SMAD2''ZNF816''ELF2''TFAP2A''ZNF423''UBP1''SREBF1'

#'KLF4''JUN''JUND''FOS''JUNB''FOSL2''KLF1''FOSL1''KLF5''TFAP2A''KLF12''MECOM,PRDM16,PRDM13''KLF3''KLF2''ATF3''ZFY''SREBF1''TFAP2C''PURA''SMARCC1''SP1''FOSB''JDP2''KLF8''ELF1''SNAI2,SNAI3,SNAI1''ETV4''KLF6''ZNF148''PLAGL1''ELK1''TCF12''FOS,FOSB,JUNB,JUN,JUND,FOSL1,FOSL2''TFCP2''SMC3''ZNF711''TFAP2B''MZF1''JUN,FOS''PAX4''THAP1''NFE2''SREBF2''ZNF524''FLI1''KLF12,KLF3,KLF8''RBM17''NAP1L1''HNF4A''SMAD3''SMARCB1''FGF19''HAND1''AKR1A1''NXPH3''AFF4''HLCS''TRIM21''SP110''ZNF503''RAD21''MCTP2''ETS2''SALL4''SPI1''GPD1''FOSB,JUNB,FOSL1,JUND,FOS''SOCS4''NANOG''SMAP2''NFE2L2,NFE2L3,BACH2,BACH1,PURA,MAFK,NFE2,MAF,MAFB,NFE2L1,MAFF,MAFG''TEAD2''TCEAL6''PIR''ZNF671''RELA''FOXN3''RBM7''ETV5''PDS5A''DUS3L''GLIS3''POLR2A''PURA,TFAP2C''POLR3A''EVX1,EVX2''NFE2L2,BACH2,BACH1,MAFK,NFE2,MAF,MAFB,NRF1,NFE2L1,MAFG''SALL2''APEX1,EP300''FHL2''PRDM16''HDAC1''VEZF1''TBX21''EIF5A2''ZNF554''ZDHHC5''ZNF100''ZFX''CUX1''ZNF202''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA''TAGLN2''PURA,ZBTB14''SMAD4''EBF1''AGGF1''ETV7''ZSCAN10''TBX3''BACH1''EZR''ZNF345''SMARCC2''ZXDB''PTPMT1''NFE4''HSF4''CEBPG''TRMO''RXRA,PPARA''KLF13,KLF12,KLF11,KLF10,KLF17,KLF16,KLF15,KLF7,KLF6,KLF5,KLF4,KLF3,KLF2,KLF1,KLF9,KLF8''ETV7,ELK1,FLI1,ETS2,ERF,ERG,ELK4,ELF1,ELF2,ETS1''BATF,FOS,JUNB,JUN,JUND,FOSL1,FOSL2''ZCCHC17''NR2C2''NELFE''BATF''ZSCAN31''GRHPR''UBB''ZIC5''ZSCAN26''TOB2''PEG3''ZBTB2''CHD1''RFX1''ZIC1''FEV,ETV2,FLI1,ETS2,ERF,ERG,GABPA,ETS1''ELF5''THRB''RXRA''ETS1''RELB,RELA,REL''FIZ1''CREB1''MXI1''HSPA1L''SMPX''RFX5,ETV2''SMAD5''ASCL2''ETV6''GABPA''PAX5''SPIB,GABPB1,ETV4,ETV7,ETV6,FLI1,SPI1,ERG,PSMD12,ETS2,ELK3,ELK1,ELK4,ELF1,GABPA,ELF2,ELF4,ETS1''ZSCAN20''KLF14''SMAD1''NR0B1,SREBF1''ZBTB7B''ZNF281''ZNF831,HIVEP1,HIVEP2,HIVEP3''SMARCA4'



enr.c2 <- enrichTable.list[['c2']]
names(sort(table(enr.c2$tf),decreasing = TRUE))
head(enr.c2[,1:3],n=20)

'FOS''STAT1''JUND''JUN''FOSL2''KLF4''JUNB''STAT5A''FOSL1''STAT6''GATA2''POU4F1''STAT4''GATA5''RUNX1''JUN,FOS''GATA6,GATA4,GATA5''RUNX3''BCL6''KLF5''FOSB''ZIC1''IRF9''RUNX2''NFIB''SMARCC1''POU4F3''TEAD3,TEAD2,TEAD1,TEAD4''AR''CEBPB''STAT5B''GRHL2''RUNX2,RUNX3,CBFB,RUNX1''KLF6''ELK1,HOXD12''PROX1''ZNF345''SP2''SIN3A''ATF2''BACH1''SP1''ESRRG''GLI3''GRHL1''GATA4''ETV1,HOXD12''ZNF384''STAT3''SP3''PLAG1''TBL1XR1''SMAD3''GATA6''NR2F2''HOXA9''FOS,FOSB,JUNB,JUN,JUND,FOSL1,FOSL2''IRF3''SMAD4''GATA1''NR3C1''GATA2,GATA3,GATA1''BACH2''HDAC2''RUNX3,GABPB1''PRDM9''ZNF658''ZNF658B''ZNF813''ZNF614''TRPS1,GATA6,GATA4,GATA5,GATA2,GATA3,GATA1''EP300''TFCP2''TROVE2''E2F1,SOX17''ZNF449''ZNF217''GLIS2,GLIS3,GLIS1,GLI2,GLI3,GLI1,ZIC2,ZIC3,ZIC1''KLF1''HSFY1''SALL2''AFF4''ZBTB7A''TSC22D4''CDX1,CDX2,CDX4''CEBPB,STAT6''RXRA,PPARA''HNRNPH3''NFIC''AKR1A1''STAT6,STAT4,STAT3,STAT2,STAT1,STAT5B,STAT5A''GLI1''GPD1''ZXDA''IKZF3''HDAC1''FOS,FOSB,JUNB,JUN,JUND,FOSL1''MTHFD1''ZNF761''ATF7''KLF3''TEAD4,CLOCK''RUNX2,RUNX3,RUNX1''ZNF599''PDS5A''ZNF329''SMAD1''ZNF546''FOSB,JUNB,FOSL1,JUND,FOS''TP53''HP1BP3''ZNF354C,POU2F1''SMARCB1''ZNF461''IRX5''STAT3,STAT1''ZMAT2''DHX36''BATF,FOS,JUNB,JUN,JUND,FOSL1,FOSL2''ZNF226''SMAD9,SMAD4,SMAD5,SMAD6,SMAD1,SMAD2,SMAD3''ELF1,HOXB2''PLG''HSFY2''IKZF1''RFX5''GLI2''NFE2''RCOR1''TBX21,ERF''MORN1''JAZF1''SPATS2''NXPH3''BCL3''ZNF430''POLR2A''SPDEF,TEAD4''HOXB2,PAX1''ZNF628''GAR1''PIR''NFIC,CRX,NFIA,RAX''CHD1''ZDHHC5''CREB1''ZNF404''ZNF160''TFAP2A,TFAP2C,TFAP2B,TFAP2E,TFAP2D''SMARCA4''ELK3,HOXD12''FLI1''SPIB''TCF3''CREM''MEF2C''AGGF1''CEBPA''TOB2''RPS10''KLF8''MYLK''TBX21,ELK1''NR3C2''DBP''PGR''ZNF708''CEBPD''PARP1''BATF''JDP2''GATA2,GATA1''SETDB1''ZIC2''EBF1''SREBF2''ZNF429''ETV2,DLX3''SREBF2,ELK1''HOXD12,ETV4''EGR2''VDR''TFAP2A''RAB18''ZNF766''EGR1''ZXDB''CTCFL''GATA3''MYC''CREB1,CREM,ELF1,ATF4,ATF7,ATF1,ATF3,ATF2''MITF,TFEC,TFE3,TFEB''PAX9''NFATC1''DLX3''IRF8''ARNT2''MEIS1''PLAGL1''PAX5''NR4A3''POU1F1''ATF1'

#'STAT1''POU4F3''ZNF143''STAT3''POU1F1''POU4F1''POU3F2''POU3F1''SIX5''CEBPB''POU2F2''POU3F3''SP2''ETS1''HOXA9''STAT5A''POU4F2''EZH2''NR5A2''ESRRB''STAT4''POU3F4''SMARCC2''ESRRA''STAT5B''NR2F2''ZNF683''JUND''HOXD11''HOXC9''XBP1''PAX2''JUN''EGR2''SETDB1''THAP11''CDC5L''PAX4''SPIC''ESRRG''NFE2L1''CEBPD''KDM2A,PHF8,KDM7A,FBXL19,KDM2B,PHF2''MEIS1''KLF17''ETV7''SP6''KLF2''POU2F3''RARB''POU2F1''KLF6''TCF12''SP1''BATF''CEBPA''ELK1,POU2F1''HMGB3,HMGB2,HMGB1,HMGB4''FOSB''CEBPB,STAT6''CDX1,CDX2,CDX4''STAT3,STAT1''ZNF782''HOXA13,ATF1''TEAD1,YY1''ETV2,HES7''SMAD2''STAT6,STAT4,STAT3,STAT2,STAT1,STAT5B,STAT5A''EGR4''CPTP''ZNF83''ZNF76''ZNF675''EGR1''SCRT1,SCRT2''JDP2''NFAT5,NFATC2,NFATC3,NFATC4,NFATC1''TBX2''RPS10''THAP1''SREBF2''POU6F2,POU6F1''SRF''CRX,PURA''NR2E1''GTF2F1''PITX1,TEAD4''ZNF559''GTF3C2''HOXB2,ELK3''KLF3''GPANK1''ZSCAN29''CREB1,CREM,ELF1,ATF4,ATF7,ATF1,ATF3,ATF2''CRX,RAX,E2F1''GATA1''POLR3A''ZNF384,YY1''POLR2A''EP300''RXRA,PPARA''ZNF714''ATF2''HMGN3''ELF2''NKX3-2''HCFC1''ELK1,GATA6,GATA4,GATA5,GATA2,GATA3,GATA1''ARID3A,PAX4''POU5F1B,POU5F1''NANOG''KLF1''SREBF2,ELK1''RPS4X''CNOT4''STAT6''CDX2,JUN''MAFG''ATF3''TEF''IRX5''LHX4''STAT5B,STAT5A''BDP1''ZNF100''ZFX''JUNB''DBP''NKX2-8''HOXD4,HOXA4,HOXC4,HOXB4''TBP,TBPL1,TBPL2''ZNF180''ZNF571''PDX1,CDX2''POU6F1''ETV5,CEBPD''LHX1''SOX8''ZFY''ETV2,DLX3''SPI1,MYB''CBX7''PAX6''HMG20B''HOXD10''TBL1XR1''PBX4,HOXA10''RBPJ''THRA''UNCX''ELK1,TEF''TRPS1,GATA6,GATA4,GATA5,GATA2,GATA3,GATA1''ELF3''SALL2''ZNF281''TBX21,E2F3''ZNF286B''SCRT2''GMEB1''DBX2'

#'STAT1''CEBPB''STAT3''STAT5A''STAT5B''TBP''STAT4''NR5A2''MECP2''STAT6,STAT4,STAT3,STAT2,STAT1,STAT5B,STAT5A''POLR2A''STAT3,STAT1''KLF17''KLF2''KLF6''SP6''POU3F2''SPIB''SRF''JUND''JUNB''FOS''HMGB3,HMGB2,HMGB1,HMGB4''RXRA''POU2F3''FOSB''ZFP62''TBP,TBPL1,TBPL2''STAT6,STAT4,STAT2''POU3F4''POU1F1''ZNF714''NFIC''HOXD11''HOMEZ''POU3F3''SP100''HOXA11''SETBP1''EGR1''TLX2''KDM2B''XBP1''HOXD10''ZHX1''TEF''DMRTA2''ATF4''CEBPB,STAT6''ZNF782''TAF1''SP2''DBX2''HOXD4,HOXA4,HOXC4,HOXB4''SREBF2,ELK1''HMGA2''STAT5B,STAT5A''JDP2''HOXB6''ELF2''IRF9''ZNF571''RARB''GTF2F1''POU4F1''EGR2''HMBOX1''HLX''ATF2''IRX5''SP1,SP2''HOXD13''EGR4''SP3''ARID5B''IL21,STAT3''FOSB,FOSL2,FOSL1,FOS''HOXA2''RARG''HNF1B''CNOT4''ZNF785''ZNF675''FOSL1''ELF1,TEAD4''EP300''TCF4,MYF5,MYF6,MYOD1,ASCL1,CREB1,TFF3,TCF3,MYOG,CREM,TCF12,ATF4,ATF7,ATF1,ATF3,ATF2''HOXA9''ESRRA''RXRB''ZSCAN10''HMGN3''GCM1,SPDEF''CDX2''RARA'


enr.c8 <- enrichTable.list[['c8']]
names(sort(table(enr.c8$tf),decreasing = TRUE))
'RXRA''NR2F6''RXRG''SRF''ETS1''SIX5''ZNF143''RXRB''NR2F1''SP1''ZNF263''EZH2''PPARG''SMARCC2''KLF5''PPARD''HNF4A''CHD1''NR2C2''SP4''SP5''PPARA''HLF''MAZ''NR1H2''WT1''ATF2''TFAP2C''KLF1''FOXP2''ATF6''TAF1''THAP11''MYOD1''DBP,TEF,HLF''POLR2A''ATF1''ZNF341''SP2''KLF17''ZNF429''OVOL2,OVOL1''MAX,ERF''MAX,FLI1''HNF4G,HNF4A''HDAC1,HDAC3,HDAC2''VEZF1''ZNF823''PATZ1''ZNF281''ATF5,CREB3''ZNF76''KLF11''UNCX''KLF4''RBBP9''IRF4''PAX5''CXXC1''MZF1''HCFC1''EGR1''EZR''RBBP5''IRX6''NFIL3''DBP''SETDB1''FLI1''TBX2''SMC3''ANXA11''NANOS1''NR2F1,NR2F2''JDP2''CDX2''ZNF467''ZBTB2''RFX5''THRB''ETV2,FIGLA''ZNF814''ASCC1''CREB1''RARB''HNF4A,HNF4G,NR2F1,RXRA,NR2F2,PPARG''HOXA3''GCM1,SOX2''PGAM2''TBP''BRF1''ZNF521''MCTP2''EBF1''SP3''BCL3''HOXB6''SP1,SP3''SP1,SP3,SP4''ERF,FIGLA''ZNF577''PDLIM5''PITX1,HES7''DMRT1''ARG1,ARG2,AGMAT''RXRG,RXRA,RXRB''ZKSCAN8''TAF1,TAF1L''ELK3,HOXD12''KLF12''TCEAL6''SPIC''ZNF135''PARP1'

#'SP1''E2F4''EGR1''MAZ''SP1,SP2,SP3,SP4''MZF1''SP4''E2F6''KLF4''ZNF281''EGR2,EGR3,EGR1,EGR4,WT1''EZH2''E2F1''SP8,SP9,SP6,SP7''MECOM,PRDM16,PRDM13''CHURC1''TEAD2''POLR2A''EGR2''STAT1''MECP2''SREBF2''ZNF436''ZSCAN10''E2F3''ZNF148''SP2''KLF15''SP3''GPD1''ZDHHC5''PTPMT1''ASCC1''AGGF1''PDLIM5''AKR1A1''ZMAT2''NAP1L1''PGAM2''EZR''KLF6''RBM7''RAD21''ZNF460''AFF4''UBB''ANXA11''SP110''WT1''PURA''VEZF1''ASCL2''GTF2I''NXPH3''ZNF823''MCTP2''HAND1''ZNF503''ZCCHC14''PLAGL1''TCEAL6''RBM17''ZIC1''ZNF134''PRDM5''ETS2''ZBTB7B''NELFE''SMC3''NR1H4,RXRA,MZF1''E2F6,E2F4''PATZ1''ZBTB33''TAF1,TAF1L''TCF12''SREBF1''LIN28B''SP5''TFDP1''ZNF596''ZNF580''PPARG''HDAC2''NFE4''HNF4A''ZNF202''PAX5''SP1,SP3''MXI1''ESR1''THRB''ZNF135''SALL4''E2F7''ZNF429''SMAP2''ZNF184''ZNF697''TBP''PAX4''TFAP2C''ZNF385D''KLF5''KLF2''ZNF773''LUZP1''RBBP5''NFYB''EGR3''TAF1''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA,PAX4''DEAF1''SP1,SP2''ZNF316''ZNF208''SPIC''ZNF468''ZBTB17''ZFX''IRF4''KLF8''SREBF2,WT1''ZNF607,ZNF189,ZNF84,ZNF23,ZNF471,ZNF16,ZNF502,ZNF429''KDM5B'


enr.c4 <- enrichTable.list[['c4']]
names(sort(table(enr.c4$tf),decreasing = TRUE))
'SRF''TP53''TP63''SREBF1''TP73''STAT1''STAT6''STAT3''THRB''SREBF2''EZH2''THRA''TFCP2''RORA''IRF7''HNF4A''NR1H4,RXRA''TP63,TP53,TP73''ZNF331''SMAD3''TFCP2L1''NR1I2''MEF2C,MEF2B,MEF2A,BORCS8-MEF2B,MEF2D''LHX3''SMAD4''RHOXF1''SREBF2,SREBF1''GTF3A,ZXDB,ZXDC,ZXDA''BHLHE41''PLAG1''HDAC2''ZNF8''SMAD9,SMAD4,SMAD5,SMAD6,SMAD7,SMAD1,SMAD2,SMAD3''TCF7L2''TRIM28''NR2C2,NR2C1''RAB7A''ZBTB2''PAX5''NHLH2,NHLH1''TP63,NR3C1,TP53,TP73''MAF''CBX7''EVX1,EVX2''RPS10''CRTC2''TP53,HIC1''NR0B1,SREBF1''NUP133''RXRB''PLAGL1''DEAF1''REST''ZNF562''SMAD9,SMAD4,SMAD5,SMAD6,SMAD1,SMAD2,SMAD3''MYB,TP53''ZNF655''ZNF675''GTF2F1''NR2C2''ZNF174''PSMC2''SRY,HMGA2,HMGA1''MYF5,MYF6,TFF3,HAND1,NHLH1,MITF,MXD4,MYOG,MXD1,MYCN,MYOD1,MXD3,BHLHE40,BHLHE41,TFEB,USF2,USF1,TFE3,MYC,TCF12,TCF4,TAL1,TCF3,TAL2,ASCL1,TP53,HAND2,MXI1,MAX''NR1I3''PPARG''EP300''ZNF600''ZNF510''ZNF384''SMAD4,NR3C1''SMAD4,SMAD1,SMAD2,SMAD3''RARB''TP53,SMAD3''RXRA''ZNF180''ZNF69''GCM1,ETV4''EVX1,ETV5''ZNF682''ZNF93''RXRA,NR1H2,NR1H3''ZNF449''TCF7L1''MEF2C''STAT4''MTHFD1''EHF''MAFB''PPARA''ZNF426''RFX5''KLF15''ZFP14''RNF114''ATF6B''NR1H4,NR1H2,NR1H3''ZNF189''NR2F6''MAFA,KLRG1''ETV7''BCL6''BACH2''HDAC6''ETV7,ELK1,FLI1,ETS2,ERF,ERG,ELK4,ELF1,ELF2,ETS1''CREB3''HDAC1''XBP1''EBF1''ETS1''POLR2A''TFAP2C,SIRT6''ETS2,ETS1''ZNF341''ZNF404''BCL3''HOXB2,ETV7''ETV4''HSF1''STAT6,STAT4,STAT3,STAT2,STAT1,STAT5B,STAT5A''SPIB''ZNF133''ZNF579''FOXJ3''STAT3,STAT1''GATA6''NR1I2,RXRA,RXRB''ATF3''RARA''SIX4''MXI1''HOXB2,PAX9''BACH1''ZC3H11A''TCF7''CTCF''RORC,RORB,RORA''CREB1''FOXD1''ZNF202'


#'RXRA''NR2F6''RXRG''SRF''ETS1''SIX5''ZNF143''RXRB''NR2F1''SP1''ZNF263''EZH2''PPARG''SMARCC2''KLF5''PPARD''HNF4A''CHD1''NR2C2''SP4''SP5''PPARA''HLF''MAZ''NR1H2''WT1''ATF2''TFAP2C''KLF1''FOXP2''ATF6''TAF1''THAP11''MYOD1''DBP,TEF,HLF''POLR2A''ATF1''ZNF341''SP2''KLF17''ZNF429''OVOL2,OVOL1''MAX,ERF''MAX,FLI1''HNF4G,HNF4A''HDAC1,HDAC3,HDAC2''VEZF1''ZNF823''PATZ1''ZNF281''ATF5,CREB3''ZNF76''KLF11''UNCX''KLF4''RBBP9''IRF4''PAX5''CXXC1''MZF1''HCFC1''EGR1''EZR''RBBP5''IRX6''NFIL3''DBP''SETDB1''FLI1''TBX2''SMC3''ANXA11''NANOS1''NR2F1,NR2F2''JDP2''CDX2''ZNF467''ZBTB2''RFX5''THRB''ETV2,FIGLA''ZNF814''ASCC1''CREB1''RARB''HNF4A,HNF4G,NR2F1,RXRA,NR2F2,PPARG''HOXA3''GCM1,SOX2''PGAM2''TBP''BRF1''ZNF521''MCTP2''EBF1''SP3''BCL3''HOXB6''SP1,SP3''SP1,SP3,SP4''ERF,FIGLA''ZNF577''PDLIM5''PITX1,HES7''DMRT1''ARG1,ARG2,AGMAT''RXRG,RXRA,RXRB''ZKSCAN8''TAF1,TAF1L''ELK3,HOXD12''KLF12''TCEAL6''SPIC''ZNF135''PARP1'

#'SRF''HMGB3,HMGB2,HMGB1,HMGB4''HNF4A''KLF4''ELF5''ZEB1''EGR2,EGR3,EGR1,EGR4,WT1''SPIB''KLF5''ETS1''ZNF281''KDM4C,KDM4B,KDM4A,KDM4F,KDM4E,KDM4D''MZF1''MAZ''KLF1''EHF''FLI1''ZNF740''ZNF148''THAP1''RREB1''RAD21''NFE2''FOS''ZBTB7A''PLAGL1''EGR1''EGR2''SPI1''CNOT4''JUND''ETV4''FOSL2''IKZF1''ZNF436''PURA''KLF2''EZH2''CTCF''KLF17''ELK1''TBP''SPDEF''MYC''SOX2''KLF3''TBP,TBPL1,TBPL2''BACH2''ZBTB18''ELF1,ELF2,ELF4''STAT3''ETV7''FOSL1''NR2C2''HNF4G''SMAD5''ZNF853''ZNF260''EGR2,EGR3,EGR4''ZBTB17''PATZ1''TRIM28''GABPA''BHLHA15''ZNF219''ZFX''SPIB,TEAD4''ZNF350''ZNF787''ZNF214,ZNF222,ZNF223,ZNF224,ZNF239,ZNF226,ZNF221,ZNF234,ZNF664,ZNF235,ZNF155,ZNF233,ZNF230,ZNF284,ZNF285,ZNF112,ZNF227''ZXDA''ZXDB''SPIC''MEF2C,MEF2B,MEF2A,BORCS8-MEF2B,MEF2D''ELK3''ZNF672''BCL3''KLF16''ZNF580''POU3F2''NKX3-1''ETV1''KLF7''FOXL1''RFX5,ETV2''PAX5''SMAD4''ELK1,TEAD4''LEF1,TCF7,TCF7L2,TCF7L1''SCRT1''RUNX3''ERG''ELF1''MEF2D''VEZF1''YY1''SREBF2''EP300''ZSCAN2''ETS2''ZFP30''TBX5''ETV5''HNF4G,HNF4A''RCOR1''CHD1''SCRT1,SCRT2''ATF3''SP1''THRB''NFKB1''STAT6,STAT4,STAT3,STAT2,STAT1,STAT5B,STAT5A''ZNF180''ZNF675''ZNF746''ZNF256''ZNF526''SREBF2,WT1''TBX15''GSC2''CTCFL''SALL2''ZXDC''SMAD1''TFAP2B''KLF8''RPS4X''NFE2L2,NFE2L3,BACH2,BACH1,PURA,MAFK,NFE2,MAF,MAFB,NFE2L1,MAFF,MAFG''ZNF404''ZC3H11A''NR3C1,PBX1''PURA,EP300''SREBF1''RXRA,PPARA''NONO''SALL4''SMAD2''MTF1,MZF1''ZNF248''ZDHHC5''ELK1,ETV7''TOB2''ERF,ETV7''FLI1,EWSR1''KLF12''NR3C1,ZFP42,PGR''POU5F1,GTF2IRD1''ETV5,ETV7''ZNF160''FLI1,ETV7''TSC22D4''ETV2,ETV7''EBF1''EBF3,EBF2,EBF1''ZNF449''ISL1,LHX3''SOX11''LARP4''ELF1,STAT1''SOX9''SREBF2,SREBF1,ZNF354C''ARID3A,BCL6''PAX4,E2F1''ZNF780B''EIF5A2''GTF2IRD1''ZBTB6''ZNF45''ZNF735''ETS1,YY1''ZNF100''ZNF624''ZNF628''ZBTB14,YY2''ZNF227''VDR''STAT2''NANOG,NANOGP8''HHAT''ZNF69''KLF7,KLF6''ZNF845''SMAD9,SMAD4,SMAD5,SMAD6,SMAD7,SMAD1,SMAD2,SMAD3''ZSCAN4''ZNF33A''ETV2''PURA,YY2''GPAM''FOXI1,ERF''ISL2''ZFY''ZIC1''ZNF263''ZBTB41''ISL1''NXPH3''ZNF532''NR2F1''WT1''MAFG''OVOL2''ZNF614''SP5''GABPA,GABPB1''ETS1,ZNF354C''GLIS2,GLIS3,GLIS1,GLI2,GLI3,GLI1,ZIC2,ZIC3,ZIC1''NR1D2''ZNF195''PDS5A''ZNF83''PAX4,GATA6,GATA4,GATA5,GATA2,GATA3,GATA1''ZBTB7B''THAP12''IKZF1,MZF1''TAF1''PURA,ETS1''SMAD3''ZNF878''NR4A2''STAT6''SOX30''GRHPR''ZNF714''GTF2I''ZNF44''ZNF253''NFE2L2''MEF2C''NR2F1,HNF4A''SMAD9,SMAD4,SMAD5,SMAD6,SMAD1,SMAD2,SMAD3''SOX7''GTF2B''TCF12''NFE4''IRF2,IRF1,IRF7''PROX1''GLI2''KLF13,KLF12,KLF11,KLF10,KLF17,KLF16,KLF15,KLF7,KLF6,KLF5,KLF4,KLF3,KLF2,KLF1,KLF9,KLF8''HES5''ELK1,SPDEF''SOD1''BATF,FOS,JUNB,JUN,JUND,FOSL1,FOSL2''ETV2,SPDEF''ZNF189''FOXQ1''ZNF174''NFATC3,ELF5''ZNF611''NELFE''ZNF579''GATA2''NKX2-2''ZNF41''EGR2,EGR3,EGR1''ZNF419''PITX1,HES7''ZNF420''ZFP3''SCRT2''E2F1,POU2F1''MYB,TP53''ZNF418''TBX2,TBX3,TBX4,TBX5''SP2''TCF4,TCF3,TCF12''KIF22''NR1H3''CRX''SMC3''NR2C1''SPIB,GABPB1,ETV4,ETV7,ETV6,FLI1,SPI1,ERG,PSMD12,ETS2,ELK3,ELK1,ELK4,ELF1,GABPA,ELF2,ELF4,ETS1''IRF3''IRF4''ELK4''PURA,ZBTB14''HDAC2''FOS,FOSB,JUNB,JUN,JUND,FOSL1''REST''SNAI2''ZNF524''ZSCAN29''ZNF25''ZNF808''SOX21,SOX14''FGF19''SPDEF,TEAD4''BHLHE41''ZNF236''BATF''KLF14''ZNF407''SP3''TRIM21''SIX1,SIX2''HOXB2,TCF3''SPI1,NFATC3''ZNF304''GLIS3''ETS2,ETS1''FOXO1,ETV5''ONECUT2,ELK1''VDR,FLI1''ZNF837''ARFGAP1''JUN''SND1''OVOL3,OVOL2,OVOL1''CEBPB''PRDM1''KLF6''AKR1A1''GLI1''DDX53''FOXP2,FOXP3,FOXP1,FOXP4''PLAG1''SREBF1,MZF1''CEBPG''EOMES,TEAD4''KLF11''ZNF282''MKX''DBP,TEF,HLF''E2F4''ZNF335''NR2F1,HNF4A,NR2F2''RBM17''ETS1,MZF1''IRF7''ZNF706''NR3C1,SPZ1''ZNF716''RBM8A''SP4''ESRRA''ISL1,ISL2''SOX17''A1CF''ZNF224''ZNF585A''NR1H4,RXRA''FOXF1,FOXF2''ZBTB11''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA,PAX4''EZR''SMARCC1''ZNF544''CELF5''PPARA''MECOM''PPARG,PPARD,PPARA''ZNF311''MLX''SMAP2''MAX,MYC''ZNF814''STAT1''MAFK,NFE2'

#'SP1''ZNF740''MAZ''MZF1''KLF5''EGR2''EGR2,EGR3,EGR1,EGR4,WT1''MECOM,PRDM16,PRDM13''ZNF281''WT1''ZNF148''ZEB1''KLF15''IKZF1''EGR1''CTCFL''KLF4''PATZ1''SP4''NR2C2''NR2F6''PAX4''ZIC1''ELF5''KLF17''SPIC''SP1,SP2,SP3,SP4''KLF6''SP3''ZBTB7B''SREBF1''RAD21''PURA''ZNF675''CTCF''ZNF180''SMAD3''ZBTB17''KLF2''KLF16''TP53''SP5''VEZF1''SPZ1''SPI1''ZSCAN10''ZNF341''KLF3''KLF1''SPDEF''PAX5''EGR2,EGR3,EGR4''ZIC3''PLAG1''KLF8''ZBTB7A''ZNF436''KDM4C,KDM4B,KDM4A,KDM4F,KDM4E,KDM4D''SREBF2,WT1''SALL4''TEAD2''GRHPR''KLF11''REST''RCOR1''GPD1''SP7''TBX15''ZDHHC5''ASCL2''EGR2,EGR3,EGR1''NXPH3''ZNF202''KLF13,KLF12,KLF11,KLF10,KLF17,KLF16,KLF15,KLF7,KLF6,KLF5,KLF4,KLF3,KLF2,KLF1,KLF9,KLF8''CUX1''OVOL2''BCL6B''ZCCHC14''SREBF2,SREBF1,ZNF354C''RARA,RARB,RARG,RXRG,THRB,RXRA,RXRB,THRA,PAX4''ZMAT2''NFE2L2,NFE2L3,BACH2,BACH1,PURA,MAFK,NFE2,MAF,MAFB,NFE2L1,MAFF,MAFG''ASCC1''EZH2''PLAGL2''AFF4''KLF10''ZNF746''ZNF853''MTF1,MZF1''SP1,SP2,SP3,SP4,ZBTB14''GCM1,PITX1''SALL2''CHURC1''PDLIM5''LUZP1''RXRA,PPARA''ZFX''GTF2I''GLI2''PROX1''PURA,ZBTB14''RREB1''ZNF404''SMAD1''SREBF1,MZF1''PURA,EP300''AKR1A1''NFE4''NONO''EZR''KLF12''EGR3''ZXDC''TRMO''ANXA11''VDR''ZC3H11A''ELK1,ETV7''ZNF219''PRDM5''TRIM28''SREBF2''PGAM2''SMAD5''GLI1''KLF13''ZNF580''ZBTB33''ZNF266''FLI1,ETV7''NR1H4''ZNF443''ERF,ETV7''SMAD2''ZNF460''AGGF1''GSC2''FEV''TRIM21''RXRA''PRDM1''SIX2''ZNF503''ETS1''E2F1,HES7''FOXI1,ETV5''RBM7''SOCS4''ZNF699''MED30''EGR4''GLIS2,GLIS3,GLIS1,GLI2,GLI3,GLI1,ZIC2,ZIC3,ZIC1''ETV5''SPIB,TEAD4''PTPMT1''ZXDB''MYC''ZNF526''ZNF248''SMAP2''SP1,SP3,SP4''RXRB''GLIS3''ETS2''TFCP2''ZNF761''ZNF260''ZNF117''ZNF256''HAND1''ETV2,ETV7''NR2F1''TFAP2C''FLI1''ZNF714''SRF''ZNF431''PIR''TWIST1''PLAGL1''ZXDA''SMAD4,SMAD1,SMAD2,SMAD3''ZNF282''TFAP2B''GLIS2''ZNF575''RPS4X''PDS5A''NR3C1,PBX1''GABPA''BHLHA15''SNAI2''ZIC2''CTBP2''SP2''TCEAL6'




########2 transform motif centered record to tf centered, that is to say, group by tf

tfmotif.list <- list()

for(cid in names(enrichTable.list) ){
    cat('do group_by_tf for dar ',cid,'\n')
    enr.dar <- enrichTable.list[[cid]]
    for(i in seq_along(1:nrow(enr.dar))  ){
        enrich.df <- enr.dar[i,]
        rownames(enrich.df) <- NULL
        for(i in colnames(enrich.df)){  enrich.df[,i] <- as.character(enrich.df[,i])  }

        enr.line <- unlist(enrich.df)

        motif <- enr.line['motif']
        nes <- enr.line['NES']
        tf <- enr.line['tf']
        peak <- enr.line['peaks']
        nearGene <- enr.line['nearGene']

        tfs <- strsplit(x = tf,split = ',')[[1]]
        for(tfid in tfs){
          tfmotif.list[[cid]][[tfid]][[motif]] <- enrich.df #enr.line
        }

    }

}

saveRDS(tfmotif.list,'tfmotif.list.rds')


###########3 select the best tf-motif record####


##is motif a unique annotation with this tf?
##is the nes max?
##overlap of peaks and nearGenes of all motif records




list2freqDF <- function(datalist = NULL){
    len.list <- length(datalist)

    #stopifnot(sum(sapply(datalist,FUN = function(x){ sum(duplicated(x))  }  ) ) ==0 )
    #0, no duplications within each elements

    #length(unique(unlist(test.list) ) ) #19

    res.stat <- table(unlist(datalist) )

    allid <- unique(unlist(datalist))

    #all.equal(sort(names(res.stat)),sort(allid) ) #TRUE
    
    freq.df <- data.frame(matrix(data = 0, nrow = length(allid), ncol = len.list ),row.names = allid )
    colnames(freq.df ) <- names(datalist )
    
    for(i in names(datalist) ){
        for(j in datalist[[i]]){  
            freq.df[j,i] <- freq.df[j,i] + 1
        
        }
        
        
    }
    
    return(freq.df)

}

#datalist <- tfmotif.nes.list

list2DF <- function(datalist = NULL){
    len.list <- length(datalist)
    allid.list <- sapply( datalist, function(x){ names(x) }  )
    allid <- Reduce(f = c, x = allid.list)
    allid <- unique(allid)
    
    nes.df <- data.frame(matrix(data = 0, nrow = length(allid), ncol = len.list ),row.names = allid )
    colnames(nes.df) <- names(datalist)
    
    for(i in names(datalist) ){
        for(j in names(datalist[[i]]) ){  
            nes.df[j,i] <- datalist[[i]][[j]]
        
        }
        
        
    }
    return(nes.df)
    
}





examFilterMotif <- function(data = NULL, cid = NULL, tfname = NULL, plot = NULL){

    n_motif <- length(data[[cid]][[tfname]])
    is_multi <- ifelse(n_motif > 2, TRUE, FALSE)
    motifid <- names(data[[cid]][[tfname]])
    
    col_cl <- ifelse(is_multi, TRUE, FALSE )
    
    #get data
    peaklist <- lapply(data[[cid]][[tfname]], FUN = function(x){strsplit(x$peaks,split = ',')[[1]] }  )
    
    nearGenelist <- lapply(data[[cid]][[tfname]], FUN = function(x){
        strsplit(x$nearGene,split = ',')[[1]] }  )
    nearGenelist_names <- lapply( nearGenelist, FUN = function(x){
        sapply(strsplit(x,split = '\\|'), function(y){y[2]}  ) }  )
    
    neslist <- lapply(data[[cid]][[tfname]], FUN = function(x){ x$NES}  )
    maxvalue <- neslist[[which.max(neslist)]]
    maxid <- names(which.max(neslist))
     #for list find max 

    peaklist.freqDF <- list2freqDF(peaklist)
    nearGenelist.freqDF <- list2freqDF(nearGenelist)
    nearGenelist_names.freqDF <- list2freqDF(nearGenelist_names)
    
    if(plot){
        ##plot peak combined    
        options(repr.plot.width = 7.5, repr.plot.height = 7.5)
        par(oma=c(3,3,3,3))
        if(sd(unlist(peaklist.freqDF)) !=0  ){
            pheatmap::pheatmap(peaklist.freqDF,main = paste(cid,'_dar-',tfname,'-peaks',sep=''),cluster_cols = col_cl)
        }else{
           pheatmap::pheatmap(peaklist.freqDF,main = paste(cid,'_dar-',tfname,'-peaks',sep=''),cluster_cols = col_cl,breaks = c(min(peaklist.freqDF),max(peaklist.freqDF)*1.01))
        }

        ##plot near gene combined

        if(sd(unlist(nearGenelist.freqDF)) !=0  ){
            pheatmap::pheatmap(nearGenelist_names.freqDF,main = paste(cid,'_dar-',tfname,'-nearGene',sep=''),cluster_cols = col_cl)
        }else{
           pheatmap::pheatmap(nearGenelist_names.freqDF,main = paste(cid,'_dar-',tfname,'-nearGene',sep=''),cluster_cols = col_cl,breaks = c(min(nearGenelist.freqDF),max(nearGenelist.freqDF)*1.01))
        }

        #pheatmap::pheatmap(nearGenelist.freqDF,main = paste(cid,'_dar-',tfname,'-near_genes',sep=''),cluster_cols = col_cl,breaks = c(min(nearGenelist.freqDF),max(nearGenelist.freqDF)*1.01))

        #plot nes

        par(oma=c(10,1,1,1))
        options(repr.plot.width = 7.5, repr.plot.height = 7.5)
        res.bp <- barplot(as.numeric(unlist(neslist)), xaxt = 'n',las = 2, ylab = 'TF NES (normalized enrichment score)' , xlab = '')
        text(x = res.bp[,1], y = 0,adj=1, srt = 45, xpd = TRUE ,labels = names(neslist) )
        text(x = res.bp[,1], y = as.numeric(unlist(neslist)),adj=1, offset = 0.5, xpd = TRUE ,labels = as.numeric(unlist(neslist)) )

        #plot tfmotif logo
        #c1_DAR_overlap_p2g_0.005/icistarget/logos/*.logo.png

        if(n_motif > 9){
            sample(motifid,size = 9, replace = FALSE)
        }

        options(repr.plot.width = 15, repr.plot.height = 15)
        par(oma=c(1,1,1,1))
        plot(NA, xlim = c(0, 3), ylim = c(0, 3), type = "n", xaxt = "n", yaxt = "n", xlab = "", ylab = "", main = paste(cid,' ',tfname,' ','motif logos (downsample to 9 if too many)',sep=''), frame.plot = FALSE)


        for(i in 1:length(motifid)){
            logofile <- paste('dar_',cid,'_overlap_p2g_auc_0.001/icistarget/logos/',motifid[i],'.logo.png',sep='')
            if(file.exists(logofile)){   
               img <- readPNG(logofile)
               rowi <- ceiling(i/3)
                coli <- i %% 3
                if(coli == 0){coli = 3}
               #cat(i,',',rowi,',',coli,':',coli-1, ':',rowi-1, ':',coli, ':',rowi,'\n')
               rasterImage(img, coli-1, rowi-1, coli, rowi) #xleft  ybottom  xright  ytop
            }
        }

    }
    

    return(list(cid=cid, tf=tfname, 'NES'= maxvalue, 'motif'=maxid ,'peaks' = data[[cid]][[tfname]][[maxid]]$peaks, 'nearGenes'= data[[cid]][[tfname]][[maxid]]$nearGene,'peaks_combined' = rownames(peaklist.freqDF), 'nearGenes_combined'= rownames(nearGenelist.freqDF)))
    
    
}

# #https://stackoverflow.com/questions/43817698/reading-png-files-into-r-and-create-a-multiple-plot
# img <- readPNG(system.file("img", "Rlogo.png", package="png"))

# plot(NA, xlim = c(0, 2), ylim = c(0, 4), type = "n", xaxt = "n", yaxt = "n", xlab = "", ylab = "")
# rasterImage(img, 1, 3, 2, 4)
# rasterImage(img, 1, 2, 2, 3)
# rasterImage(img, 1, 1, 2, 2)
# rasterImage(img, 1, 0, 2, 1)




# data <- tfmotif.list
# cid <- 'c2'
# tfname <- 'STAT4'
# tfname <- 'STAT5A'
# #tfname <- 'ESRRA'

# #motif of this tf has more than 2 
# #n_motif <- lapply (tfmotif.list[[cid]], FUN = length )
# #names(n_motif > 2)


res.tfmotif.choose.c2.STAT5A <- examFilterMotif(data = tfmotif.list, cid = 'c2', tfname = 'STAT5A',plot = TRUE)
res.tfmotif.choose.c2.STAT4 <- examFilterMotif(data = tfmotif.list, cid = 'c2', tfname = 'STAT4',plot = TRUE)
#res.tfmotif.choose.c2.CDX2 <- examFilterMotif(data = tfmotif.list, cid = 'c2', tfname = 'CDX2',plot = TRUE)

res.tfmotif.choose.c1.FOSL2 <- examFilterMotif(data = tfmotif.list, cid = 'c1', tfname = 'FOSL2',plot = TRUE)

length(res.tfmotif.choose.c2.STAT5A$peaks_combined) #58 peaks with 7 motfifs


#res.tfmotif.choose.c2.STAT5A[['peaks_combined']]
#grep('chr17:63931182-63931743',res.tfmotif.choose.c2.STAT5A[['peaks_combined']])



####readin dorc and p2g table

cisCorr.filt <- readRDS('/sda/mjwang/pwdex/placenta_10X_combine/03.snRNA_snATAC/liger/figR_Kartha/FigR/PLA-8w-combined/cisCorr.filt.rds') #19674 
p2g.filter <- readRDS('/sda/mjwang/pwdex/placenta_10X_combine/03.snRNA_snATAC/liger/peak_gene_link/result_good/gene_peak_links/out.p2g.filter.rds') #93053
p2g.filter <- as.data.frame(p2g.filter)


p2g.filter.uniq <- readRDS('/sda/mjwang/pwdex/placenta_10X_combine/03.snRNA_snATAC/liger/peak_gene_link/result_good/gene_peak_links/out.p2g.filter.uniq.rds') #46558


length(unique(p2g.filter$gene))
#9922

length(unique(p2g.filter.uniq$gene))
#7678



##readin dar
regionSets <- readRDS('../../../RcisTarget/regionSets.rds')
regionSets.dar <- lapply(regionSets,function(x){  mcols(x)$name   } )
regionSets[['c2']]
regionSets.dar[['c2']]

sapply(regionSets.dar, length)
#c6 7000 c3 7000 c9 5244 c5 6188 c7 6140 c1 5276 c8 2722 c2 4929 c4 1158

all.equal(sapply(regionSets, length),sapply(regionSets.dar, length)) #TRUE



##readin deg top1000
deg.list <- readRDS('../../../../../02.seurat_harmony/DEGs_extend/geneList.loose_use_all.top1000.rds')
#fold change with name, loose_use_all
names(deg.list)
#'up', 'down'

deg.list.up <- deg.list[['up']]

sapply(deg.list.up, length)
#c8 1000 c5 1000 c9 894 c10 694 c6 137 c1 145 c3 242 c2 133 c4 123 c7 246

##rna cluster -vs- atac cluster

names(regionSets.dar)
'c6''c3''c9''c5''c7''c1''c8''c2''c4'
names(deg.list.up)
'c8''c5''c9''c10''c6''c1''c3''c2''c4''c7'



map_cluster <- list( #atac cid to rna cid
       'c6' = 'c9',
       'c3' = 'c5',
       'c9' = 'c10',
       'c5' = 'c6',
       'c1' = 'c7',
       'c2' = 'c3',
       'c4' = 'c2' 
)


#######################readin go result of deg top1000 and extract go id member gene#####################
ck.go.df <- readRDS('/sda/mjwang/pwdex/placenta_10X_combine/02.seurat_harmony/DEGs_extend/result_do_GO_Pathway_quickonestep/ck.go.df.pvalue_0.05.qvalue_0.1.rds')

ck.go.df.list <- split(ck.go.df,f = ck.go.df$Cluster)

ck.go.df.member.list <- lapply(ck.go.df.list, FUN = function(x){  
    df2list <- list()
    for(i in 1:nrow(x) ){  
      df2list[[x[i,'Description']]] <-  strsplit(x[i,'geneID'],split = '/')[[1]] #x[i,'geneID'] #
    }
    df2list
    
} )

names(ck.go.df.member.list)
'c10''c6''c1''c3''c2''c4''c7'


goid_extract_genes.list <- list()

###extract growth hormone related go id members
ck.go.df.member.list[['c3']]
idx <- grep("growth",names(ck.go.df.member.list[['c3']]))
goid_extract <- ck.go.df.member.list[['c3']][idx]
names(goid_extract)
#'response to growth hormone' 'growth hormone receptor signaling pathway' 'cellular response to growth hormone stimulus' 'growth hormone receptor signaling pathway via JAK-STAT'

goid_extract_genes <- unique(unlist(goid_extract) )
#'CPS1''CSHL1''CSH1''CSH2''GH2''JAK2''PRLR'

goid_extract_genes.list[['c3']] <- goid_extract_genes



##extract hypoxia (hypoxia, oxygen ) ,angiogenesis related go id members

ck.go.df.member.list[['c7']]
idx <- grep("hypoxia|oxygen|angiogenesis",names(ck.go.df.member.list[['c7']]))

goid_extract <- ck.go.df.member.list[['c7']][idx]
names(goid_extract)
#'response to oxygen levels' 'positive regulation of angiogenesis' 'response to hypoxia' 'response to decreased oxygen levels''regulation of angiogenesis'

for(i in names(goid_extract) ){cat( i,"\n", goid_extract[[i]],'\n',sep=' ' ) }

goid_extract_genes <- unique(unlist(goid_extract) )
'LEP''EGLN3''LIMD1''HDAC2''ANGPTL4''ATG7''BNIP3L''NDRG1''CARD16''HK2''SLC2A1''PRKCE''NPEPPS''PAM''EGLN1''SERPINE1''FLT1''SASH1''ITGA5''ENG''GATA2''ADAM12''JUP''ALOX5'

goid_extract_genes.list[['c7']] <- goid_extract_genes





#########




#######

make_bold_names <- function(mat, rc_fun, rc_names) { #different font boldness and color for rowname/colnames
  bold_names <- rc_fun(mat)
  ids <- rc_names %>% match(rc_fun(mat))
  ids %>%
    walk(
      function(i)
        bold_names[i] <<-
        bquote(bold(.(rc_fun(mat)[i]))) %>%
        as.expression()
    )
  bold_names
}


tfmotif.list.filter <- list() #iterative for all dar tfmotif list and add targetGene_dorc, targetGene_dorc_combined; targetGene_p2g, targetGene_p2g_combined
for(cid in names(tfmotif.list) ){
    cat('cluster ',cid,'\n')
    for(tfname in names(tfmotif.list[[cid]]) ){
        cat('  tf ',tfname,'\n')
        res.tfmotif.choose <- examFilterMotif(data = tfmotif.list, cid = cid, tfname = tfname,plot = FALSE)
        #choose the motif with the maximum NES 
        
        #stopifnot (sum(res.tfmotif.choose$peaks_combined %in% cisCorr.filt$PeakRanges) == length(res.tfmotif.choose$peaks_combined) ) #should all within
        
        stopifnot (sum(res.tfmotif.choose$peaks_combined %in% p2g.filter$peak) == length(res.tfmotif.choose$peaks_combined) ) #should all within
        
        motifid <- res.tfmotif.choose[['motif']]
        
        
        #split string for this tf
        res.tfmotif.choose$peaks <- strsplit(res.tfmotif.choose$peaks,split = ',')[[1]] 
        res.tfmotif.choose$nearGenes <- strsplit(res.tfmotif.choose$nearGenes,split = ',')[[1]]
        
        ##add targetGene_dorc, targetGene_dorc_combined; targetGene_p2g, targetGene_p2g_combined
        idx<- which (cisCorr.filt$PeakRanges %in% res.tfmotif.choose$peaks ) #match all
        res.tfmotif.choose[['targetGenes_dorc']] <- paste(cisCorr.filt[idx,'PeakRanges'],
                                                     cisCorr.filt[idx,'Gene'],
                                                     sep = '|')
        
        idx<- which (cisCorr.filt$PeakRanges %in% res.tfmotif.choose$peaks_combined ) #match all
        res.tfmotif.choose[['targetGenes_dorc_combined']] <- paste(cisCorr.filt[idx,'PeakRanges'],
                                                     cisCorr.filt[idx,'Gene'],
                                                     sep = '|')
        
#         idx<- which (p2g.filter$peak %in% res.tfmotif.choose$peaks ) #match all
#         res.tfmotif.choose[['targetGenes_p2g']] <- paste(p2g.filter[idx,'peak'],
#                                                      p2g.filter[idx,'gene'],
#                                                      sep = '|')
        
#         idx<- which (p2g.filter$peak %in% res.tfmotif.choose$peaks_combined ) #match all
#         res.tfmotif.choose[['targetGenes_p2g_combined']] <- paste(p2g.filter[idx,'peak'],
#                                                      p2g.filter[idx,'gene'],
#                                                      sep = '|')
        
        
        idx<- which (p2g.filter.uniq$peak %in% res.tfmotif.choose$peaks ) #match all
        res.tfmotif.choose[['targetGenes_p2g']] <- paste(p2g.filter.uniq[idx,'peak'],
                                                     p2g.filter.uniq[idx,'gene'],
                                                     sep = '|')
        
        idx<- which (p2g.filter.uniq$peak %in% res.tfmotif.choose$peaks_combined ) #match all
        res.tfmotif.choose[['targetGenes_p2g_combined']] <- paste(p2g.filter.uniq[idx,'peak'],
                                                     p2g.filter.uniq[idx,'gene'],
                                                     sep = '|')
        
        
        
        ##overlap with deg of this cluster
        nearGenes <- sapply( res.tfmotif.choose$nearGenes, FUN = function(x){
        sapply(strsplit(x,split = '\\|'), function(y){y[2]}  ) }  )
        
        nearGenes_combine <- sapply( res.tfmotif.choose$nearGenes_combine, FUN = function(x){
        sapply(strsplit(x,split = '\\|'), function(y){y[2]}  ) }  )
        
        
        targetGenes_dorc <- sapply( res.tfmotif.choose$targetGenes_dorc, FUN = function(x){
        sapply(strsplit(x,split = '\\|'), function(y){y[2]}  ) }  )
        
        targetGenes_dorc_combine <- sapply( res.tfmotif.choose$targetGenes_dorc_combine, FUN = function(x){
        sapply(strsplit(x,split = '\\|'), function(y){y[2]}  ) }  )
        
        
        targetGenes_p2g <- sapply( res.tfmotif.choose$targetGenes_p2g, FUN = function(x){
        sapply(strsplit(x,split = '\\|'), function(y){y[2]}  ) }  )
        
        targetGenes_p2g_combine <- sapply( res.tfmotif.choose$targetGenes_p2g_combine, FUN = function(x){
        sapply(strsplit(x,split = '\\|'), function(y){y[2]}  ) }  )
        
        
        if(is.null(map_cluster[[cid]])){  
            tftarget.genesets <- list(
             #'nearGenes'=nearGenes,
             'nearGenes' = nearGenes_combine,
             #'targetGenes_dorc' = targetGenes_dorc,
             'targetGenes_dorc' = targetGenes_dorc_combine,
             #'targetGenes_p2g' = targetGenes_p2g,
             'targetGenes_p2g' = targetGenes_p2g_combine
             #'DEG' = names(deg.keep) #names(deg.keep)
            
            )

        }else{
          deg <- deg.list.up[[map_cluster[[cid]]]]
          deg.keep <- deg[names(deg) %in% c(nearGenes_combine, targetGenes_dorc_combine,targetGenes_p2g_combine ) ]
            tftarget.genesets <- list(
             #'nearGenes'=nearGenes,
             'nearGenes_combine' = nearGenes_combine,
             #'targetGenes_dorc' = targetGenes_dorc,
             'targetGenes_dorc' = targetGenes_dorc_combine,
             #'targetGenes_p2g' = targetGenes_p2g,
             'targetGenes_p2g' = targetGenes_p2g_combine,
             'DEG' = names(deg.keep) #names(deg.keep)
            
            )
            
        }
        

        tftarget.genesets.freqDF <- list2freqDF(tftarget.genesets)

        if(is.null(map_cluster[[cid]])){
           rowid_hi = "" 
        }else{
           if( is.null(goid_extract_genes.list[[map_cluster[[cid]]]])  ){
                rowid_hi = ""
            }else{
                hi_genes <- goid_extract_genes.list[[map_cluster[[cid]]]]
                rowid_hi = hi_genes[hi_genes %in% rownames(tftarget.genesets.freqDF)]
            }
        }

        
        if(nrow(tftarget.genesets.freqDF) < 2){  cat('too few rows in freqDF: ',nrow(tftarget.genesets.freqDF),' at tf ',tfname,', skip this tf\n',sep='' );next }
        
        #options(repr.plot.height = 35, repr.plot.width = 5.5)
        options(repr.plot.height = 10, repr.plot.width = 5.5)
        #options(repr.plot.width = width, repr.plot.height = height)
        res.p <- pheatmap(
                       tftarget.genesets.freqDF, 
                       cluster_cols = TRUE,
                       cluster_rows = TRUE, 
                       treeheight_col = 2.5,
                       treeheight_row = 2.5,
                       #cellwidth = 20, 
                       #cellheight = 2.8,
                       na_col = 'white', 
                       color = colorset,#rev(colorset_pathway),
                       border =TRUE,
                       border_color = 'black',
                       #labels_col = colid, 
                       labels_row = make_bold_names(tftarget.genesets.freqDF,
                                                    rownames, 
                                                    rowid_hi
                                                   ),
            
                       angle_col = 315,
                       #display_numbers = data.df.text,
                       #number_color = 'white',
                       #fontsize_number = 10,
                       fontsize_col = 20,
                       fontsize_row = 12,
                       main = paste('ATAC_',cid,'_dar-',tfname,'-targetGeneSet',sep=''),
                       silent = TRUE
                       #legend_breaks = c(2,4,6,8,10),
                       #legend_labels = c(2,4,6,8,10),
                       #legend = TRUE
#                        height = height,
#                        width = width,
#                        filename = paste('result_do_GO_Pathway_quickonestep/ck.GO_BP.pvalue_',
#                                         pvalue,'.qvalue_', qvalue,'.heatmap.pdf' ,sep = '' 
#                                        )
                       
                      )
        
        
#         options(repr.plot.height = 5.5, repr.plot.width = 5.5)
#         res.p.modified <- add.flag(res.p, 
#                                    kept.labels = growth_hormone_goid_genes[growth_hormone_goid_genes %in% rownames(tftarget.genesets.freqDF)],
#                                    repel.degree = 0.5
#                                   )
        
        
        #res.tfmotif.choose[['overlap']] <- tftarget.genesets.freqDF

        
        res.tfmotif.choose[['shared_targets']] <- rownames(tftarget.genesets.freqDF)[res.p$tree_row$order]
        res.tfmotif.choose[['heatmap']] <- res.p
        
        tfmotif.list.filter[[cid]][[tfname]] <- res.tfmotif.choose
        
    }
    
}


#saveRDS(tfmotif.list.filter,'tfmotif.list.filter.rds')
saveRDS(tfmotif.list.filter,'tfmotif.list.filter.uniq.rds')

tfmotif.list.filter.withdup <- readRDS('tfmotif.list.filter.rds')


###plot  dtf of interest in c2
names(tfmotif.list.filter[['c2']])
'BACH2''STAT5A''HDAC2''RUNX3''GABPB1''CEBPB''PRDM9''BCL6''STAT4''ZNF658''ZNF658B''ZNF813''FOS''ZNF614''TRPS1''GATA6''GATA4''GATA5''GATA2''GATA3''GATA1''STAT6''EP300''TFCP2''KLF4''STAT5B''TROVE2''JUND''GRHL2''E2F1''SOX17''RUNX2''CBFB''RUNX1''ZNF449''ZNF217''GLIS2''GLIS3''GLIS1''GLI2''GLI3''GLI1''ZIC2''ZIC3''ZIC1''KLF1''HSFY1''SALL2''KLF5''AFF4''KLF6''JUNB''ZBTB7A''TSC22D4''STAT1''CDX1''CDX2''CDX4''ELK1''HOXD12''JUN''FOSL2''FOSL1''RXRA''PPARA''HNRNPH3''NFIC''PROX1''ZNF345''AKR1A1''STAT3''STAT2''SP2''SIN3A''GPD1''ATF2''ZXDA''IKZF3''HDAC1''FOSB''MTHFD1''BACH1''SP1''ESRRG''ZNF761''ATF7''GRHL1''KLF3''TEAD4''CLOCK''ZNF599''PDS5A''ZNF329''IRF9''SMAD1''ZNF546''TP53''HP1BP3''ZNF354C''POU2F1''ETV1''SMARCB1''ZNF384''ZNF461''IRX5''ZMAT2''NFIB''SMARCC1''DHX36''BATF''SP3''ZNF226''SMAD9''SMAD4''SMAD5''SMAD6''SMAD2''SMAD3''ELF1''HOXB2''PLG''HSFY2''IKZF1''RFX5''PLAG1''POU4F3''NFE2''TBL1XR1''RCOR1''TBX21''ERF''MORN1''JAZF1''SPATS2''NXPH3''BCL3''ZNF430''POLR2A''SPDEF''PAX1''ZNF628''POU4F1''GAR1''PIR''CRX''NFIA''RAX''NR2F2''CHD1''HOXA9''ZDHHC5''CREB1''ZNF404''ZNF160''IRF3''TFAP2A''TFAP2C''TFAP2B''TFAP2E''TFAP2D''SMARCA4''ELK3''FLI1''SPIB''TCF3''TEAD3''TEAD2''TEAD1''CREM''MEF2C''AGGF1''CEBPA''TOB2''RPS10''KLF8''MYLK''NR3C2''DBP''PGR''ZNF708''CEBPD''PARP1''JDP2''SETDB1''EBF1''SREBF2''ZNF429''ETV2''DLX3''ETV4''EGR2''AR''VDR''RAB18''ZNF766''EGR1''ZXDB''CTCFL''MYC''ATF4''ATF1''ATF3''MITF''TFEC''TFE3''TFEB''PAX9''NFATC1''IRF8''ARNT2''MEIS1''PLAGL1''PAX5''NR3C1''NR4A3''POU1F1'

#'FOSB''CEBPB''STAT6''POU4F1''POU1F1''POU4F3''POU3F2''XBP1''STAT1''PAX2''CDX1''CDX2''CDX4''ESRRB''STAT3''STAT5A''ZNF782''POU4F2''POU3F1''POU2F2''HOXA13''ATF1''EZH2''JUN''TEAD1''YY1''POU3F3''ETV2''HES7''ZNF143''SMAD2''STAT4''STAT2''STAT5B''EGR4''CPTP''ZNF83''SIX5''SP2''POU3F4''EGR2''ZNF76''NR5A2''SETDB1''ZNF675''ETS1''THAP11''SMARCC2''EGR1''HOXA9''CDC5L''SCRT1''SCRT2''JDP2''ESRRA''PAX4''SPIC''ESRRG''NFAT5''NFATC2''NFATC3''NFATC4''NFATC1''TBX2''NFE2L1''RPS10''THAP1''SREBF2''POU6F2''POU6F1''SRF''CRX''PURA''NR2E1''NR2F2''GTF2F1''PITX1''TEAD4''ZNF559''GTF3C2''HOXB2''ELK3''KLF3''GPANK1''ZSCAN29''CREB1''CREM''ELF1''ATF4''ATF7''ATF3''ATF2''CEBPD''RAX''E2F1''GATA1''POLR3A''ZNF384''POLR2A''EP300''KDM2A''PHF8''KDM7A''FBXL19''KDM2B''PHF2''MEIS1''RXRA''PPARA''ZNF714''HMGN3''KLF17''ELF2''NKX3-2''HCFC1''ELK1''GATA6''GATA4''GATA5''GATA2''GATA3''ARID3A''POU5F1B''POU5F1''ETV7''ZNF683''NANOG''KLF1''JUND''SP6''RPS4X''CNOT4''KLF2''POU2F3''RARB''POU2F1''MAFG''TEF''KLF6''HOXD11''IRX5''LHX4''BDP1''ZNF100''ZFX''JUNB''DBP''HOXC9''NKX2-8''HOXD4''HOXA4''HOXC4''HOXB4''TBP''TBPL1''TBPL2''ZNF180''ZNF571''PDX1''TCF12''SP1''ETV5''LHX1''SOX8''ZFY''DLX3''SPI1''MYB''CBX7''BATF''PAX6''HMG20B''CEBPA''HOXD10''TBL1XR1''PBX4''HOXA10''RBPJ''THRA''UNCX''TRPS1''ELF3''SALL2''HMGB3''HMGB2''HMGB1''HMGB4''ZNF281''TBX21''E2F3''ZNF286B''GMEB1''DBX2'

#'FOSB''STAT5B''STAT5A''STAT1''ZFP62''TBP''TBPL1''TBPL2''CEBPB''STAT6''STAT4''STAT2''POU3F4''POU1F1''STAT3''ZNF714''NFIC''HOXD11''HOMEZ''POU3F3''SP100''HOXA11''SETBP1''EGR1''TLX2''POLR2A''KDM2B''XBP1''HOXD10''ZHX1''TEF''KLF17''DMRTA2''ATF4''KLF2''ZNF782''KLF6''TAF1''SP2''SP6''DBX2''HOXD4''HOXA4''HOXC4''HOXB4''POU3F2''SPIB''SREBF2''ELK1''HMGA2''JDP2''SRF''HOXB6''ELF2''IRF9''ZNF571''RARB''GTF2F1''POU4F1''EGR2''HMBOX1''NR5A2''HLX''MECP2''ATF2''IRX5''SP1''HOXD13''JUND''JUNB''EGR4''SP3''FOS''ARID5B''IL21''FOSL2''FOSL1''HOXA2''RARG''HNF1B''CNOT4''ZNF785''ZNF675''ELF1''TEAD4''HMGB3''HMGB2''HMGB1''HMGB4''RXRA''EP300''TCF4''MYF5''MYF6''MYOD1''ASCL1''CREB1''TFF3''TCF3''MYOG''CREM''TCF12''ATF7''ATF1''ATF3''HOXA9''ESRRA''RXRB''ZSCAN10''HMGN3''GCM1''SPDEF''POU2F3''CDX2''RARA'

options(repr.plot.height = 25, repr.plot.width = 5.5)
tfmotif.list.filter[['c2']][['STAT4']][['heatmap']]
tfmotif.list.filter[['c2']][['STAT5A']][['heatmap']] #hocomoco__STA5A_MOUSE.H11MO.0.A'

#tfmotif.list.filter[['c2']][['ESRRA']][['heatmap']]

options(repr.plot.height = 10, repr.plot.width = 5.5)
tfmotif.list.filter[['c2']][['ESRRG']][['heatmap']]

#'cid''tf''NES''motif''peaks''nearGenes''peaks_combined''nearGenes_combined''targetGenes_dorc''targetGenes_dorc_combined''targetGenes_p2g''targetGenes_p2g_combined''shared_targets''heatmap'
tfmotif.list.filter[['c2']][['STAT5A']][['peaks_combined']]
tfmotif.list.filter[['c2']][['STAT5A']][['targetGenes_p2g_combined']]

options(repr.plot.height = 25, repr.plot.width = 5.5)
tfmotif.list.filter[['c2']][['RUNX1']][['heatmap']]

tfmotif.list.filter[['c2']][['RUNX1']][['targetGenes_p2g_combined']]
'chr1:24990464-24991656|CLIC4''chr10:120950214-120950624|WDR11''chr11:10376840-10377309|AMPD3''chr11:78325335-78326012|GAB2''chr15:38593595-38594079|RASGRP1''chr17:69757849-69758374|LINC01483''chr2:213104394-213105729|SPAG16''chr2:28557043-28558282|SPDYA''chr22:32539647-32540533|FBXO7''chr3:133882229-133883028|SLCO2A1''chr3:41117402-41118060|CTNNB1''chr4:16082498-16082979|TAPT1-AS1''chr5:76457955-76458717|IQGAP2''chr6:150134860-150135160|PPP1R14C''chr6:46012766-46013390|CLIC5''chr8:91248978-91249582|SLC26A7''chr9:107471409-107472520|LINC01509''chr9:115993018-115993390|PAPPA''chr9:116058462-116059057|PAPPA''chr9:116161112-116161355|PAPPA''chr9:121239262-121240422|GSN''chr9:124098122-124098818|DENND1A''chrX:129877050-129877883|ZDHHC9'

unique(unlist(lapply(strsplit(x = tfmotif.list.filter[['c2']][['RUNX1']][['targetGenes_p2g_combined']], split = '\\|'), function(x){x[2]}) ))

'CLIC4', 'WDR11' ,'AMPD3', 'GAB2', 'RASGRP1', 'LINC01483' ,'SPAG16', 'SPDYA' ,'FBXO7' ,'SLCO2A1', 'CTNNB1', 'TAPT1-AS1' ,'IQGAP2', 'PPP1R14C', 'CLIC5' ,'SLC26A7', 'LINC01509' ,'PAPPA' ,'GSN', 'DENND1A' ,'ZDHHC9'

CLIC4
WDR11
AMPD3
GAB2
RASGRP1
LINC01483
SPAG16
SPDYA
FBXO7
SLCO2A1
CTNNB1
TAPT1-AS1
IQGAP2
PPP1R14C
CLIC5
SLC26A7
LINC01509
PAPPA
GSN
DENND1A
ZDHHC9


###plot tf of interest in c1
names(tfmotif.list.filter[['c1']])
'SMARCC1''BACH2''FOS''RCOR1''TCF7L2''FOSL2''ZNF554''FOSB''JUNB''FOSL1''JUND''SIN3A''MYC''JUN''STAT3''JDP2''TFAP2C''SMARCB1''MEF2A''DUS3L''KLF1''GATA2''BATF''ATF3''NFE2L2''BACH1''MAFK''NFE2''MAF''MAFB''NRF1''NFE2L1''MAFG''EP300''NAP1L1''NR1H4''KLF5''BCL3''GRHL1''EVX1''EVX2''PAX5''FGF19''ZFP62''TFCP2''EIF5A2''HNF4A''RORA''NXPH3''ZNF503''TCF12''SMAD3''KLF8''MEF2C''KDM5A''NR2C2''RBM17''NR2F1''NR2F2''POLR3A''CRTC2''TRIM21''AFF4''CEBPB''SALL4''KLF17''RAB7A''RXRA''ZNF384''CEBPA''ZBTB14''YY2''EPAS1''KLF4''RELA''PURA''HNF4G''NR2F6''HLCS''HAND1''PDS5A''KLF6''ZEB1''SREBF2''PPARG''CELF6''PPARD''PPARA''AKR1A1''PIR''MCTP2''KLF7''SMARCA4''ZBTB7B''ZBTB7A''CEBPD''ZNF708''CEBPG''TERF1''RARA''RARB''RARG''RXRG''THRB''RXRB''THRA''MXI1''ZNF660''NFIC''PPARGC1A''ZFY''ZFX''KLF12''SMAP2''FHL2''FLI1''TCF3''RAD21''MECOM''PRDM16''PRDM13''SRF''SOCS4''ZNF214''ZNF222''ZNF223''ZNF224''ZNF239''ZNF226''ZNF221''ZNF234''ZNF664''ZNF235''ZNF155''ZNF233''ZNF230''ZNF284''ZNF285''ZNF112''ZNF227''RREB1''ATF2''ZNF658''ZNF658B''VDR''ARNT2''ARNT''SREBF1''RFX1''ATF7''AVEN''TBX2''UGP2''MYBL1''MYB''MYBL2''SNAPC4''CREM''NR4A3''HLF''MEIS2''THAP1''ZBTB49''FOXO3''ESRRA''TP53''ATF1'

#'TFAP2C''NAP1L1''EVX1''EVX2''HNF4A''SMAD3''RBM17''FGF19''FOSL2''DUS3L''TRIM21''PURA''NXPH3''SMAP2''HLCS''RAD21''AKR1A1''AFF4''SP110''CELF6''TCF12''BACH1''CERS4''SMC3''FOSB''JUNB''FOSL1''JUND''FOS''MECOM''PRDM16''PRDM13''HAND1''THAP1''KLF5''SREBF2''MTHFD1''MCTP2''ZNF503''PAX5''FOXN3''KLF4''SMARCB1''JUN''ZCCHC17''TFCP2''KLF8''ZXDA''NFE2''RARA''RARB''RARG''RXRG''THRB''RXRA''RXRB''THRA''SP1''KDM5A''ZNF28''ZNF430''GPD1''ZXDB''TEAD2''PLAGL1''BACH2''ZNF671''KLF1''NR0B1''SREBF1''SOD1''KLF13''KLF12''KLF11''KLF10''KLF17''KLF16''KLF15''KLF7''KLF6''KLF3''KLF2''KLF9''TCF4''TCF3''FHL2''SMARCC1''APEX1''EP300''MYBL2''SOCS4''JDP2''ELK1''ZNF780A''ERG''GPAM''EBF1''ZNF554''RCOR1''RAB7A''SPIC''TCF7L2''ZNF583''ZNF148''ZBTB7A''SPI1''ZNF770''BATF''PDS5A''ATF2''UBB''STAT3''MAZ''ZNF692''NKX2-2''NFE2L2''CHD1''VEZF1''ZNF711''TRMO''PIR''SALL4''PLG''RORB''MAF''RBM7''SPIB''LYL1''ZBTB41''MYC''ATF3''NR1H4''MZF1''RELA''ZBTB7B''MEIS2''CUX1''ATF4''ZFX''HIVEP2''MAFB''RELB''REL''STAT6''STAT4''STAT2''YY1''HDAC1''ASCL2''ASCL1''PGAM2''NR2C2''STAT1''ZNF846''POLR3A''E2F1''ZXDC''ETV7''ATF7''SMAD1''OVOL2''EZH2''PATZ1''NR1D1''PRDM10''ZNF814''RORA''TGIF2LY''TGIF2LX''TGIF2''TGIF1''ETS1''ATF1''HNRNPH3''MESP1''HINFP''ZNF354C''ZSCAN20''NR4A2''ZNF480''MYBL1''HSF4''TFAP2B''UGP2''NFE4''AHR''ARNT''HIF1A''ARNT2''AHRR''ZNF682''MEIS1''AGGF1''MYB''SNAPC4''CHURC1''NFATC3''GMEB2''EIF5A2''SMAD4''SMAD2''ZNF816''ELF2''TFAP2A''ZNF423''UBP1'

#'RBM17''TFAP2C''NAP1L1''HNF4A''SMAD3''SMARCB1''MECOM''PRDM16''PRDM13''FGF19''HAND1''AKR1A1''NXPH3''PURA''KLF1''SMARCC1''AFF4''HLCS''TRIM21''SP110''ZNF503''RAD21''MCTP2''KLF4''JUND''KLF3''KLF6''ETS2''JUN''SALL4''ZNF148''PLAGL1''SP1''ELK1''SPI1''GPD1''FOSB''JUNB''FOSL1''FOS''SOCS4''NANOG''SMAP2''KLF2''NFE2L2''NFE2L3''BACH2''BACH1''MAFK''NFE2''MAF''MAFB''NFE2L1''MAFF''MAFG''TCF12''ATF3''KLF5''JDP2''TEAD2''FOSL2''TCEAL6''PIR''ZNF671''RELA''FOXN3''TFCP2''RBM7''ETV5''PDS5A''DUS3L''TFAP2A''GLIS3''POLR2A''POLR3A''EVX1''EVX2''NRF1''SMC3''SALL2''APEX1''EP300''FHL2''KLF12''HDAC1''ZNF711''KLF8''VEZF1''TBX21''EIF5A2''ZNF554''ZDHHC5''ZNF100''TFAP2B''ELF1''ZFY''ZFX''CUX1''ZNF202''RARA''RARB''RARG''RXRG''THRB''RXRA''RXRB''THRA''TAGLN2''ZBTB14''SMAD4''EBF1''AGGF1''MZF1''ETV7''SNAI2''SNAI3''SNAI1''ZSCAN10''TBX3''EZR''PAX4''ZNF345''SMARCC2''ZXDB''PTPMT1''SREBF1''THAP1''NFE4''HSF4''CEBPG''TRMO''PPARA''KLF13''KLF11''KLF10''KLF17''KLF16''KLF15''KLF7''KLF9''FLI1''ERF''ERG''ELK4''ELF2''ETS1''BATF''ZCCHC17''NR2C2''NELFE''ZSCAN31''GRHPR''UBB''ZIC5''ZSCAN26''TOB2''ETV4''PEG3''ZBTB2''CHD1''RFX1''ZIC1''SREBF2''FEV''ETV2''GABPA''ELF5''ZNF524''RELB''REL''FIZ1''CREB1''MXI1''HSPA1L''SMPX''RFX5''SMAD5''ASCL2''ETV6''PAX5''SPIB''GABPB1''PSMD12''ELK3''ELF4''ZSCAN20''KLF14''SMAD1''NR0B1''ZBTB7B''ZNF281''ZNF831''HIVEP1''HIVEP2''HIVEP3''SMARCA4'

#options(repr.plot.height = 20, repr.plot.width = 5.5)
options(repr.plot.height = 35 ,repr.plot.width = 5.5)
tfmotif.list.filter[['c1']][['FOSL2']][['heatmap']]



###plot tf of interest in c9
names(tfmotif.list.filter[['c9']])

'TP73''TP53''TP63''FOXO3''NAP1L1''PURA''SP7''MECOM''PRDM16''PRDM13''SMAD3''PTPMT1''KLF3''ASCC1''KLF6''SP110''PDS5A''GRHL1''UBB''EGR2''EGR3''EGR1''EGR4''WT1''EZR''SALL2''BCL6''ZDHHC5''HINFP''RFX5''ZNF714''TFAP2C''CEBPA''TFCP2''HAND1''PRDM1''RBM17''ZNF217''PPARA''MZF1''CEBPD''TRMO''KLF5''PLG''CEBPB''AKR1A1''RUNX3''GABPB1''LUZP1''TBX15''ZSCAN10''ZNF148''ZNF341''ZNF526''KLF4''SREBF2''AFF4''CHURC1''ZBTB7B''RAD21''SOCS4''TRIM21''ZIC1''GPD1''ZNF281''ZNF354C''CHD1''PAX5''MCTP2''TCF7L2''FGF19''GATA3''TEAD4''RBPJ''E2F1''TAF1''HIC1''KLF2''PROX1''ZNF503''ZCCHC14''ETV5''SP2''ZIC2''IRF2''MAZ''KLF16''ZNF607''ZNF189''ZNF84''ZNF23''ZNF471''ZNF16''ZNF502''ZNF429''PDLIM5''NXPH3''NANOS1''ASCL2''ZNF134''MTHFD1''ZBTB14''KLF1''SOX4''ZNF579''ZMIZ1''FOXO4''ERF''DLX3''OVOL2''TEF''SP1''POLR2A''STAT3''ZBTB33''RELA''MYB''GTF2I''CPSF4''ZNF460''RCOR1''SPIB''SP4''SALL4''EP300''ZNF418''MECP2''ETV2''OTUD4''RARA''RARB''RARG''RXRG''THRB''RXRA''RXRB''THRA''PAX4''PPARG''SPDEF'

#'EZR''RAD21''MECOM''PRDM16''PRDM13''FGF19''HSF4''HLCS''PURA''PLG''POLR3A''HNF1B''APEX1''EP300''RARA''RARB''RARG''RXRG''THRB''RXRA''RXRB''THRA''EBF1''RELA''SMC3''SP110''NAP1L1''SMARCB1''KLF6''PLAGL1''PAX5''RBM7''TCF12''JUND''PRDM1''HAND1''SP3''SMAD3''RBM17''EZH2''ASCC1''TRIM21''BCL11A''ZNF683''TP73''ETV2''SOX15''GCM1''HOXA2''GTF2F1''ZNF616''BACH1''SP2''CEBPB''SP6''ANXA11''TFCP2''AFF4''EVX1''EVX2''TRMO''UBTF''ZNF134''FOXR1''FOXR2''FOXN4''FOXN1''FOXN2''FOXN3''SRF''KLF5''SMARCA4''AKR1A1''ZNF554''FOSL1''TOB2''ARID3A''USF1''GTF2I''ZNF652''ZDHHC5''SPIB''NFE2''ZFY''ZNF239''CEBPD''ZNF263''CEBPE''ZNF510''SOD1''ZIC1''ZXDA''AHR''ARNT''HIF1A''ARNT2''AHRR''ZXDB''SMAP2''HNF4A''MAFK''BACH2''ZNF83''PTPMT1''ZBTB17''ZBTB14''UGP2''TAF1''E2F2''RBPJ''RBPJL''NFE2L2''TP53''SMAD4''ZNF589''BRF2''JUN''IRF2''CHD1''ZNF354C''MZF1''SIN3A''IRF4''KLF1''VEZF1''UBB''PDLIM5''ZXDC''SP100''BCL6''PAX4''IKZF2''SMARCC2''E2F1''ZNF672''ZNF579''ELK3''TP63''ZMAT2''ZNF341''ZNF503''ZNF595''ZSCAN22''MAZ''ZNF709''ZBTB2''ATF4''ZNF48''FOSL2''CEBPG''PRDM5''KLF2''FOSB''JUNB''FOS''KLF4''NFE2L1''MAFG''SPI1''HOXA3''ZNF670''ETV5''POLR2A''ZBTB33''ZNF621''ZNF460''GSC2''MYCN''SP9''MECP2''SP5''CUX1'

#'HNF1B''KLF15''UBB''KLF7''KLF8''PTPMT1''MECOM''PRDM16''PRDM13''RAD21''ZCCHC14''MXI1''CREB1''KLF6''ZNF202''RELA''RBBP5''AKR1A1''AFF4''JUND''EZH2''PURA''EBF1''RBM17''NAP1L1''NELFE''KLF11''HLCS''POLR3A''SP110''KLF13''SMC3''SP5''KLF10''HNF4A''TFAP2C''HAND1''SP3''TOB2''SP2''ZNF48''EGR2''EGR3''EGR1''ERF''HES7''GCM1''ELK3''SP6''SP9''CTCF''TP53''FGF19''ATF3''E2F1''SP1''KLF2''TP73''SP8''ELK1''EZR''ZNF341''TRMO''ZNF460''PLAGL1''ATF2''RBM7''HDAC2''RARA''RARB''RARG''RXRG''THRB''RXRA''RXRB''THRA''TP63''ZBTB7A''KLF12''ZNF219''POLR2A''ZNF134''ZIC1''GPD1''ATF4''ZDHHC5''MAZ''ZNF66''SREBF2''TFAP2A''KLF4''ZNF503''ZNF595''JUNB''SMPX''WT1''SREBF1''TBP''ZBTB14''UBTF''PAX5''CHURC1''SMAD1''SMARCB1''SP7''ZNF281''NXPH3''KLF17''NFE2L2''NFE2L3''BACH2''BACH1''MAFK''NFE2''MAF''MAFB''NFE2L1''MAFF''MAFG''FOSL2''MBD2''EGR4''FOXN3''ZNF676''PLAG1''TRIM21''ZFP64''PRDM5''JUN''KLF14''GLIS2'

options(repr.plot.height = 15, repr.plot.width = 5.5)
tfmotif.list.filter[['c9']][['RELA']][['heatmap']]
tfmotif.list.filter[['c9']][['GCM1']][['heatmap']]

names(tfmotif.list.filter[['c5']])

'NAP1L1''TFAP2C''RBM17''PURA''HAND1''HNF4A''SMAD4''SMAD2''FGF19''RAD21''HLCS''TEAD1''UBP1''ZNF557''TEAD2''TEAD3''HSF4''SMAD9''SMAD5''SMAD6''SMAD1''SMAD3''SP2''TEAD4''AKR1A1''ZFX''ERG''KLF16''NXPH3''NKX2-2''RFX5''CEBPG''SMAP2''TRMO''RARA''RARB''RARG''RXRG''THRB''RXRA''RXRB''THRA''ELF1''NANOG''TP73''SPI1''FLI1''TRIM21''TFCP2''GRHL2''RUNX1''PIR''NR2C2''PAX5''ETS1''DUS3L''MYB''ZEB1''ZNF91''EVX1''EVX2''AFF4''EBF1''PLG''ELF5''FOXI1''PRDM16''MECOM''PRDM13''SMC3''KDM5A''PLAGL1''TCF12''PGR''IKZF1''KLF6''ZNF503''SPDEF''KLF5''HDAC1''RBM7''JUN''EZH2''KLF4''BCL6''NR4A2''SOCS4''VDR''HINFP''EGR2''WISP2''FOXO1''ELK1''TFAP2A''FOXB1''UBB''SPIB''SP110''ZNF776''RORA''KLF12''FOXP2''RELA''MCTP2''ZNF483''SMARCA4''KLF3''PTPMT1''FOXC1''TP63''NKX2-1''TP53''ZXDA''ZDHHC5''TFAP4''GPD1''ZSCAN26''RBPJ''EHF''NFIC''NFIB''NFIA''NR2F6''ETV5''ZBTB7A''ERF''HNRNPH3''FIZ1''DEAF1''GRHL1''ZEB2''GLIS3''ZIC1''ESR2''APEX1''EP300''FOXK1''NR4A3'

#'NAP1L1''HLCS''RBM17''TRIM21''SMAP2''TRMO''PURA''RAD21''HNF4A''FGF19''SMAD3''SMAD1''HAND1''UBP1''NANOG''TFAP2C''RBM7''ESR2''KLF4''SMC3''FOXI1''ELK1''EZR''RARA''RARB''RARG''RXRG''THRB''RXRA''RXRB''THRA''GLIS3''AKR1A1''APEX1''EP300''HSF4''DUS3L''KDM5A''GRHL1''EVX1''EVX2''SMAD5''TEAD3''TEAD2''TEAD1''TEAD4''ETS1''FLI1''ETV5''MECOM''PRDM16''PRDM13''ZNF837''CTBP1''ZBTB33''ERG''KLF5''ELF1''ZBTB17''PAX4''KLF6''EGR2''ZNF23''RUNX2''RUNX3''RUNX1''SP110''ZNF557''NXPH3''TCF12''AFF4''INSM1''KLF12''PAX5''VEZF1''SMAD2''GMEB2''ZNF71''ZNF66''EGR1''ZBTB2''SMARCA4''KLF18''KLF17''CUX1''ZNF554''ELF2''NR2F1''NR2F2''NR1I3''NR1I2''NR1H2''NR1H3''MCTP2''AHR''ARNT2''ARNT''FHL2''IRF1''NR3C1''DPF1''ZFX''ZNF445''WT1''YY1''TCEAL6''PATZ1''ESR1''SPIB''RELA''ESRRG''SMAD4''ELF5''RORA''EIF5A2''PRDM5''OSR2''FOXS1''FOXC2''FOXC1''PRDM15''ZDHHC5''ZNF671''SPIC''BHLHE40''PROX1''PROX2''ZNF708''SPI1''ZNF563''ETS2''ESRRB''ZNF341''TAGLN2''ZNF354A''GRHL2''TBX21''ERF''ZBTB45''TAF1''TFAP4''IKZF1''ZNF546''PTPMT1''EGR3''GABPB1''ELK4''CEBPG''GPD1''CREB1''EOMES''MEF2C''MEF2B''MEF2A''BORCS8-MEF2B''MEF2D''ZNF263''HSF1''ZNF579''GABPA''NR2E1''ZNF513''ETV4''BRCA1''WRNIP1''FOXF2''CHD2''ETV1''POLR2A''ZNF26''CBFB''ZMAT2''GFI1B''HSPA1L''SRF''MAZ''SP3''UGP2''TCF4''ZFY''ZNF483''FOXA1''TFCP2''ELF4''MEIS3''MEIS2''MEIS1''TGIF2''TGIF2LY''TGIF2LX''TGIF1''PKNOX1''PKNOX2''EBF1''RBBP5''RORC''RORB''POLR3A''HDAC2''CTBP2''STAT1''XBP1''SREBF2''TP63''NKX2-1''TP53''TP73''ZNF639''CAT''ZNF267''MAFA''ZNF502'



options(repr.plot.height = 35, repr.plot.width = 5.5)
tfmotif.list.filter[['c5']][['TEAD3']][['heatmap']]
tfmotif.list.filter[['c5']][['NAP1L1']][['heatmap']]
tfmotif.list.filter[['c5']][['RELA']][['heatmap']]

# #https://stackoverflow.com/questions/52599180/partial-row-labels-heatmap-r
# #by Z.lin in Singapore, modify pheatmap returned gtable for row label
# add.flag <- function(pheatmap,
#                      kept.labels,
#                      repel.degree) {

#   # repel.degree = number within [0, 1], which controls how much 
#   #                space to allocate for repelling labels.
#   ## repel.degree = 0: spread out labels over existing range of kept labels
#   ## repel.degree = 1: spread out labels over the full y-axis

#   heatmap <- pheatmap$gtable

#   new.label <- heatmap$grobs[[which(heatmap$layout$name == "row_names")]] 

#   # keep only labels in kept.labels, replace the rest with ""
#   new.label$label <- ifelse(new.label$label %in% kept.labels, 
#                             new.label$label, "")

#   # calculate evenly spaced out y-axis positions
#   repelled.y <- function(d, d.select, k = repel.degree){
#     # d = vector of distances for labels
#     # d.select = vector of T/F for which labels are significant

#     # recursive function to get current label positions
#     # (note the unit is "npc" for all components of each distance)
#     strip.npc <- function(dd){
#       if(!"unit.arithmetic" %in% class(dd)) {
#         return(as.numeric(dd))
#       }

#       d1 <- strip.npc(dd$arg1)
#       d2 <- strip.npc(dd$arg2)
#       fn <- dd$fname
#       return(lazyeval::lazy_eval(paste(d1, fn, d2)))
#     }

#     full.range <- sapply(seq_along(d), function(i) strip.npc(d[i]))
#     selected.range <- sapply(seq_along(d[d.select]), function(i) strip.npc(d[d.select][i]))

#     return(unit(seq(from = max(selected.range) + k*(max(full.range) - max(selected.range)),
#                     to = min(selected.range) - k*(min(selected.range) - min(full.range)), 
#                     length.out = sum(d.select)), 
#                 "npc"))
#   }
#   new.y.positions <- repelled.y(new.label$y,
#                                 d.select = new.label$label != "")
#   new.flag <- segmentsGrob(x0 = new.label$x,
#                            x1 = new.label$x + unit(0.15, "npc"),
#                            y0 = new.label$y[new.label$label != ""],
#                            y1 = new.y.positions)

#   # shift position for selected labels
#   new.label$x <- new.label$x + unit(0.2, "npc")
#   new.label$y[new.label$label != ""] <- new.y.positions

#   # add flag to heatmap
#   heatmap <- gtable::gtable_add_grob(x = heatmap,
#                                    grobs = new.flag,
#                                    t = 4, 
#                                    l = 4
#   )

#   # replace label positions in heatmap
#   heatmap$grobs[[which(heatmap$layout$name == "row_names")]] <- new.label

#   # plot result
#   grid.newpage()
#   grid.draw(heatmap)

#   # return a copy of the heatmap invisibly
#   invisible(heatmap)
# }
                         
                       
######https://github.com/raivokolde/pheatmap/issues/48
# use this function to make row or column names bold
# parameters:
#   mat: the matrix passed to pheatmap
#   rc_fun: either rownames or colnames (function)
#   rc_names: vector of names that should appear in boldface
make_bold_names <- function(mat, rc_fun, rc_names) { #different font boldness and color for rowname/colnames
  bold_names <- rc_fun(mat)
  ids <- rc_names %>% match(rc_fun(mat))
  ids %>%
    walk(
      function(i)
        bold_names[i] <<-
        bquote(bold(.(rc_fun(mat)[i]))) %>%
        as.expression()
    )
  bold_names
}

# # Create test matrix
# test = matrix(rnorm(200), 20, 10)
# test[1:10, seq(1, 10, 2)] = test[1:10, seq(1, 10, 2)] + 3
# test[11:20, seq(2, 10, 2)] = test[11:20, seq(2, 10, 2)] + 2
# test[15:20, seq(2, 10, 2)] = test[15:20, seq(2, 10, 2)] + 4
# colnames(test) = paste("Test", 1:10, sep = "")
# rownames(test) = paste("Gene", 1:20, sep = "")

# pheatmap(
#   test,
#   labels_row = make_bold_names(test, rownames, c("Gene2", "Gene5", "Gene14")),
#   labels_col = make_bold_names(test, colnames, c("Test2", "Test3", "Test4")))


make_color_names <- function(){ #different color within the same string, use phantom to place emtpy label but not plotted, not ok
#https://blog.revolutionanalytics.com/2009/01/multicolor-text-in-r.html
    plot(rnorm(20),rnorm(20),col=rep(c("red","blue"),c(10,10)))
    title(expression("Hair color" * phantom(" and Eye color")),col.main="red")
    title(expression(phantom("Hair color and ") * "Eye color"),col.main="blue")
    title(expression(phantom("Hair color ") * "and " * phantom("Eye color")),col.main="black")
    
}




matchMulti <- function(df1 = NULL, df2 = NULL, by = NULL){
    idx <- which(df1[,by] %in% df2[,by] ) #find all (include dup) then merge

    df1.sel <- df1[idx,by]

    #link map.df.sel and nearGene.df by merge
    colnames(nearGene.df) <- c('x','id_meta','gene')
    df.merge <- merge(x=nearGene.df,y=map.df.sel, by = 'id_meta')


    return(df.merge)
    
}




res.tfmotif.c2.STAT4 <- tfmotif.list.filter[['c2']][['STAT4']]
res.tfmotif.c1.FOSL2 <- tfmotif.list.filter[['c1']][['FOSL2']]

#saveRDS(tfmotif.list.filter,'tfmotif.list.filter.rds')



######quick query taret gene for upstream TF regulators#######

tgene <- 'CSHL1'
tgene <- 'PAPPA'
tgene <- 'GH2'


tgene <- 'ERV'

tgene <- 'SH3TC2'
tgene <- 'PSG8'


tgene <- 'FLT1'
tgene <- 'INHBA'
tgene <- 'FSTL3' #empty (MYCN according to dorc method)
tgene <- 'LVRN'
tgene <- 'EGLN3'

tgene <- 'MYCN'
tgene <- 'POU2F3'

tgene <- 'ESRRG'

tgene <- 'STAT5A'

tgene <- 'STAT5B'

for(cid in names(tfmotif.list.filter) ){
    for(tfid in names(tfmotif.list.filter[[cid]]) ){
        peak_tgene <-  tfmotif.list.filter[[cid]][[tfid]][['targetGenes_p2g_combined']]
        ix <- grep (tgene,peak_tgene,value=TRUE)
        if(length(ix) != 0){ cat('Found tgene: ', paste(ix,collapse = ','), ' for gene ',tgene, ' in cid: ',cid, ' in tf: ',tfid,'\n',sep='') }
        
    }

}


for(cid in names(tfmotif.list.filter.withdup) ){
    for(tfid in names(tfmotif.list.filter.withdup[[cid]]) ){
        peak_tgene <-  tfmotif.list.filter.withdup[[cid]][[tfid]][['targetGenes_p2g_combined']]
        ix <- grep (tgene,peak_tgene,value=TRUE)
        if(length(ix) != 0){ cat('Found tgene: ', paste(ix,collapse = ','), ' for gene ',tgene, ' in cid: ',cid, ' in tf: ',tfid,'\n',sep='') }
        
    }

}



Found tgene: chr17:64115878-64116410|CSHL1 for gene CSHL1 in cid: c9 in tf: CEBPA
Found tgene: chr17:64115878-64116410|CSHL1 for gene CSHL1 in cid: c9 in tf: CEBPD
Found tgene: chr17:64115878-64116410|CSHL1 for gene CSHL1 in cid: c9 in tf: CEBPB
Found tgene: chr17:64115878-64116410|CSHL1 for gene CSHL1 in cid: c9 in tf: TCF7L2
Found tgene: chr17:64115878-64116410|CSHL1 for gene CSHL1 in cid: c9 in tf: KLF1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: RUNX3
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: GABPB1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: PRDM9
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: KLF4
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: ZIC1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: KLF1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: SALL2
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: KLF5
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: AFF4
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: KLF6
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: AKR1A1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: SP2
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: GPD1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: ZXDA
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: SP1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: KLF3
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: TEAD4
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: PDS5A
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: SP3
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: SMAD4
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: ELF1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: HOXB2
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: POLR2A
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: SPDEF
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: ZDHHC5
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: FLI1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: KLF8
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: EGR2
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: EGR1
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: ZXDB
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: CTCFL
Found tgene: chr17:63931182-63931743|CSHL1 for gene CSHL1 in cid: c2 in tf: PLAGL1

Found tgene: chr17:64115878-64116410|GH2 for gene GH2 in cid: c9 in tf: CEBPA
Found tgene: chr17:64115878-64116410|GH2 for gene GH2 in cid: c9 in tf: CEBPD
Found tgene: chr17:64115878-64116410|GH2 for gene GH2 in cid: c9 in tf: CEBPB
Found tgene: chr17:64115878-64116410|GH2 for gene GH2 in cid: c9 in tf: TCF7L2
Found tgene: chr17:64115878-64116410|GH2 for gene GH2 in cid: c9 in tf: KLF1


Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: SMARCC1
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: BACH2
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: FOS
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: RCOR1
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: TCF7L2
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: FOSL2
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: FOSB
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: JUNB
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: FOSL1
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: JUND
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: SIN3A
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: MYC
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: JUN
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: STAT3
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: JDP2
Found tgene: chr13:28574821-28575743|FLT1 for gene FLT1 in cid: c1 in tf: MEF2A

Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: JUND
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: JUN
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: NR1H4
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: KLF5
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: FGF19
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: EIF5A2
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: NXPH3
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: TCF12
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: SMAD3
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: NR2C2
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: NR2F2
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: SALL4
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: PURA
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: TERF1
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: MECOM
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: PRDM16
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: PRDM13
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: RFX1
Found tgene: chr2:15782231-15782991|MYCNOS for gene MYCN in cid: c1 in tf: MEIS2

Found tgene: chr11:120270535-120270947|POU2F3 for gene POU2F3 in cid: c1 in tf: SOCS4


Found tgene: chr7:41681375-41681768|INHBA,chr7:41681375-41681768|INHBA-AS1 for gene INHBA in cid: c1 in tf: CRTC2
Found tgene: chr7:41590177-41591044|INHBA,chr7:41590177-41591044|INHBA-AS1 for gene INHBA in cid: c2 in tf: TEAD4
Found tgene: chr7:41590177-41591044|INHBA,chr7:41590177-41591044|INHBA-AS1 for gene INHBA in cid: c2 in tf: SPDEF
Found tgene: chr7:41590177-41591044|INHBA,chr7:41590177-41591044|INHBA-AS1 for gene INHBA in cid: c2 in tf: PARP1
Found tgene: chr7:41590177-41591044|INHBA,chr7:41590177-41591044|INHBA-AS1 for gene INHBA in cid: c2 in tf: ZNF766

Found tgene: chr5:116046131-116046796|LVRN for gene LVRN in cid: c1 in tf: FOSL1
Found tgene: chr5:115948301-115948967|LVRN for gene LVRN in cid: c1 in tf: CEBPB
Found tgene: chr5:115948301-115948967|LVRN for gene LVRN in cid: c1 in tf: CEBPA
Found tgene: chr5:115948301-115948967|LVRN for gene LVRN in cid: c1 in tf: CEBPD

Found tgene: chr14:34303748-34304529|EGLN3 for gene EGLN3 in cid: c9 in tf: CEBPB
Found tgene: chr14:34252425-34252739|EGLN3 for gene EGLN3 in cid: c1 in tf: JUN
Found tgene: chr14:34252425-34252739|EGLN3 for gene EGLN3 in cid: c1 in tf: NR1H4
Found tgene: chr14:34303748-34304529|EGLN3 for gene EGLN3 in cid: c1 in tf: CEBPB
Found tgene: chr14:34231665-34232166|EGLN3 for gene EGLN3 in cid: c1 in tf: EPAS1
Found tgene: chr14:34231665-34232166|EGLN3 for gene EGLN3 in cid: c1 in tf: ZEB1
Found tgene: chr14:34231665-34232166|EGLN3 for gene EGLN3 in cid: c1 in tf: FLI1
Found tgene: chr14:34231665-34232166|EGLN3 for gene EGLN3 in cid: c1 in tf: TCF3
Found tgene: chr14:34252425-34252739|EGLN3 for gene EGLN3 in cid: c1 in tf: AVEN
Found tgene: chr14:34303748-34304529|EGLN3 for gene EGLN3 in cid: c1 in tf: TBX2
Found tgene: chr14:34223395-34224159|EGLN3 for gene EGLN3 in cid: c8 in tf: SRF


Found tgene: chr1:187877455-187877793|ERVMER61-1 for gene ERV in cid: c6 in tf: TP53
Found tgene: chr1:187877455-187877793|ERVMER61-1 for gene ERV in cid: c6 in tf: TEAD1
Found tgene: chr1:187877455-187877793|ERVMER61-1 for gene ERV in cid: c3 in tf: TP53
Found tgene: chr6:11129969-11130442|ERVFRD-1 for gene ERV in cid: c9 in tf: SP7
Found tgene: chr6:11111160-11112626|ERVFRD-1 for gene ERV in cid: c9 in tf: GATA3
Found tgene: chr6:11111160-11112626|ERVFRD-1 for gene ERV in cid: c9 in tf: TEAD4

Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: SMARCC1
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: BACH2
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: FOS
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: RCOR1
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: TCF7L2
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: FOSL2
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: FOSB
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: JUNB
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: FOSL1
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: JUND
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: SIN3A
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: MYC
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in tf: JUN
Found tgene: chr5:149134560-149134991|SH3TC2 for gene SH3TC2 in cid: c1 in 



Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: STAT5A
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: HDAC2
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: RUNX3
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: GABPB1
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: CEBPB
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: STAT4
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: STAT6
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: TFCP2
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: KLF4
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: STAT5B
...
Found tgene: chr17:42286567-42287058|STAT5A for gene STAT5A in cid: c2 in tf: PPARA


Found tgene: chr17:42274110-42274539|STAT5B for gene STAT5B in cid: c8 in tf: ZNF341
Found tgene: chr17:42274110-42274539|STAT5B for gene STAT5B in cid: c8 in tf: SMARCC2
Found tgene: chr17:42274110-42274539|STAT5B for gene STAT5B in cid: c8 in tf: SP4
Found tgene: chr17:42274110-42274539|STAT5B for gene STAT5B in cid: c8 in tf: TFAP2C
Found tgene: chr17:42274110-42274539|STAT5B for gene STAT5B in cid: c8 in tf: OVOL2
Found tgene: chr17:42274110-42274539|STAT5B for gene STAT5B in cid: c8 in tf: OVOL1



subset(p2g.filter,peak == 'chr5:116046131-116046796')

subset(p2g.filter.uniq,peak == 'chr5:116046131-116046796')





##########4 format tf-nes-cluster mat among all cluster and compared to tf-expr mat######


# lapply( tfmotif.list.filter[['c9']], function(x){ 
#       #grep('DUSP8',x[['shared_targets']],value=TRUE)
#       x[['NES']]
# }   )

# tfmotif.list.filter[['c9']][['EGR1']][['NES']]
# tfmotif.list.filter[['c9']][['EGR1']][['shared_targets']]



tfmotif.nes.list <- list()
for(cid in names(tfmotif.list.filter) ){
    tfmotif.nes.list[[cid]] <- lapply( tfmotif.list.filter[[cid]], function(x){ 
          #grep('DUSP8',x[['shared_targets']],value=TRUE)
          x[['NES']]
    }   )
    
    tfmotif.nes.list[[cid]] <- unlist(tfmotif.nes.list[[cid]])
}




tfmotif.nes.df <- list2DF(datalist = tfmotif.nes.list)
rowid <- rownames(tfmotif.nes.df)
tfmotif.nes.df <- apply(tfmotif.nes.df,2,as.numeric)
rownames(tfmotif.nes.df) <- rowid
604 x 9
#486 x 9


options(repr.plot.height = 45, repr.plot.width = 5.5)
#options(repr.plot.width = width, repr.plot.height = height)
nes.p <- pheatmap(
               tfmotif.nes.df, 
               cluster_cols = TRUE,
               cluster_rows = TRUE, 
               treeheight_col = 2.5,
               treeheight_row = 2.5,
               #cellwidth = 20, 
               #cellheight = 2.8,
               na_col = 'white', 
               color = c('white',colorset),#rev(colorset_pathway),
               border =TRUE,
               border_color = 'black',
               #labels_col = colid, 
#                labels_row = make_bold_names(tftarget.genesets.freqDF,
#                                             rownames, 
#                                             rowid_hi
#                                            ),

               ##angle_col = 315,
               #display_numbers = data.df.text,
               #number_color = 'white',
               #fontsize_number = 10,
               fontsize_col = 20,
               fontsize_row = 8,
               main = paste('ATAC_TF_enrichment_in_dar',sep=''),
               silent = FALSE
               #legend_breaks = c(2,4,6,8,10),
               #legend_labels = c(2,4,6,8,10),
               #legend = TRUE
#                        height = height,
#                        width = width,
#                        filename = paste('result_do_GO_Pathway_quickonestep/ck.GO_BP.pvalue_',
#                                         pvalue,'.qvalue_', qvalue,'.heatmap.pdf' ,sep = '' 
#                                        )

              )




####readin expression mat to add TF gene expression #####

exprMat.ave.z <-readRDS("/sda/mjwang/pwdex/placenta_10X_combine/02.snapATAC_harmony/chromVAR_TF_specific/full12526/TF_dev-vs-TF_expression/exprMat.ave.z.rds")


table(rownames(tfmotif.nes.df) %in% rownames(exprMat.ave.z))

FALSE  TRUE 
   55   549 

FALSE  TRUE 
   95   686 

FALSE  TRUE 
   50   436


tfmotif.exp.df <-  exprMat.ave.z[rownames(tfmotif.nes.df)[rownames(tfmotif.nes.df) %in% rownames(exprMat.ave.z)] ,]

colnames(tfmotif.exp.df) <- paste('c',colnames(tfmotif.exp.df),sep='')
#549 x 10

#686 x 10

#436 x 10


grep('ESR',rownames(tfmotif.exp.df), value = TRUE )
'ESRRA''ESR2''ESRRG'

#'ESR2''ESR1''ESRRG''ESRRB''ESRRA'
#'ESRRA''ESR2''ESR1'

grep('TBX3',rownames(tfmotif.exp.df), value = TRUE )
#not found

rowsds <- matrixStats::rowSds(as.matrix(tfmotif.exp.df)) 
quantile(rowsds)
0% 0 25% 0.0338533934349017 50% 0.0928136970060022 75% 0.173794584630954 100% 0.745274926860947

#0 %0 25% 0.0340040628757603 50% 0.092181068482616 75% 0.179327464130576 100% 0.745274926860947

#0% 0 25% 0.0343054017574775 50% 0.0964485621056574 75% 0.18938655327171 100% 0.745274926860947

tfmotif.exp.df.sel <- tfmotif.exp.df[rowsds >= 0.17,]
#tfmotif.exp.df.sel <- tfmotif.exp.df[rowsds >= 0.18,]
#115

options(repr.plot.height=20,repr.plot.width=6)
hp = Heatmap(tfmotif.exp.df.sel, name = "tfmotif.exp.df.sel", 
             cluster_rows = TRUE, 
             cluster_columns = TRUE, 
             show_row_names = TRUE,
             use_raster = FALSE,
             ##col = circlize::colorRamp2(seq(-1,1,by=2/10), viridis(n = 11,option = "C")),
             row_title = "TFmotif expression (sd >= 0.18,hclust: ward.D)",
             clustering_method_row = "ward.D", ##ward.D,complete
             #clustering_distance_rows  = "pearson",
             row_names_gp = gpar(fontsize = 8),
             #clustering_distance_columns  = function (m) dist(m,method="manhattan") #euclidean, manhattan,minkowski  #clustering_distance_columns  = "pearson",
             #heatmap_legend_param = list(color_bar = "continuous")
             ##right_annotation = ha#,heatmap_width=unit(8, "cm")
) 
#width = max(grobWidth(textGrob(labels))))

##pdf('TF_gene_expression.heatmap.pdf',height=15,width=4,useDingbats=FALSE)
set.seed(1);draw(hp, heatmap_legend_side = "left",gap = unit(0.1, "cm")) #8.895833 8.843750



###

table(rownames(tfmotif.exp.df) %in% rownames(tfmotif.nes.df) )
TRUE 
 549

#TRUE 
# 686

#436 TRUE

tfmotif_nes_order <- tfmotif.nes.df[rownames(tfmotif.exp.df),c('c6','c3','c9','c5','c1','c2','c4')] #549 x 7 #686 x 7#486 x 9
tfmotif_expr_order <-  tfmotif.exp.df[, c('c9','c5','c10','c6','c7','c3','c2')] #549 x 7 #686 x 7 #436 x 9


all.equal(rownames(tfmotif_nes_order),rownames(tfmotif_expr_order))
#TRUE

ncol(tfmotif_nes_order) == ncol(tfmotif_expr_order)
#TRUE




######################1 filter nes mat by sd only and output aligned mat################


rowsds <- matrixStats::rowSds(as.matrix(tfmotif_nes_order)) 
quantile(rowsds)
0% 0 25% 1.24365301086292 50% 1.69424550107429 75% 2.04032305229728 100% 4.4030093758897

#0 %0 25% 1.32590404210684 50% 1.72133111753708 75% 2.1376933095494 100% 4.50416386545977
#0 %0 25% 1.23685615927494 50% 1.69746221464998 75% 2.05908753706225 100% 3.16393933673023
#0% 0.802997574237999 25% 1.14129335983333 50% 1.58681988662351 75% 2.01129544418732 100% 3.12067000400633

sd.cutoff <- 1.69 #1.72#1.69
tfmotif_nes_order.sel <- tfmotif_nes_order[rowsds >= sd.cutoff,]
275
#353
#219
#183


options(repr.plot.height = 35, repr.plot.width = 5.5)
#options(repr.plot.width = width, repr.plot.height = height)
nes.sel.p <- pheatmap(
               tfmotif_nes_order.sel, 
               cluster_cols = FALSE,
               cluster_rows = FALSE, 
               treeheight_col = 2.5,
               treeheight_row = 2.5,
               #cellwidth = 20, 
               #cellheight = 2.8,
               na_col = 'white', 
               color = c('white',colorset),#rev(colorset_pathway),
               border =TRUE,
               border_color = 'black',
               #labels_col = colid, 
#                labels_row = make_bold_names(tftarget.genesets.freqDF,
#                                             rownames, 
#                                             rowid_hi
#                                            ),

               ##angle_col = 315,
               #display_numbers = data.df.text,
               #number_color = 'white',
               #fontsize_number = 10,
               fontsize_col = 20,
               fontsize_row = 8,
               main = paste('ATAC_TF_enrichment_in_dar ( sd cutoff:',sd.cutoff,' )',sep=''),
               silent = FALSE
               #legend_breaks = c(2,4,6,8,10),
               #legend_labels = c(2,4,6,8,10),
               #legend = TRUE
#                        height = height,
#                        width = width,
#                        filename = paste('result_do_GO_Pathway_quickonestep/ck.GO_BP.pvalue_',
#                                         pvalue,'.qvalue_', qvalue,'.heatmap.pdf' ,sep = '' 
#                                        )

              )


# rowid <- rownames(tfmotif_nes_order.sel)[nes.sel.p$tree_row$order]

# table(rowid %in% rownames(tfmotif_expr_order))
# #219 TRUE

# tfmotif_expr_order.sel <- tfmotif_expr_order[rowid,]

tfmotif_expr_order.sel <- tfmotif_expr_order[rownames(tfmotif_nes_order.sel),]

# options(repr.plot.height = 35, repr.plot.width = 5.5)
# #options(repr.plot.width = width, repr.plot.height = height)
# exrp.sel.p <- pheatmap(
#                tfmotif_expr_order.sel, 
#                cluster_cols = FALSE,
#                cluster_rows = FALSE, 
#                treeheight_col = 2.5,
#                treeheight_row = 2.5,
#                #cellwidth = 20, 
#                #cellheight = 2.8,
#                na_col = 'white', 
#                #color = c('white',colorset),#rev(colorset_pathway),
#                border =TRUE,
#                border_color = 'black',
#                #labels_col = colid, 
# #                labels_row = make_bold_names(tftarget.genesets.freqDF,
# #                                             rownames, 
# #                                             rowid_hi
# #                                            ),

#                ##angle_col = 315,
#                #display_numbers = data.df.text,
#                #number_color = 'white',
#                #fontsize_number = 10,
#                fontsize_col = 20,
#                fontsize_row = 8,
#                main = paste('TF_expression ( aligned to TF_NES )',sep=''),
#                silent = FALSE
#                #legend_breaks = c(2,4,6,8,10),
#                #legend_labels = c(2,4,6,8,10),
#                #legend = TRUE
# #                        height = height,
# #                        width = width,
# #                        filename = paste('result_do_GO_Pathway_quickonestep/ck.GO_BP.pvalue_',
# #                                         pvalue,'.qvalue_', qvalue,'.heatmap.pdf' ,sep = '' 
# #                                        )

#               )


options(repr.plot.height=35,repr.plot.width=5.5)
exrp.sel.p = Heatmap(tfmotif_expr_order.sel, name = 'expression', 
             cluster_rows = FALSE, 
             cluster_columns = FALSE, 
             show_row_names = TRUE,
             use_raster = FALSE,
             ##col = circlize::colorRamp2(seq(-1,1,by=2/10), viridis(n = 11,option = "C")),
             row_title = paste('TF_expression ( aligned to TF_NES )',sep=''),
             clustering_method_row = "ward.D", ##ward.D,complete
             #clustering_distance_rows  = "pearson",
             row_names_gp = gpar(fontsize = 8),
             #clustering_distance_columns  = function (m) dist(m,method="manhattan") #euclidean, manhattan,minkowski  #clustering_distance_columns  = "pearson",
             #heatmap_legend_param = list(color_bar = "continuous")
             ##right_annotation = ha#,heatmap_width=unit(8, "cm")
) 
set.seed(1);draw(exrp.sel.p, heatmap_legend_side = "left",gap = unit(0.1, "cm")) 




######################2 filter by correlation with TF expr mat and output aligned mat###################

###plot tfmotif-nes vs tfmotif-expr for each cluster
map_cluster.full <- list(
#atac-rna
#CTB
'c6' = 'c9',
'c3' = 'c5',

#fusion
'c9' = 'c10',#'c11',


#STB
'c5' = 'c6',
'c2' = 'c3',
'c4' = 'c2',
#'c1' = 'c4',
'c1' = 'c7'



##EVT and STR
#'c12' = 'c12',
#'c10' = 'c10'
#13-13
#15-15
#14-14
#11-16

)

##pair cluster with liger integration result
tfmotif_nes_order <- tfmotif.nes.df[,c('c6','c3','c9','c5','c1','c2','c4')] #604 x 9 #486 x 9
tfmotif_expr_order <-  tfmotif.exp.df[, c('c9','c5','c10','c6','c7','c3','c2')] #549 x 7 #436 x 9



all.equal (nrow(tfmotif_nes_order), nrow(tfmotif_expr_order) )#no 

length(intersect(rownames(tfmotif_nes_order),rownames(tfmotif_expr_order)))
549
#686
#436

i <- 'STAT4'
i <- 'STAT5A'
unlist(tfmotif_expr_order[i,])
unlist(tfmotif_nes_order[i,])

cor(unlist(tfmotif_expr_order[i,]), unlist(tfmotif_nes_order[i,]))
#0.819235091282893 STAT4
#0.75978 STAT5A


res.tfmotif.cor <- vector()
for(i in rownames(tfmotif_expr_order)){
  res.tfmotif.cor[i] <- cor(unlist(tfmotif_expr_order[i,]), unlist(tfmotif_nes_order[i,]))
  
}

res.tfmotif.cor <- res.tfmotif.cor[!is.na(res.tfmotif.cor)]
res.tfmotif.cor.sel.cor0.6 <- res.tfmotif.cor[res.tfmotif.cor >= 0.6]
res.tfmotif.cor.sel.cor0.5 <- res.tfmotif.cor[res.tfmotif.cor >= 0.5]
res.tfmotif.cor.sel.cor0.4 <- res.tfmotif.cor[res.tfmotif.cor >= 0.4]
res.tfmotif.cor.sel.cor0.25 <- res.tfmotif.cor[res.tfmotif.cor >= 0.25]

tf.cor.sel.cor0.5 <- names(res.tfmotif.cor.sel.cor0.5)
tf.cor.sel.cor0.4 <- names(res.tfmotif.cor.sel.cor0.4)
tf.cor.sel.cor0.25 <- names(res.tfmotif.cor.sel.cor0.25)

tf.cor.sel <- tf.cor.sel.cor0.25










##manually add these tf in case of filtered by low cor

tf.cor.sel <- c(tf.cor.sel,'ESRRG','PPARD','REL','MYCN','FOSL2','TBX3','BCL6','GCM1','OVOL1','POU2F3')

tf.cor.sel[duplicated(tf.cor.sel)]
#no dup
#tf.cor.sel <- tf.cor.sel[!duplicated(tf.cor.sel)]

tf.cor.sel[!tf.cor.sel %in% rownames(tfmotif_nes_order) | !tf.cor.sel %in% rownames(tfmotif_expr_order)]
#TBX3, POU2F3

tf.cor.sel <- tf.cor.sel[tf.cor.sel %in% rownames(tfmotif_nes_order) & tf.cor.sel %in% rownames(tfmotif_expr_order)]


#for(i in names(res.tfmotif.cor.sel)){
#for(i in names(res.tfmotif.cor.sel.cor0.5)){
for(i in tf.cor.sel){
  options(repr.plot.height=5.5,repr.plot.width=5.5)
  plot(unlist(tfmotif_expr_order[i,]), unlist(tfmotif_nes_order[i,]) , xlab = 'TF gene expression (normalized)',ylab = 'TF NES', pch = 19, cex = 0.5, main = paste('TF gene ',i) )
}



#options(repr.plot.height=5.5,repr.plot.width=5)
options(repr.plot.height=15.5,repr.plot.width=5)
#hp.expr = Heatmap(tfmotif_expr_order[tf.cor.sel.cor0.5,], name = "tfmotif.exp.high_corr", 
hp.expr = Heatmap(tfmotif_expr_order[tf.cor.sel,], name = "tfmotif.exp.high_corr",
             cluster_rows = TRUE, 
             cluster_columns = FALSE, 
             show_row_names = TRUE,
             use_raster = FALSE,
             #col = circlize::colorRamp2(seq(-1,1,by=2/10), viridis::viridis(n = 11,option = "C")),
             row_title = "TFmotif expression (aligned,hclust: ward.D)",
             clustering_method_row = "ward.D", ##ward.D,complete
             #clustering_distance_rows  = "pearson",
             row_names_gp = gpar(fontsize = 8),
             rect_gp = gpar(col = 'black', lwd=0.5),
             ##border_gp = gpar(col = 'black', lwd = 2),
             #clustering_distance_columns  = function (m) dist(m,method="manhattan") #euclidean, manhattan,minkowski  #clustering_distance_columns  = "pearson",
             #heatmap_legend_param = list(color_bar = "continuous")
             ##right_annotation = ha#,heatmap_width=unit(8, "cm")
) 
#width = max(grobWidth(textGrob(labels))))

##pdf('TF_gene_expression.heatmap.pdf',height=15,width=4,useDingbats=FALSE)
set.seed(1);draw(hp.expr, heatmap_legend_side = "left",gap = unit(0.1, "cm")) #8.895833 8.843750



set.seed(1);res.row.order <- row_order(hp.expr) 
#rowid.order <- rownames(tfmotif_expr_order[tf.cor.sel.cor0.5,])[res.row.order]
rowid.order <- rownames(tfmotif_expr_order[tf.cor.sel,])[res.row.order]

#set.seed(1);res.col.order <- column_order(hp.expr) 
#colid.order <- colnames(tfmotif.exp.df[tf.cor.sel,])[res.col.order]
#colid.order[map_cluster.full[colid.order]  ]

#options(repr.plot.height=5.5,repr.plot.width=5)
options(repr.plot.height=15.5,repr.plot.width=5)
hp.nes = Heatmap(tfmotif_nes_order[rowid.order ,], name = "tfmotif.NES.high_corr", 
             cluster_rows = FALSE, 
             cluster_columns = FALSE, 
             show_row_names = TRUE,
             use_raster = FALSE,
             ##col = circlize::colorRamp2(seq(-1,1,by=2/10), viridis(n = 11,option = "C")),
             #col = c('white',colorset),
             row_title = "TFmotif NES (aligned,hclust: ward.D)",
             clustering_method_row = "ward.D", ##ward.D,complete
             #clustering_distance_rows  = "pearson",
             row_names_gp = gpar(fontsize = 8),
             rect_gp = gpar(col = 'black', lwd=0.5),
             #clustering_distance_columns  = function (m) dist(m,method="manhattan") #euclidean, manhattan,minkowski  #clustering_distance_columns  = "pearson",
             #heatmap_legend_param = list(color_bar = "continuous")
             ##right_annotation = ha#,heatmap_width=unit(8, "cm")
) 
#width = max(grobWidth(textGrob(labels))))

##pdf('TF_gene_expression.heatmap.pdf',height=15,width=4,useDingbats=FALSE)
set.seed(1);draw(hp.nes, heatmap_legend_side = "left",gap = unit(0.1, "cm")) #8.895833 8.843750



##output aligned tf gene nes and expr mat as long format


tfmotif_nes_align <- hp.nes@matrix
tfmotif_expr_align <- hp.expr@matrix


all.equal( sort(rownames(tfmotif_expr_align)),sort(rownames(tfmotif_nes_align)) ) #TRUE

tfmotif_expr_align <- tfmotif_expr_align[rownames(tfmotif_nes_align),]

all.equal( rownames(tfmotif_expr_align),rownames(tfmotif_nes_align) ) #TRUE


colnames(tfmotif_nes_align)
'c6''c3''c9''c5''c1''c2''c4'

colnames(tfmotif_expr_align)
'c9''c5''c10''c6''c7''c3''c2'


##select
##rowid_mannual <- c('ATF3','BACH1','MAFK','MAFF','STAT4','STAT5A','TWIST1','ELK1','ETV4','JUNB','CEBPG','JUND','STAT6','ZNF66','ZNF134','RELA','MAFG','SUZ12','CHD1','ZNF304','HINFP','BRF1','E2F4','ZNF676','NFYA','TP53','MTHFD1','EZH2','SETBP1','TP63','PPARG','NR2F2')


##exclude

idx1 <- grep ('^CPTP$',rownames(tfmotif_expr_align) )
idx2 <-  grep ('^EBF1$',rownames(tfmotif_expr_align) )
rowid_exclude <- c(rownames(tfmotif_expr_align)[idx1:idx2])



##

tfmotif_expr_align <- tfmotif_expr_align[!rownames(tfmotif_expr_align) %in% rowid_exclude,]
tfmotif_nes_align <- tfmotif_nes_align[!rownames(tfmotif_nes_align) %in% rowid_exclude,]

#tfmotif_expr_align <- tfmotif_expr_align[rowid_mannual,]
#tfmotif_nes_align <- tfmotif_nes_align[rowid_mannual,]

all.equal( rownames(tfmotif_expr_align),rownames(tfmotif_nes_align) )
#TRUE

#options(repr.plot.height=5.6,repr.plot.width=5)
options(repr.plot.height=10.5,repr.plot.width=5)
hp.expr.manual = Heatmap(tfmotif_expr_align, name = "tfmotif.exp.high_corr", 
             cluster_rows = FALSE, 
             cluster_columns = FALSE, 
             show_row_names = TRUE,
             use_raster = FALSE,
             col = circlize::colorRamp2(seq(-0.8,0.8,length.out = 255), colorRampPalette(c('blue','white','red'))(255) ),
             row_title = "TF gene expression (aligned)",
             clustering_method_row = "ward.D", ##ward.D,complete
             #clustering_distance_rows  = "pearson",
             row_names_gp = gpar(fontsize = 8),
             rect_gp = gpar(col = 'black', lwd=0.5),
             ##border_gp = gpar(col = 'black', lwd = 2),
             #clustering_distance_columns  = function (m) dist(m,method="manhattan") #euclidean, manhattan,minkowski  #clustering_distance_columns  = "pearson",
             #heatmap_legend_param = list(color_bar = "continuous")
             ##right_annotation = ha#,heatmap_width=unit(8, "cm")
) 
#width = max(grobWidth(textGrob(labels))))

##pdf('TF_gene_expression.heatmap.pdf',height=15,width=4,useDingbats=FALSE)
set.seed(1);draw(hp.expr.manual, heatmap_legend_side = "left",gap = unit(0.1, "cm")) #8.895833 8.843750

zScore <- function (m = NULL, min = -2, max = 2, limit = FALSE) 
{
    z <- sweep(m - rowMeans(m), 1, matrixStats::rowSds(m), `/`)
    if (limit) {
        z[z > max] <- max
        z[z < min] <- min
    }
    return(z)
}#ArchR


tfmotif_nes_align.z <- zScore(tfmotif_nes_align,limit = TRUE)


#options(repr.plot.height=5.5,repr.plot.width=5)
options(repr.plot.height=10.5,repr.plot.width=5)
hp.nes.manual = Heatmap(tfmotif_nes_align.z , name = "tfmotif.NES.high_corr", 
#hp.nes.manual = Heatmap(tfmotif_nes_align , name = "tfmotif.NES.high_corr", 
             cluster_rows = FALSE, 
             cluster_columns = FALSE, 
             show_row_names = TRUE,
             use_raster = FALSE,
             ##col = circlize::colorRamp2(seq(-1,1,by=2/10), viridis(n = 11,option = "C")),
             col = c('white',colorset),
             ##col = circlize::colorRamp2(seq(-2,2,by=4/4), c('white',colorset)),
             row_title = "TF motif NES (aligned)",
             clustering_method_row = "ward.D", ##ward.D,complete
             #clustering_distance_rows  = "pearson",
             row_names_gp = gpar(fontsize = 8),
             rect_gp = gpar(col = 'black', lwd=0.5),
             #clustering_distance_columns  = function (m) dist(m,method="manhattan") #euclidean, manhattan,minkowski  #clustering_distance_columns  = "pearson",
             #heatmap_legend_param = list(color_bar = "continuous")
             ##right_annotation = ha#,heatmap_width=unit(8, "cm")
) 
#width = max(grobWidth(textGrob(labels))))

##pdf('TF_gene_expression.heatmap.pdf',height=15,width=4,useDingbats=FALSE)
set.seed(1);draw(hp.nes.manual, heatmap_legend_side = "right",gap = unit(0.1, "cm")) #8.895833 8.843750


##############output to long format for dotplot with grid#################

wide2long <- function(df1 = NULL,df2=NULL){
    ##turn a wide dataframe to long format, with row_idx, col_idx and row_id, col_id and value 
    row_name <- rownames(df1)
    col_name <- colnames(df1)
    
    #make sure that df2 has the same rowname with df1, the same ncol with df1 (but no need to have the same colname, but be paired)
    
    stopifnot( all.equal(rownames(df1), rownames(df2) ) )
    stopifnot( ncol(df1) == ncol(df2) )
    
    df.long <- data.frame()
    for(i in 1:nrow(df1)){
        for(j in 1:ncol(df1) ){
            row_idx <- i
            col_idx <- j
            row_id <- row_name[i]
            col_id <- col_name[j]
            value1 <- df1[i,j]
            value2 <- df2[i,j]
            df.long <- rbind.data.frame(df.long, data.frame( row_idx = row_idx, col_idx = col_idx, row_id = row_id, col_id= col_id,  value1=value1, value2=value2  ) )
            
        }
        
    }
    
    return(df.long)
}




tf_keep <- unique(unlist(tf_select))
#tf_keep <- c('STAT5A','STAT4','AR','MITF',)


all.equal(rownames(tfmotif_nes_align.z),rownames(tfmotif_expr_align) )
#all.equal(rownames(as.data.frame(tfmotif_nes_align.z)[tf_keep,]),rownames(as.data.frame(tfmotif_expr_align)[tf_keep,]) )

#tfmotif_align_long <- wide2long(df1 = as.data.frame(tfmotif_nes_align.z)[tf_keep,], df2 = as.data.frame(tfmotif_expr_align)[tf_keep,])



##reorder rowid before wide2long

##tf before ordered
tf_order <- c('MCTP2','TP63','PPARG','NR2F2','TEAD4','TEAD1','ESRRG','PPARD','REL','DLX3','POLR2A','SP4','SP110','ZNF16','RAD21','CPSF4','MZF1','RBBP5','VEZF1','HNF4G','TCF7','TEF','ZNF263','ZMAT2','UBB','EGR1','TFAP2B','POLE4','NFYA','PATZ1','KLF13','SUPT20H','APEX1','NFYC','BRF1','ZNF808','ZNF148','E2F4','GRHPR','ASCC1','TP53','ZCCHC14','ZNF43','PPARA','ANXA11','CTBP2','TCF7L1','GATA2','NR4A3','AFF4','FOSL2','NR2C2','JUNB','ZDHHC5','ZBTB7B','OVOL1','ZNF217','EGR3','SP2','ZNF134','ZNF460','KLF16','STAT3','RELA','TRIM21','CEBPG','ZFP62','HOXB2','RNF114','KLF15','BATF','CEBPD','BCL3','VDR','NFE2','GLIS2','STAT6','ATF3','BACH1','MXD1','RAB7A','CEBPB','GCM1','POU2F1','ZNF91','MAFK','STAT4','AR','ELK1','ARNT2','RUNX1','STAT5A','XBP1','JUND','ETV5','SMARCB1','ETV4','MITF','HP1BP3','SMAD6','MYCN','GLIS3','TFAP2A','MXI1','SREBF1','BCL6','MYLK','TBL1XR1','STAT5B')


rowid.nes <- rownames(tfmotif_nes_align.z)
rowid.expr <- rownames(tfmotif_expr_align)

all.equal(rowid.nes,rowid.expr) #TRUE
all.equal(tf_order, rowid.nes) #TRUE




########calculate correlation plot and value for selected TFs  below

i <- 'STAT4'
i <- 'STAT5A'
unlist(tfmotif_expr_order[i,])
unlist(tfmotif_nes_order[i,])

cor(unlist(tfmotif_expr_order[i,]), unlist(tfmotif_nes_order[i,]))
#0.819235091282893 STAT4
#0.75978 STAT5A


res.tfmotif.cor.new <- vector()
for(i in order_list){
  res.tfmotif.cor.new[i] <- cor(unlist(tfmotif_expr_align[i,]), unlist(tfmotif_nes_align.z[i,]))
  
}



color_set = list('c6'='#74add1',
           'c3'='#4575b4',
           'c9'='darkgreen',
           'c5'='#ffffbf',
           'c7'='#fdae61',
           'c1'='#f46d43',
           'c8'='#fee090',
           'c2'='#d73027',
           'c4'='#a50026')



#par(mfrow = c(4,5))
par(mfrow = c(3,3))
options(repr.plot.height = 12, repr.plot.width = 12)

#options(repr.plot.height=5.5,repr.plot.width=5.5)
for(i in order_list){
  #options(repr.plot.height=5.5,repr.plot.width=5.5)
  plot(unlist(tfmotif_expr_align[i,]), unlist(tfmotif_nes_align.z[i,]) , 
       xlab = 'TF gene expression (normalized)',ylab = 'TF NES', pch = 19, cex = 2, 
       main = paste('TF gene ',i,' correlation: \n r = ',round(res.tfmotif.cor.new[i],digits = 3),sep=''),
       col = unlist(color_set[colnames(tfmotif_nes_align.z)]),
       cex.main = 2
       #cex.xlab = 3
      )
  abline(h=0.3,lty=2)
  text(unlist(tfmotif_expr_align[i,]), unlist(tfmotif_nes_align.z[i,]), 
       labels = colnames(tfmotif_nes_align.z),
       pos = 1,
       col = unlist(color_set[colnames(tfmotif_nes_align.z)]),
       cex = 2
      )
  
}


########







order_list <- c('MCTP2','DLX3','TEAD4','RELA','ZNF217','GLIS3','BCL6','FOSL2','JUNB','CEBPG','MXI1','NR4A3','GATA2','MITF','STAT5A','STAT4','AR','ELK1','RUNX1','VDR')


#order part of tf_order according to order_list


sum(duplicated(order_list)) #0
table(order_list %in% tf_order )
TRUE 
  20

idx <- which(tf_order %in% order_list)
#for(i in idx){cat(i,',',sep='')}
#1,5,10,48,49,51,53,57,64,66,74,87,88,89,91,92,98,102,104,106

tf_order[idx]

'MCTP2''TEAD4''DLX3''GATA2''NR4A3''FOSL2''JUNB''ZNF217''RELA''CEBPG''VDR''STAT4''AR''ELK1''RUNX1''STAT5A''MITF''GLIS3''MXI1''BCL6'

all.equal(sort(order_list),sort(tf_order[idx])  )
#TRUE

tf_order[idx] <- order_list


sum(duplicated(tf_order)) #0


tfmotif_align_long <- wide2long(df1 = tfmotif_nes_align.z[tf_order,], df2 = tfmotif_expr_align[tf_order,])


#tfmotif_align_long <- wide2long(df1 = tfmotif_nes_align, df2 = tfmotif_expr_align)
#tfmotif_align_long <- wide2long(df1 = tfmotif_nes_align.z, df2 = tfmotif_expr_align)



colnames(tfmotif_align_long) <- c('y_idx','x_idx','y_id','x_id','NES','expression')


#tf_hi <- c('STAT4','STAT5A','JUNB','CEBPG','JUND','RELA','MAFG','TP63','PPARG','NR2F2','AR','MYCN','MITF','VDR')
#tf_hi <- c('STAT4','STAT5A','JUNB','CEBPG','JUND','ZNF66','ZNF134','RELA','MAFG','TP63','PPARG','NR2F2')
##chosen for network construction

tf_hi <- c('STAT4','STAT5A','AR','MITF','FOSL2','JUNB','CEBPG','JUND','RELA','ZNF217','GRHL1','TEAD4','TP63','PPARG')



tfmotif_align_long$edge <- ifelse(tfmotif_align_long$y_id %in% tf_hi, 'black','grey' )





#####################ggplot dot plot with grid or use external python script####################

# ggplot(data = tfmotif_align_long, aes(x= x_id, y = y_id, size = NES, col = expression) ) +
#   geom_point()


temp <- tfmotif_align_long$expression

temp[temp < 0] <- 0

temp.min <- min(temp)
temp.max <- max(temp)

# (temp - temp.min)/(temp.max-temp) = x/(1-x)
# x*(temp.max-temp) = (temp - temp.min)*1-  (temp - temp.min)* x

# x*temp.max-x*temp = temp - temp.min - x*temp + x* temp.min

# x*temp.max = temp - temp.min + x* temp.min

# x*temp.max - x* temp.min = temp - temp.min
# x * (temp.max -temp.min) = temp - temp.min

x = (temp - temp.min)/(temp.max -temp.min)

tfmotif_align_long$expression.transform <- x

# options(repr.plot.height=5.5,repr.plot.width=8) #too urgly, use python script
# ggplot(data = tfmotif_align_long, aes(x= x_id, y = y_id, size = expression.transform, col = NES) ) +
#   geom_point() +
#   scale_y_discrete(lim=rev)


#tfmotif_align_long <- tfmotif_align_long[!tfmotif_align_long$y_id %in% c('EBF1','GLI2'),]



tfmotif_align_long.keep <- subset(tfmotif_align_long, y_id %in% order_list)


saveRDS(tfmotif_align_long,'tfmotif_align_long.rds')


#write.table(tfmotif_align_long,file = 'tfmotif_align_long.txt',sep = '\t',col.names = TRUE,row.names = FALSE)

write.table(tfmotif_align_long.keep,file = 'tfmotif_align_long.txt',sep = '\t',col.names = TRUE,row.names = FALSE)




###################extract data from tfmotif.list.filter and output tf-peak-targetgene table###############


##tf_hi <- c('STAT4','STAT5A','JUNB','CEBPG','JUND','ZNF66','ZNF134','RELA','MAFG','TP63','PPARG','NR2F2')
tfmotif.list #TF-motif with multiple
tfmotif.list.filter #TF-motif-choose one and get heatmap, targetgene
tfmotif.list.filter.withdup


# names(tfmotif.list.filter[['c5']] )

# options(repr.plot.width = 7.5, repr.plot.height = 10)
# tfmotif.list.filter[['c5']][['GCM1']][['heatmap']]

# options(repr.plot.width = 7.5, repr.plot.height = 15)
# tfmotif.list.filter[['c9']][['GCM1']][['heatmap']]



# options(repr.plot.width = 7.5, repr.plot.height = 15)
# tfmotif.list.filter[['c2']][['STAT5A']][['heatmap']]

# options(repr.plot.width = 7.5, repr.plot.height = 35)
# tfmotif.list.filter.withdup[['c2']][['STAT5A']][['heatmap']]



# options(repr.plot.width = 7.5, repr.plot.height = 35)
# tfmotif.list.filter[['c1']][['FOSL1']][['heatmap']]

# options(repr.plot.width = 7.5, repr.plot.height = 45)
# tfmotif.list.filter.withdup[['c1']][['FOSL1']][['heatmap']]

# grep ('FLT1',tfmotif.list.filter.withdup[['c1']][['FOSL1']][['targetGenes_p2g_combined']],value=TRUE)
# grep ('LVRN',tfmotif.list.filter.withdup[['c1']][['FOSL1']][['targetGenes_p2g_combined']],value=TRUE)
# #'chr5:116046131-116046796|LVRN'
# grep ('LVRN',tfmotif.list.filter.withdup[['c1']][['FOSL1']][['targetGenes_p2g']],value=TRUE)
# #not found
# grep ('LVRN',tfmotif.list.filter.withdup[['c1']][['FOSL1']][['shared_targets']],value=TRUE) 
# #LVRN


# grep ('FLT1',tfmotif.list.filter[['c1']][['FOSL1']][['targetGenes_p2g_combined']],value=TRUE)
# #'chr13:28574821-28575743|FLT1'
# grep ('FLT1',tfmotif.list.filter[['c1']][['FOSL1']][['targetGenes_p2g']],value=TRUE)
# #'chr13:28574821-28575743|FLT1'
# grep ('FLT1',tfmotif.list.filter[['c1']][['FOSL1']][['shared_targets']],value=TRUE) 
# #'FLT1'


# grep ('LVRN',tfmotif.list.filter[['c1']][['FOSL1']][['targetGenes_p2g_combined']],value=TRUE)
# #not found
# grep ('LVRN',tfmotif.list.filter[['c1']][['FOSL1']][['targetGenes_p2g']],value=TRUE)
# #not found
# grep ('LVRN',tfmotif.list.filter[['c1']][['FOSL1']][['shared_targets']],value=TRUE) 
# #not found





#'cid''tf''NES''motif''peaks''nearGenes''peaks_combined''nearGenes_combined''targetGenes_dorc''targetGenes_dorc_combined''targetGenes_p2g''targetGenes_p2g_combined''shared_targets''heatmap'


############choose according to the tf-nes-expr dot-heatmap ? ################
tf_select_full <- list(
   'c6' = c('TP63','PPARG','TEAD4'),#,'NR2F2'), #NR2F2 expressed (highly), but has NES = 0 (filled value, no found actually)
    'c3' = c('TP63','PPARG','NR2F2'),
    'c9' = c('RELA','ZNF217','TEAD3','REL','MAFG','TP63','GRHL1','EGR3'),
    'c5' = c('BCL6','GCM1','TEAD3','MAFK'),
    'c1' = c('JUNB','JUND','CEBPG','FOSL2','PPARD','GRHL1','NR4A3'),
    'c2' = c('STAT4', 'STAT5A' ,'AR','MITF','ESRRG' )

)



#check 
for(cid in names(tf_select_full) ){
    tfs <- tf_select_full[[cid]]
    tfs <- tfs[tfs %in% names(tfmotif.list.filter[[cid]])]
    #tfs <- tfs[tfs %in% names(tfmotif.list.filter.withdup[[cid]])]
    cat(cid,' tfs found: ',paste(tfs,collapse = ','),'\n',sep='' )
    
    
}

c6 tfs found: TP63,PPARG,TEAD4
c3 tfs found: TP63,PPARG,NR2F2
c9 tfs found: RELA,ZNF217,TP63,GRHL1,EGR3
c5 tfs found: BCL6,TEAD3
c1 tfs found: JUNB,JUND,CEBPG,FOSL2,PPARD,GRHL1
c2 tfs found: STAT4,STAT5A,AR,MITF,ESRRG

##after filtering
tf_select <- list(
   'c6' = c('TP63','PPARG','TEAD4'),#,'NR2F2'), #NR2F2 expressed (highly), but has NES = 0 (filled value, no found actually)
    'c3' = c('TP63','PPARG','NR2F2'),
    'c9' = c('RELA','ZNF217','TP63','GRHL1','EGR3'),
    'c5' = c('BCL6','TEAD3'),
    'c1' = c('JUNB','JUND','CEBPG','FOSL2','PPARD','GRHL1'),
    'c2' = c('STAT4', 'STAT5A' ,'AR','MITF','ESRRG' )

)


##TF bundle 1

tf_select_1 <- list(
   'c6' = c('TP63','PPARG','TEAD4'),#,'NR2F2'), #NR2F2 expressed (highly), but has NES = 0 (filled value, no found actually)
    'c3' = c('TP63','PPARG','NR2F2'),
    'c9' = c('RELA','ZNF217','TP63','GRHL1'),
    'c5' = c('BCL6','TEAD3'),
    'c1' = c('JUNB','CEBPG','FOSL2','PPARD','GRHL1'),
    'c2' = c('STAT4', 'STAT5A' ,'AR','MITF','ESRRG' )

)


tf_select_2 <- list( #for c1, c2 network
   'c6' = c('TP63','PPARG','TEAD4'),#,'NR2F2'), #NR2F2 expressed (highly), but has NES = 0 (filled value, no found actually)
    'c3' = c('TP63','PPARG','NR2F2'),
    'c9' = c('RELA','ZNF217','TP63','GRHL1'),
    'c5' = c('BCL6','TEAD3'),
    'c1' = c('JUNB','CEBPG','FOSL2'),
    'c2' = c('STAT4', 'STAT5A' ,'AR','MITF')

)


##TF bundle 3 #add ERVFRD-1 regulatory TF TEAD4

tf_select_3 <- list( #for c5 network
   'c6' = c('TP63','PPARG','TEAD4'),#,'NR2F2'), #NR2F2 expressed (highly), but has NES = 0 (filled value, no found actually)
    'c3' = c('TP63','PPARG','NR2F2'),
    'c9' = c('RELA','ZNF217','GRHL1','TEAD4'),
    'c5' = c('BCL6','TEAD3'),
    'c1' = c('JUNB','CEBPG','FOSL2'),
    'c2' = c('STAT4', 'STAT5A' ,'AR','MITF' )

)


#tf_select <- tf_select

#tf_select <- tf_select_1

#tf_select <- tf_select_2

tf_select <- tf_select_3

##outdir

#outdir <- 'tf_peak_gene-network_useuniq/'
#outdir <- 'tf_peak_gene-network_useuniq/tf_select_test'

#outdir <- 'tf_peak_gene-network_useuniq/tf_select_1'
#outdir <- 'tf_peak_gene-network_useuniq/tf_select_2'
outdir <- 'tf_peak_gene-network_useuniq/tf_select_3'

if(!dir.exists(outdir) ){dir.create(outdir)}


#awk '($6 > 0.6 && $6 != "Cor"){print $6}' network.table.df.c6.txt |wcl
#796

awk '($6 > 0.6 && $6 != "Cor"){print $6}' network.table.df.c3.txt |wcl
#654

awk '($6 > 0.5 && $6 != "Cor"){print $6}' network.table.df.c1.txt |wcl
160 of total 218


cutoff.cor <- list(
    'c6' = 0.82,
    'c3' = 0.82,
    'c9' = 0,
    'c5' = 0,
    'c1' = 0,
    'c2' = 0


)



#::::::::::::::::::::##################start to build network ##############::::::::::::::::::::::#


##sum(duplicated(rownames(p2g.filter.uniq))) #0
##all.equal(rownames(p2g.filter.uniq),p2g.filter.uniq$peak)
#TRUE

p2g.filter.uniq.modify <- p2g.filter.uniq
rownames(p2g.filter.uniq.modify) <-  paste0(p2g.filter.uniq$peak,'|',p2g.filter.uniq$gene)
#sum(duplicated(rownames(p2g.filter.uniq.modify)))  #0
##modify rownames to extract correlation for edge table

saveRDS(p2g.filter.uniq.modify,'data_rds/p2g.filter.uniq.modify.rds')


#tf_select
saveRDS(tf_select,'data_rds/tf_select.rds')

#tfmotif.list.filter
saveRDS(tfmotif.list.filter,'data_rds/tfmotif.list.filter.rds')

#map_cluster
saveRDS(map_cluster,'data_rds/map_cluster.rds')


##exprPerc mat
exprMat.perc <- readRDS('/sda/mjwang/pwdex/placenta_10X_combine/02.snapATAC_harmony/chromVAR_TF_specific/full12526/TF_dev-vs-TF_expression/exprMat.perc.rds')
#243070 x 5

exprMat.perc$id <- paste0('c',exprMat.perc$id)
table(exprMat.perc$id)
  c1   c10    c2    c3    c4    c5    c6    c7    c8    c9 
24307 24307 24307 24307 24307 24307 24307 24307 24307 24307

exprMat.perc$features.plot <- as.character(exprMat.perc$features.plot)


#exprMat

colnames(exprMat.ave.z) <- paste0('c',colnames(exprMat.ave.z))

subset(exprMat.perc, features.plot == 'PAPPA')
exprMat.ave.z['PAPPA',]

cor(subset(exprMat.perc, features.plot == 'PAPPA')$avg.exp.scaled, unlist(exprMat.ave.z['PAPPA',]))
#0.982409310602705

cor(subset(exprMat.perc, features.plot == 'PAPPA')$avg.exp, unlist(exprMat.ave.z['PAPPA',]))
#0.982409310602705

cor(subset(exprMat.perc, features.plot == 'PAPPA')$pct.exp, unlist(exprMat.ave.z['PAPPA',]))
#0.967340836587857

sum(duplicated(rownames(exprMat.ave.z)))  #0
exprMat.ave.z$geneid <- rownames(exprMat.ave.z) #to avoid rowid duplication when extract data


##peak accessibility mat
peakMat.aggre <- readRDS('/sda/mjwang/pwdex/placenta_10X_combine/02.snapATAC_harmony/cicero_Granja/peakMat.aggre.rds')

colnames(peakMat.aggre) <- paste( 'c', colnames(peakMat.aggre),sep='')

sum(duplicated(rownames(peakMat.aggre))) #0

peakMat.aggre$peakid <- rownames(peakMat.aggre) #to avoid rowid duplication 




saveRDS(exprMat.perc,'data_rds/exprMat.perc.rds')
saveRDS(exprMat.ave.z,'data_rds/exprMat.ave.z.rds')
saveRDS(peakMat.aggre,'data_rds/peakMat.aggre.rds')
saveRDS(deg.list.up,'data_rds/deg.list.up.rds')


##deg labeling

length(deg.list.up[['c7']]) #246 total
length(deg.list.up[['c10']]) #694 total
length(deg.list.up[['c3']]) #242


lapply(deg.list.up,FUN = function(x){ length(x)  } )
$c8 1000 $c5 1000 $c9 894 
$c10 694 
$c6 137 
$c1 145 $c3 242 
$c2 133 $c4 123 $c7 246




#######automatic function#######

# ####read in presaved rds######
# ##p2g filter uniq modify table
# p2g.filter.uniq.modify <- readRDS('data_rds/p2g.filter.uniq.modify.rds')

# #map_cluster
# map_cluster <- readRDS('data_rds/map_cluster.rds')

# ##data rds
# exprMat.perc <- readRDS('data_rds/exprMat.perc.rds')
# exprMat.ave.z <- readRDS('data_rds/exprMat.ave.z.rds')
# peakMat.aggre <- readRDS('data_rds/peakMat.aggre.rds')

# deg.list.up <- readRDS('data_rds/deg.list.up.rds')




source('eGRN_build.r')


eGRN_build(tf_select = tf_select,
            outdir = outdir,
            p2g.filter.uniq.modify = p2g.filter.uniq.modify,
            map_cluster = map_cluster,
            exprMat.perc  = exprMat.perc,
            exprMat.ave.z  = exprMat.ave.z,
            peakMat.aggre = peakMat.aggre,
            deg.list.up = deg.list.up,
            cutoff.cor = cutoff.cor
          )

########tf_select_1#############
# 1. get tf-peak-targetgene table
# cluster  c6 
#  tfid  TP63 
#    filter peak-to-gene with cor cutoff: 0.82, 557 of total 683 filtered 
#  tfid  PPARG 
#    filter peak-to-gene with cor cutoff: 0.82, 174 of total 198 filtered 
#  tfid  TEAD4 
#    filter peak-to-gene with cor cutoff: 0.82, 525 of total 607 filtered 
# cluster  c3 
#  tfid  TP63 
#    filter peak-to-gene with cor cutoff: 0.82, 517 of total 664 filtered 
#  tfid  PPARG 
#    filter peak-to-gene with cor cutoff: 0.82, 225 of total 284 filtered 
#  tfid  NR2F2 
#    filter peak-to-gene with cor cutoff: 0.82, 204 of total 254 filtered 
# cluster  c9 
#  tfid  RELA 
#  tfid  ZNF217 
#  tfid  TP63 
#  tfid  GRHL1 
# cluster  c5 
#  tfid  BCL6 
#  tfid  TEAD3 
# cluster  c1 
#  tfid  JUNB 
#  tfid  CEBPG 
#  tfid  FOSL2 
#  tfid  PPARD 
#  tfid  GRHL1 
# cluster  c2 
#  tfid  STAT4 
#  tfid  STAT5A 
#  tfid  AR 
#  tfid  MITF 
#  tfid  ESRRG 
# 2. add data for node table
# add data for node of cluster  c6  with matched rna cluster  c9 
# add data for node of cluster  c3  with matched rna cluster  c5 
# add data for node of cluster  c9  with matched rna cluster  c10 
# add data for node of cluster  c5  with matched rna cluster  c6 
# add data for node of cluster  c1  with matched rna cluster  c7 
# add data for node of cluster  c2  with matched rna cluster  c3 
# 'ok to outdir: tf_peak_gene-network_useuniq/tf_select_1\n'















#######################
#####below code has been exported to eGRN_build.r, only for debuging##################
######################

############construction of  tf-cisElement-targetGene enhancer gene regulatory network##########

##get tf-peak-target_gene table
tf_peak_gene.df <- data.frame()
for(cid in names(tf_select) ){
    cat('cluster ',cid,'\n')

    for(tfid in tf_select[[cid]]){
      cat(' tfid ', tfid, '\n')
    
      stopifnot(!is.null(tfmotif.list.filter[[cid]][[tfid]]))
        
      hp <- tfmotif.list.filter[[cid]][[tfid]][['heatmap']]
      motifid <- tfmotif.list.filter[[cid]][[tfid]][['motif']]
      nes <-  tfmotif.list.filter[[cid]][[tfid]][['NES']]
        
      peaks_combined <- tfmotif.list.filter[[cid]][[tfid]][['peaks_combined']]
        
      targetGenes_p2g_combined <- tfmotif.list.filter[[cid]][[tfid]][['targetGenes_p2g_combined']]
      targetGenes_dorc_combined <- tfmotif.list.filter[[cid]][[tfid]][['targetGenes_dorc_combined']]
      nearGene_combined <- tfmotif.list.filter[[cid]][[tfid]][['nearGenes_combined']]
    
      shared_targets <- tfmotif.list.filter[[cid]][[tfid]][['shared_targets']]
        
      targetGenes_all <- unique(targetGenes_p2g_combined)  #use all motif targets hit p2g_filter_uniq
      #targetGenes_all <- unique(targetGenes_dorc_combined) #p2g has many one-to-many peak2gene links
      #targetGenes_all <- unique(c(targetGenes_p2g_combined,targetGenes_dorc_combined,nearGene_combined))
      targetGenes_all.peaks <- sapply(strsplit(targetGenes_all,split = '\\|'), function(x){x[1]}  )
      targetGenes_all.genes <- sapply(strsplit(targetGenes_all,split = '\\|'), function(x){x[2]}  )
      targetGenes_all.peaks <- unique(targetGenes_all.peaks)
      targetGenes_all.genes <- unique(targetGenes_all.genes) 

    
      stopifnot( isTRUE(all.equal(sort(targetGenes_all.peaks),sort(peaks_combined)) ) )#TRUE
      ##stopifnot(all.equal(sort(targetGenes_all.genes),sort(shared_targets)) )#TRUE
      
      #peaks_combined and shared_targets are the complete set
        
      ##choose targetGenes_all (if combined equal to peaks_combined and shared_targets, because use dar_overlap_p2g)
      
      if( isTRUE(all.equal(sort(targetGenes_all.peaks) ,sort(peaks_combined)) ) ){ 
          #for(i in targetGenes_all){    
              targetGenes_all.df <- as.data.frame(Reduce(rbind,strsplit(x = targetGenes_all, split = "\\|")),stringsAsFactors = FALSE)
              rownames(targetGenes_all.df) <- NULL
              colnames(targetGenes_all.df) <- c('peak','target')
              targetGenes_all.df$tf <- tfid
              targetGenes_all.df$dar <- cid
              targetGenes_all.df$NES <- nes
              
              #get peak2gene correlation from p2g.filter.uniq table
              p2gid <- paste0(targetGenes_all.df$peak,'|',targetGenes_all.df$target)
              stopifnot(all.equal(targetGenes_all,p2gid))
              stopifnot(sum(p2gid %in% paste0(p2g.filter.uniq.modify$peak,'|',p2g.filter.uniq.modify$gene) ) == length(p2gid) ) 
              
              cor.df <- p2g.filter.uniq.modify[p2gid,c('peak','Correlation')]
              stopifnot(all.equal(targetGenes_all.df$peak, cor.df$peak))
              cor.value <- cor.df$Correlation
              stopifnot(sum(is.na(cor.value)) == 0)
              targetGenes_all.df$cor <- cor.value
          
              #edge table
              targetGenes_all.df <- targetGenes_all.df[,c('dar','tf','peak','target','NES','cor')]
          #}
      }else{ #use peaks_combined as the whole set, add target gene as 'NA'
          cat("targetGenes_p2g_combined not eq peak_combined, use peak_combined and fill target gene as NA \n")
          peaks.df <- data.frame(peak = peaks_combined, stringsAsFactors = FALSE)
          temp.df <- as.data.frame(Reduce(rbind,strsplit(x = targetGenes_all, split = "\\|")),stringsAsFactors = FALSE)
          rownames( temp.df ) <- NULL
          colnames( temp.df ) <- c('peak','target')
           temp.df$tf <- tfid
           temp.df$dar <- cid
          stopifnot(  sum(temp.df$peak %in% peaks_combined) == nrow(temp.df) )
          #merge
          targetGenes_all.df <- base::merge(x = peaks.df, y = temp.df, by = 'peak', all.x = TRUE)
          
          targetGenes_all.df$tf <- tfid
          targetGenes_all.df$dar <- cid
          #targetGenes_all.df$NES <- nes
          
          targetGenes_all.df <- targetGenes_all.df[,c('dar','tf','peak','target')]
      }
        
      tf_peak_gene.df <- rbind.data.frame(tf_peak_gene.df,targetGenes_all.df)
        
        
    }
}



tf_peak_gene.df #3959 #2836 #6460 #2138 use all peaks combined - target/near gene

table(tf_peak_gene.df$dar)

c1   c2   c3   c5   c6   c9  #use p2g.filter.uniq with new tf_select
 500  153 1202  217 1488  399 

  c1   c2   c3   c6   c9 #use p2g.filter.uniq
 473  153 1202  881  127 

 c1   c2   c3   c6   c9 
 854  241 2872 2151  261

# c1  c2  c3  c6  c9  #docr only
# 156  57 232 226  92 

# c1  c2  c3  c6  c9  #p2g only
# 225 104 661 667 211 

# c1  c2  c3  c6  c9 #all combined
# 291 124 730 746 247

table(tf_peak_gene.df$tf)

    AR   BCL6  CEBPG  ESRRG  FOSL2  GRHL1   JUNB   JUND   MITF  NR2F2  PPARD 
    20     39     30     13    123    108    123    157     13    254     40 
 PPARG   RELA  STAT4 STAT5A  TEAD3  TEAD4   TP63 ZNF217 
   482     46     49     58    178    607   1538     81 


#     AR  CEBPG  ESRRG  FOSL2   JUNB   JUND   MITF  NR2F2  PPARD  PPARG   RELA 
#         20     30     13    123    123    157     13    254     40    482     46 
#      STAT4 STAT5A   TP63 ZNF217 
#         49     58   1347     81 

#     AR  CEBPG  ESRRG  FOSL2   JUNB   JUND   MITF  NR2F2  PPARD  PPARG   RELA 
#     32     61     22    215    221    276     15    581    81   1167    108 
#  STAT4 STAT5A   TP63 ZNF217 
#     79     93   3275    153 


# CEBPG   JUNB   JUND   MAFG  NR2F2  PPARG   RELA  STAT4 STAT5A   TP63 ZNF134 #dorc only
#     22     58     76     24     41    103     22     26     31    314     27 
#  ZNF66 
#     19

#  CEBPG   JUNB   JUND   MAFG  NR2F2  PPARG   RELA  STAT4 STAT5A   TP63 ZNF134 #p2g only
#     32     82    111     55    105    313     61     49     55    910     55 
#  ZNF66 
#     40 

#  CEBPG   JUNB   JUND   MAFG  NR2F2  PPARG   RELA  STAT4 STAT5A   TP63 ZNF134 #all combined
#     41    106    144     61    111    343     68     58     66   1022     71 
#  ZNF66 
#     47 

#table(tf_peak_gene.df$target)

##saveRDS(tf_peak_gene.df,'tf_peak_gene-network_useuniq/tf_peak_gene.df.motifcombined.p2guniq.rds')
##write.table(tf_peak_gene.df,file = 'tf_peak_gene-network_useuniq/tf_peak_gene.df.motifcombined.p2guniq.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

saveRDS(tf_peak_gene.df,paste(outdir,'/tf_peak_gene.df.motifcombined.p2guniq.rds',sep='') )
write.table(tf_peak_gene.df,file = paste(outdir,'/tf_peak_gene.df.motifcombined.p2guniq.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)


##saveRDS(tf_peak_gene.df,'tf_peak_gene-network/tf_peak_gene.df.motifcombined.rds')
##write.table(tf_peak_gene.df,file = 'tf_peak_gene-network/tf_peak_gene.df.motifcombined.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)


#saveRDS(tf_peak_gene.df,'tf_peak_gene-network/tf_peak_gene.df.dorconly.rds')
#write.table(tf_peak_gene.df,file = 'tf_peak_gene-network/tf_peak_gene.df.dorconly.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)




# tf_peak_gene.df #1813 x 4
# table(tf_peak_gene.df$dar)

#  c1  c2  c3  c6  c9 
# 134  77 628 802 172
# table(tf_peak_gene.df$tf)
#  CEBPG   JUNB   JUND   MAFG  NR2F2  PPARG   RELA  STAT4 STAT5A   TP63 ZNF134 
#     20     45     69     45    101    464     52     38     39    865     44 
#  ZNF66 
#     31

# table(tf_peak_gene.df$target_p2g_combined)


# saveRDS(tf_peak_gene.df,'tf_peak_gene.df.rds')

# write.table(tf_peak_gene.df,file = 'tf_peak_gene.df.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)


##split into pairwise format (edge table format?) as tf-peak (cistrome), peak-targetgene (regulon)

cistrome.df <- tf_peak_gene.df[,c('dar','tf','peak','NES')]
cistrome.df$type <- 'cistrome'
colnames(cistrome.df) <- c('dar','name1','name2','score','type')


regulon.df <- tf_peak_gene.df[,c('dar','peak','target','cor')]
#regulon.df <- tf_peak_gene.df[,c('dar','peak','target_p2g_combined')]
regulon.df$type <- 'regulon'
colnames(regulon.df) <- c('dar','name1','name2','score','type')


network.table.df <- rbind.data.frame(cistrome.df,regulon.df)


#########dedup#########

flag1 <- duplicated(paste(network.table.df$dar,network.table.df$name1,network.table.df$name2,sep='|') )
sum(flag1) #1081, only filter within dar duplicated edges

flag2 <- duplicated(paste(network.table.df$name1,network.table.df$name2,sep='|') )
sum(flag2) #2355, filter all duplicated edges 

flag3 <- flag1 == FALSE & flag2 == TRUE
sum(flag3) #1274 duplicated between dars (keep, c1 c2 very few!)

sum(flag1)+sum(flag3) == sum(flag2)


table(network.table.df[flag1,'dar'])
table(network.table.df[flag1,c('dar','type')])
 c1  c2  c3  c5  c6  c9 
282  55 327  16 302  99 
    type
dar  regulon
  c1     282
  c2      55
  c3     327
  c5      16
  c6     302
  c9      99

table(network.table.df[flag2,'dar'])
table(network.table.df[flag2,c('dar','type')])
c1   c2   c3   c5   c6   c9 
 284   60 1461   56  302  192
     type
dar  cistrome regulon
  c1        0     284
  c2        0      60
  c3      579     882
  c5        0      56
  c6        0     302
  c9       45     147

table(network.table.df[flag3,'dar'])
table(network.table.df[flag3,c('dar','type')])
 c1   c2   c3   c5   c9 
 2    5   1134  40   93 

    type
dar  cistrome regulon
  c1        0       2
  c2        0       5
  c3      579     555
  c5        0      40
  c9       45      48


network.table.df[flag1,] #all duplications within dar, must remove
network.table.df[flag3,] #duplications among dars, keep?

network.table.df[grep('chr9:116161112-116161355',network.table.df$name1),] #duplicated within one dar cluster, must remove!
network.table.df[grep('chr5:38808406-38808680',network.table.df$name1),] #must remove


network.table.df[grep('chr10:99615420-99615780',network.table.df$name1),] #duplicated in more than one dar cluster, must keep! (because of TF to multiple peak,  peak to the same target in multiple <similar?> clusters )
5311	c6	chr10:99615420-99615780	COX15	0.49936308425407	regulon
7121	c5	chr10:99615420-99615780	COX15	0.49936308425407	regulon

network.table.df[grep('chr11:118267463-118268021',network.table.df$name1),] 
6302	c3	chr11:118267463-118268021	KMT2A	0.675514949879478	regulon
7823	c2	chr11:118267463-118268021	KMT2A	0.675514949879478	regulon

network.table.df[grep('chr1:53682161-53682576',network.table.df$name1),]  
6201	c3	chr1:53682161-53682576	YIPF1	0.694939564534203	regulon
6479	c3	chr1:53682161-53682576	YIPF1	0.694939564534203	regulon
7704	c1	chr1:53682161-53682576	YIPF1	0.694939564534203	regulon

network.table.df[grep('chr1:115239183-115239827',network.table.df$name2),]
30	c6	TP63	chr1:115239183-115239827	7.290964824	cistrome
909	c6	TEAD4	chr1:115239183-115239827	4.716354626	cistrome
1515	c3	TP63	chr1:115239183-115239827	11.78049249	cistrome
2449	c3	NR2F2	chr1:115239183-115239827	4.687216039	cistr



####################


flag <- flag1 #only remove within dar duplicated edge, keep among dar edge duplications
#network.table.df[flag,]



#1081#if dar-name1-name2 #2355 if name1-name2

#1862 (use p2g.filter.uniq)

#7057

#2480
#2345

network.table.df.dedup <- network.table.df[!flag,]
#6837 remove within dar edge duplications
#5563 remove all edge duplications

#1043 dorc only
#1281 combined

##split score to NES and Cor, fill with NA (cytoscape will fail to read in ,use 0)

nes_value <- ifelse(network.table.df.dedup$type =='cistrome', network.table.df.dedup$score ,0)
cor_value <- ifelse(network.table.df.dedup$type =='regulon', network.table.df.dedup$score ,0)

network.table.df.dedup$NES <- sprintf("%.5f",as.numeric(nes_value))
network.table.df.dedup$Cor <- sprintf("%.5f",as.numeric(cor_value)) 

network.table.df.dedup$score <- NULL
#network.table.df.dedup$test <- 1


##save edge table to file


write.table(network.table.df.dedup,file = paste(outdir,'/network.table.df.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

write.table(subset(network.table.df.dedup,dar=='c2'),file = paste(outdir,'/network.table.df.c2.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

write.table(subset(network.table.df.dedup,dar=='c1'),file = paste(outdir,'/network.table.df.c1.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

write.table(subset(network.table.df.dedup,dar=='c9'),file = paste(outdir,'/network.table.df.c9.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

write.table(subset(network.table.df.dedup,dar=='c5'),file = paste(outdir,'/network.table.df.c5.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

write.table(subset(network.table.df.dedup,dar=='c6'),file = paste(outdir,'/network.table.df.c6.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

write.table(subset(network.table.df.dedup,dar=='c3'),file = paste(outdir,'/network.table.df.c3.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)


# write.table(network.table.df.dedup,file = 'tf_peak_gene-network/network.table.df.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

# write.table(subset(network.table.df.dedup,dar=='c2'),file = 'tf_peak_gene-network/network.table.df.c2.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

# write.table(subset(network.table.df.dedup,dar=='c1'),file = 'tf_peak_gene-network/network.table.df.c1.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

# write.table(subset(network.table.df.dedup,dar=='c9'),file = 'tf_peak_gene-network/network.table.df.c9.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

# write.table(subset(network.table.df.dedup,dar=='c5'),file = 'tf_peak_gene-network/network.table.df.c5.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

# write.table(subset(network.table.df.dedup,dar=='c6'),file = 'tf_peak_gene-network/network.table.df.c6.txt',col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)



#############add data (of this cluster) for node: tf (expression <and FC >), peak (accessibility) , gene (expression and percentage)################



#create node attribution table
##start to fill node data
tf_peak_gene.node.list <- list()

for(cid in names(tf_select)){
    
    stopifnot(!is.null(map_cluster[[cid]]))
    cid_map <- map_cluster[[cid]]
    
    cat ('add data for node of cluster ', cid, ' with matched rna cluster ',cid_map,'\n')
    tf_peak_gene.df.sel <- subset(tf_peak_gene.df, dar == cid)
    
    node.df <- rbind.data.frame(
        unique(data.frame(name = tf_peak_gene.df.sel$tf,type = 'TF',stringsAsFactors = FALSE) ),
        unique(data.frame(name = tf_peak_gene.df.sel$peak,type = 'region',stringsAsFactors = FALSE) ),
        unique(data.frame(name = tf_peak_gene.df.sel$target,type = 'gene',stringsAsFactors = FALSE) ),
        #unique(data.frame(name = tf_peak_gene.df.sel$target_p2g_combined,type = 'gene',stringsAsFactors = FALSE) ),
        stringsAsFactors = FALSE
    )

    ##add tf gene expression, and percentage
    
    sum(duplicated(node.df$name)) != 0 #target gene or peak may duplicated 
    #node.df[which(duplicated(node.df$name)),]
    
    ##isDEG or isShow ?
  
    deg <- deg.list.up[[cid_map]]
    deg.sel <- deg[node.df$name ]
  
    deg.label <- names(deg.sel)
    
    deg.label[is.na(deg.label)] <- ''
    deg.label[which(node.df$type == 'TF')] <- node.df$name[which(node.df$type == 'TF')]
    
    deg.log2FC <- deg.sel
    isDEG <- ifelse(is.na(deg.label),'No','Yes')
    names(deg.log2FC) <- NULL
    
    deg.sel.df <- data.frame('isDEG'=isDEG,'deg_label'=deg.label,'log2FC'=deg.log2FC, stringsAsFactors = FALSE)
    
    ##exprPerc
   exprMat.perc.sel <- subset(exprMat.perc, id == cid_map)
    stopifnot( sum(duplicated(exprMat.perc.sel$features.plot)) == 0 )
    rownames(exprMat.perc.sel) <- exprMat.perc.sel$features.plot
    
    expr_perc <- exprMat.perc.sel[node.df$name,c('features.plot','pct.exp')]
    
    ##exprValue
    
    #expr_value <-  exprMat.ave.z[node.df$name,cid,drop = FALSE]
    #expr_value$geneid <- rownames(expr_value) #will add .1 if duplicated
    expr_value <-  exprMat.ave.z[node.df$name,c('geneid',cid_map)]
    #expr_value$geneid <- rownames(expr_value)
    colnames(expr_value) <- c('geneid','value')
    #expr_value <- expr_value[,c(2,1)]
    
    #node.df.data <- cbind.data.frame(node.df,expr_perc,expr_value)
    node.df.data <- cbind.data.frame(node.df,deg.sel.df,expr_perc,expr_value)
     
    stopifnot(all.equal(node.df.data[!is.na(node.df.data$log2FC),'name'],node.df.data[!is.na(node.df.data$log2FC),'deg_label']) )
    stopifnot(all.equal(node.df.data[!is.na(node.df.data$pct.exp),'name'],node.df.data[!is.na(node.df.data$pct.exp),'features.plot']) ) #TRUE
    stopifnot(all.equal(node.df.data[!is.na(node.df.data$value),'name'],node.df.data[!is.na(node.df.data$value),'geneid']) )#TRUE

    ##add peak accessibility
    
    #acc_value <- peakMat.aggre[node.df$name,cid,drop = FALSE]
    #acc_value$peakid <- rownames(acc_value)
    acc_value <- peakMat.aggre[node.df$name,c('peakid',cid)]
    colnames(acc_value) <- c('peak','acc_value')
    #acc_value <- acc_value[,c(2,1)]
    
    node.df.data <- cbind.data.frame(node.df.data,acc_value)
    
    stopifnot(all.equal(node.df.data[!is.na(node.df.data$acc_value),'name'],node.df.data[!is.na(node.df.data$acc_value),'peak']) )
    
    colnames(node.df.data) <- c('name',	'type', 'isDEG', 'deg_label','log2FC','label',	'pct.exp',	'geneid', 'expr.value',	'peakid','acc_value')
    
#     #NA to 0 for value column
#     node.df.data$pct.exp <-  ifelse(is.na(node.df.data$pct.exp), 0  , node.df.data$pct.exp)
#     node.df.data$expr.value <-  ifelse(is.na(node.df.data$pct.exp), 0  , node.df.data$pct.exp)
#     node.df.data$acc_value <-  ifelse(is.na(node.df.data$pct.exp), 0  , node.df.data$pct.exp)
    
    tf_peak_gene.node.list[[cid]] <- node.df.data
}

# add data for node of cluster  c2  with matched rna cluster  c3 
# add data for node of cluster  c1  with matched rna cluster  c7 
# add data for node of cluster  c9  with matched rna cluster  c10 
# add data for node of cluster  c6  with matched rna cluster  c9 
# add data for node of cluster  c3  with matched rna cluster  c5 

add data for node of cluster  c6  with matched rna cluster  c9 
add data for node of cluster  c3  with matched rna cluster  c5 
add data for node of cluster  c9  with matched rna cluster  c10 
add data for node of cluster  c5  with matched rna cluster  c6 
add data for node of cluster  c1  with matched rna cluster  c7 
add data for node of cluster  c2  with matched rna cluster  c3 

saveRDS(tf_peak_gene.node.list,paste(outdir,'/tf_peak_gene.node.list.rds',sep='') )


write.table(tf_peak_gene.node.list[['c2']],file=paste(outdir,'/tf_peak_gene.node.c2.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)

write.table(tf_peak_gene.node.list[['c1']],file=paste(outdir,'/tf_peak_gene.node.c1.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



write.table(tf_peak_gene.node.list[['c9']],file=paste(outdir,'/tf_peak_gene.node.c9.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



write.table(tf_peak_gene.node.list[['c5']],file=paste(outdir,'/tf_peak_gene.node.c5.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



write.table(tf_peak_gene.node.list[['c6']],file=paste(outdir,'/tf_peak_gene.node.c6.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)


write.table(tf_peak_gene.node.list[['c3']],file=paste(outdir,'/tf_peak_gene.node.c3.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)


# saveRDS(tf_peak_gene.node.list,'tf_peak_gene-network/tf_peak_gene.node.list.rds')


# write.table(tf_peak_gene.node.list[['c2']],file='tf_peak_gene-network/tf_peak_gene.node.c2.txt',sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)

# write.table(tf_peak_gene.node.list[['c1']],file='tf_peak_gene-network/tf_peak_gene.node.c1.txt',sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



# write.table(tf_peak_gene.node.list[['c9']],file='tf_peak_gene-network/tf_peak_gene.node.c9.txt',sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



# write.table(tf_peak_gene.node.list[['c5']],file='tf_peak_gene-network/tf_peak_gene.node.c5.txt',sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



# write.table(tf_peak_gene.node.list[['c6']],file='tf_peak_gene-network/tf_peak_gene.node.c6.txt',sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)







###color set in cytoscape

peak border color:  505F93, alpha 100






eGRN_build <- function( tf_select = NULL,
                        outdir = NULL,
                        p2g.filter.uniq.modify = NULL,
                        map_cluster = NULL,
                        exprMat.perc  = NULL,
                        exprMat.ave.z  = NULL,
                        peakMat.aggre = NULL,
                        deg.list.up = NULL,
                        cutoff.cor = NULL ){
    ##construction of  tf-cisElement-targetGene enhancer gene regulatory network

    cat('1. get tf-peak-targetgene table\n')

    ##get tf-peak-target_gene table
    tf_peak_gene.df <- data.frame()
    for(cid in names(tf_select) ){
        cat('cluster ',cid,'\n')

        for(tfid in tf_select[[cid]]){
          cat(' tfid ', tfid, '\n')

          stopifnot(!is.null(tfmotif.list.filter[[cid]][[tfid]]))

          hp <- tfmotif.list.filter[[cid]][[tfid]][['heatmap']]
          motifid <- tfmotif.list.filter[[cid]][[tfid]][['motif']]
          nes <-  tfmotif.list.filter[[cid]][[tfid]][['NES']]

          peaks_combined <- tfmotif.list.filter[[cid]][[tfid]][['peaks_combined']]

          targetGenes_p2g_combined <- tfmotif.list.filter[[cid]][[tfid]][['targetGenes_p2g_combined']]
          targetGenes_dorc_combined <- tfmotif.list.filter[[cid]][[tfid]][['targetGenes_dorc_combined']]
          nearGene_combined <- tfmotif.list.filter[[cid]][[tfid]][['nearGenes_combined']]

          shared_targets <- tfmotif.list.filter[[cid]][[tfid]][['shared_targets']]

          targetGenes_all <- unique(targetGenes_p2g_combined)  #use all motif targets hit p2g_filter_uniq
          #targetGenes_all <- unique(targetGenes_dorc_combined) #p2g has many one-to-many peak2gene links
          #targetGenes_all <- unique(c(targetGenes_p2g_combined,targetGenes_dorc_combined,nearGene_combined))
          targetGenes_all.peaks <- sapply(strsplit(targetGenes_all,split = '\\|'), function(x){x[1]}  )
          targetGenes_all.genes <- sapply(strsplit(targetGenes_all,split = '\\|'), function(x){x[2]}  )
          targetGenes_all.peaks <- unique(targetGenes_all.peaks)
          targetGenes_all.genes <- unique(targetGenes_all.genes) 


          stopifnot( isTRUE(all.equal(sort(targetGenes_all.peaks),sort(peaks_combined)) ) )#TRUE
          ##stopifnot(all.equal(sort(targetGenes_all.genes),sort(shared_targets)) )#TRUE

          #peaks_combined and shared_targets are the complete set

          ##choose targetGenes_all (if combined equal to peaks_combined and shared_targets, because use dar_overlap_p2g)

          if( isTRUE(all.equal(sort(targetGenes_all.peaks) ,sort(peaks_combined)) ) ){ 
              #for(i in targetGenes_all){    
                  targetGenes_all.df <- as.data.frame(Reduce(rbind,strsplit(x = targetGenes_all, split = "\\|")),stringsAsFactors = FALSE)
                  rownames(targetGenes_all.df) <- NULL
                  colnames(targetGenes_all.df) <- c('peak','target')
                  targetGenes_all.df$tf <- tfid
                  targetGenes_all.df$dar <- cid
                  targetGenes_all.df$NES <- nes

                  #get peak2gene correlation from p2g.filter.uniq table
                  p2gid <- paste0(targetGenes_all.df$peak,'|',targetGenes_all.df$target)
                  stopifnot(all.equal(targetGenes_all,p2gid))
                  stopifnot(sum(p2gid %in% paste0(p2g.filter.uniq.modify$peak,'|',p2g.filter.uniq.modify$gene) ) == length(p2gid) ) 

                  cor.df <- p2g.filter.uniq.modify[p2gid,c('peak','Correlation')]
                  stopifnot(all.equal(targetGenes_all.df$peak, cor.df$peak))
                  cor.value <- cor.df$Correlation
                  stopifnot(sum(is.na(cor.value)) == 0)
                  targetGenes_all.df$cor <- cor.value

                  stopifnot(sum(cor.value <= 0) == 0)
              
                  ##add additional filter layers for network with high density
                  if(cutoff.cor[[cid]] == 'auto'){
                      qs <- quantile(x = targetGenes_all.df$cor, probs = seq(from = 0, to =1, by = 0.1) )
                      cutoff <- qs['25%']
                      targetGenes_all.df <- subset(targetGenes_all.df,cor >= cutoff)
                      
                  }else{
                    len.filtered <- sum(targetGenes_all.df$cor < cutoff.cor[[cid]])
                    len.total <- length(targetGenes_all.df$cor)
                    if(len.filtered != 0){
                      cat('   filter peak-to-gene with cor cutoff: ',cutoff.cor[[cid]],', ',len.filtered,' of total ',len.total,' filtered \n',sep='')
                    }else{  }
                      
                    targetGenes_all.df <- subset(targetGenes_all.df,cor >= cutoff.cor[[cid]])
                  }
                  stopifnot(nrow(targetGenes_all.df) != 0 )
              
                  #edge table
                  targetGenes_all.df <- targetGenes_all.df[,c('dar','tf','peak','target','NES','cor')]
              #}
          }else{ #use peaks_combined as the whole set, add target gene as 'NA'
              cat("targetGenes_p2g_combined not eq peak_combined, use peak_combined and fill target gene as NA \n")
              peaks.df <- data.frame(peak = peaks_combined, stringsAsFactors = FALSE)
              temp.df <- as.data.frame(Reduce(rbind,strsplit(x = targetGenes_all, split = "\\|")),stringsAsFactors = FALSE)
              rownames( temp.df ) <- NULL
              colnames( temp.df ) <- c('peak','target')
               temp.df$tf <- tfid
               temp.df$dar <- cid
              stopifnot(  sum(temp.df$peak %in% peaks_combined) == nrow(temp.df) )
              #merge
              targetGenes_all.df <- base::merge(x = peaks.df, y = temp.df, by = 'peak', all.x = TRUE)

              targetGenes_all.df$tf <- tfid
              targetGenes_all.df$dar <- cid
              #targetGenes_all.df$NES <- nes

              targetGenes_all.df <- targetGenes_all.df[,c('dar','tf','peak','target')]
          }
            
            
          tf_peak_gene.df <- rbind.data.frame(tf_peak_gene.df,targetGenes_all.df)


        }
    }




    saveRDS(tf_peak_gene.df,paste(outdir,'/tf_peak_gene.df.motifcombined.p2guniq.rds',sep='') )
    write.table(tf_peak_gene.df,file = paste(outdir,'/tf_peak_gene.df.motifcombined.p2guniq.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)



    ##split into pairwise format (edge table format?) as tf-peak (cistrome), peak-targetgene (regulon)

    cistrome.df <- tf_peak_gene.df[,c('dar','tf','peak','NES')]
    cistrome.df$type <- 'cistrome'
    colnames(cistrome.df) <- c('dar','name1','name2','score','type')


    regulon.df <- tf_peak_gene.df[,c('dar','peak','target','cor')]
    #regulon.df <- tf_peak_gene.df[,c('dar','peak','target_p2g_combined')]
    regulon.df$type <- 'regulon'
    colnames(regulon.df) <- c('dar','name1','name2','score','type')


    network.table.df <- rbind.data.frame(cistrome.df,regulon.df)


    #########dedup#########

    flag1 <- duplicated(paste(network.table.df$dar,network.table.df$name1,network.table.df$name2,sep='|') )
    #sum(flag1) #1081, only filter within dar duplicated edges

    flag2 <- duplicated(paste(network.table.df$name1,network.table.df$name2,sep='|') )
    #sum(flag2) #2355, filter all duplicated edges 

    flag3 <- flag1 == FALSE & flag2 == TRUE
    #sum(flag3) #1274 duplicated between dars (keep, c1 c2 very few!)

    #sum(flag1)+sum(flag3) == sum(flag2)

    flag <- flag1 #only remove within dar duplicated edge, keep among dar edge duplications
    #network.table.df[flag,]


    network.table.df.dedup <- network.table.df[!flag,]
 

    ##split score to NES and Cor, fill with NA (cytoscape will fail to read in ,use 0)

    nes_value <- ifelse(network.table.df.dedup$type =='cistrome', network.table.df.dedup$score ,0)
    cor_value <- ifelse(network.table.df.dedup$type =='regulon', network.table.df.dedup$score ,0)

    network.table.df.dedup$NES <- sprintf("%.5f",as.numeric(nes_value))
    network.table.df.dedup$Cor <- sprintf("%.5f",as.numeric(cor_value)) 

    network.table.df.dedup$score <- NULL
    #network.table.df.dedup$test <- 1


    ##save edge table to file


    write.table(network.table.df.dedup,file = paste(outdir,'/network.table.df.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

    write.table(subset(network.table.df.dedup,dar=='c2'),file = paste(outdir,'/network.table.df.c2.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

    write.table(subset(network.table.df.dedup,dar=='c1'),file = paste(outdir,'/network.table.df.c1.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

    write.table(subset(network.table.df.dedup,dar=='c9'),file = paste(outdir,'/network.table.df.c9.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

    write.table(subset(network.table.df.dedup,dar=='c5'),file = paste(outdir,'/network.table.df.c5.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

    write.table(subset(network.table.df.dedup,dar=='c6'),file = paste(outdir,'/network.table.df.c6.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)

    write.table(subset(network.table.df.dedup,dar=='c3'),file = paste(outdir,'/network.table.df.c3.txt',sep=''),col.names = TRUE, row.names = FALSE, sep='\t',quote=FALSE)


    cat('2. add data for node table\n')

    #############add data (of this cluster) for node: tf (expression <and FC >), peak (accessibility) , gene (expression and percentage)################



    #create node attribution table
    ##start to fill node data
    tf_peak_gene.node.list <- list()

    for(cid in names(tf_select)){

        stopifnot(!is.null(map_cluster[[cid]]))
        cid_map <- map_cluster[[cid]]

        cat ('add data for node of cluster ', cid, ' with matched rna cluster ',cid_map,'\n')
        tf_peak_gene.df.sel <- subset(tf_peak_gene.df, dar == cid)

        node.df <- rbind.data.frame(
            unique(data.frame(name = tf_peak_gene.df.sel$tf,type = 'TF',stringsAsFactors = FALSE) ),
            unique(data.frame(name = tf_peak_gene.df.sel$peak,type = 'region',stringsAsFactors = FALSE) ),
            unique(data.frame(name = tf_peak_gene.df.sel$target,type = 'gene',stringsAsFactors = FALSE) ),
            #unique(data.frame(name = tf_peak_gene.df.sel$target_p2g_combined,type = 'gene',stringsAsFactors = FALSE) ),
            stringsAsFactors = FALSE
        )

        ##add tf gene expression, and percentage

        sum(duplicated(node.df$name)) != 0 #target gene or peak may duplicated 
        #node.df[which(duplicated(node.df$name)),]

        ##isDEG or isShow ?

        deg <- deg.list.up[[cid_map]]
        deg.sel <- deg[node.df$name ]

        deg.label <- names(deg.sel)

        deg.label[is.na(deg.label)] <- ''
        deg.label[which(node.df$type == 'TF')] <- node.df$name[which(node.df$type == 'TF')]

        deg.log2FC <- deg.sel
        isDEG <- ifelse(deg.label == '','No','Yes')
        names(deg.log2FC) <- NULL

        deg.sel.df <- data.frame('isDEG'=isDEG,'deg_label'=deg.label,'log2FC'=deg.log2FC, stringsAsFactors = FALSE)

        ##exprPerc
       exprMat.perc.sel <- subset(exprMat.perc, id == cid_map)
        stopifnot( sum(duplicated(exprMat.perc.sel$features.plot)) == 0 )
        rownames(exprMat.perc.sel) <- exprMat.perc.sel$features.plot

        expr_perc <- exprMat.perc.sel[node.df$name,c('features.plot','pct.exp')]

        ##exprValue

        #expr_value <-  exprMat.ave.z[node.df$name,cid,drop = FALSE]
        #expr_value$geneid <- rownames(expr_value) #will add .1 if duplicated
        expr_value <-  exprMat.ave.z[node.df$name,c('geneid',cid_map)]
        #expr_value$geneid <- rownames(expr_value)
        colnames(expr_value) <- c('geneid','value')
        #expr_value <- expr_value[,c(2,1)]

        #node.df.data <- cbind.data.frame(node.df,expr_perc,expr_value)
        node.df.data <- cbind.data.frame(node.df,deg.sel.df,expr_perc,expr_value)

        stopifnot(all.equal(node.df.data[!is.na(node.df.data$log2FC),'name'],node.df.data[!is.na(node.df.data$log2FC),'deg_label']) )
        stopifnot(all.equal(node.df.data[!is.na(node.df.data$pct.exp),'name'],node.df.data[!is.na(node.df.data$pct.exp),'features.plot']) ) #TRUE
        stopifnot(all.equal(node.df.data[!is.na(node.df.data$value),'name'],node.df.data[!is.na(node.df.data$value),'geneid']) )#TRUE

        ##add peak accessibility

        #acc_value <- peakMat.aggre[node.df$name,cid,drop = FALSE]
        #acc_value$peakid <- rownames(acc_value)
        acc_value <- peakMat.aggre[node.df$name,c('peakid',cid)]
        colnames(acc_value) <- c('peak','acc_value')
        #acc_value <- acc_value[,c(2,1)]

        node.df.data <- cbind.data.frame(node.df.data,acc_value)

        stopifnot(all.equal(node.df.data[!is.na(node.df.data$acc_value),'name'],node.df.data[!is.na(node.df.data$acc_value),'peak']) )

        colnames(node.df.data) <- c('name',	'type', 'isDEG', 'deg_label','log2FC','label',	'pct.exp',	'geneid', 'expr.value',	'peakid','acc_value')

    #     #NA to 0 for value column
    #     node.df.data$pct.exp <-  ifelse(is.na(node.df.data$pct.exp), 0  , node.df.data$pct.exp)
    #     node.df.data$expr.value <-  ifelse(is.na(node.df.data$pct.exp), 0  , node.df.data$pct.exp)
    #     node.df.data$acc_value <-  ifelse(is.na(node.df.data$pct.exp), 0  , node.df.data$pct.exp)

        tf_peak_gene.node.list[[cid]] <- node.df.data
    }

 
    saveRDS(tf_peak_gene.node.list,paste(outdir,'/tf_peak_gene.node.list.rds',sep='') )


    write.table(tf_peak_gene.node.list[['c2']],file=paste(outdir,'/tf_peak_gene.node.c2.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)

    write.table(tf_peak_gene.node.list[['c1']],file=paste(outdir,'/tf_peak_gene.node.c1.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



    write.table(tf_peak_gene.node.list[['c9']],file=paste(outdir,'/tf_peak_gene.node.c9.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



    write.table(tf_peak_gene.node.list[['c5']],file=paste(outdir,'/tf_peak_gene.node.c5.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)



    write.table(tf_peak_gene.node.list[['c6']],file=paste(outdir,'/tf_peak_gene.node.c6.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)


    write.table(tf_peak_gene.node.list[['c3']],file=paste(outdir,'/tf_peak_gene.node.c3.txt',sep=''),sep='\t',col.names = TRUE, row.names = FALSE, quote = FALSE)

  return(paste('ok to outdir: ',outdir,'\n',sep=''))


}
